{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 6-arm study design\n",
    "- there is an observational factor with 3 values, which is age_group (young, middle-aged, elderly)\n",
    "- a crossover of two treatments, a drug and a biological treatment\n",
    "- three non-treatment phases: screen, washout and follow-up\n",
    "- two sample types colllected: blood and saliva\n",
    "- two assay types: metabolite profiling through (1) mass spectrometry on the saliva sample and (2) NMR spectroscopy on the blood samples. A combination of technical replicates and protocol parameters are defined in the assay configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOG: <Logger isatools (DEBUG)>\n"
     ]
    }
   ],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design_from_config\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'observationalFactors': [{'name': 'age group',\n",
       "   'values': ['young', 'middle-aged', 'elderly'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None}],\n",
       " 'generatedSubjectGroups': [{'name': 'SubjectGroup_0',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'young',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]},\n",
       "  {'name': 'SubjectGroup_1',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'middle-aged',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]},\n",
       "  {'name': 'SubjectGroup_2',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'elderly',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]}],\n",
       " 'selectedSubjectGroups': [{'name': 'SubjectGroup_0',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'young',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]},\n",
       "  {'name': 'SubjectGroup_1',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'middle-aged',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]},\n",
       "  {'name': 'SubjectGroup_2',\n",
       "   'type': 'Homo sapiens',\n",
       "   'characteristics': [{'name': 'age group',\n",
       "     'value': 'elderly',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}]}],\n",
       " 'treatmentPlan': {'screen': {'name': 'screen',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  'runIn': {'name': 'run-in', 'duration': None, 'durationUnit': ''},\n",
       "  'washout': {'name': 'washout', 'duration': 7, 'durationUnit': 'days'},\n",
       "  'followUp': {'name': 'follow-up', 'duration': 60, 'durationUnit': 'days'},\n",
       "  'treatments': [{'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'short_form': 'ERO_0000600',\n",
       "     'obo_id': 'ERO:0000600',\n",
       "     'label': 'drug intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg/day',\n",
       "    'duration': 7,\n",
       "    'durationUnit': 'days'},\n",
       "   {'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'short_form': 'ERO_0001553',\n",
       "     'obo_id': 'ERO:0001553',\n",
       "     'label': 'biologic intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'bacteriophages',\n",
       "    'intensity': 9,\n",
       "    'intensityUnit': 'log10 (phages)',\n",
       "    'duration': 7,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'interventionType': None,\n",
       "   'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'generatedStudyDesign': {'name': 'onto annotated study',\n",
       "  'description': 'study with ontology annotated assays and NMR assays',\n",
       "  'type': {'term': 'crossover design',\n",
       "   'id': 'OBI:0500003',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/OBI_0500003'},\n",
       "  'subjectType': 'Homo sapiens',\n",
       "  'elements': [{'id': '273194f1-fa4a-45a0-9bcf-09e2dfee9cf5',\n",
       "    'name': 'screen',\n",
       "    'duration': 5,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': '23bbffef-3a20-4f80-a0be-9dda20987dac',\n",
       "    'name': 'treatment_0',\n",
       "    'interventionType': {'term': 'drug intervention',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600'},\n",
       "    'duration': 7,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg/day'},\n",
       "   {'id': '6ee85545-0457-43f2-8fc7-812b3f63fca7',\n",
       "    'name': 'washout',\n",
       "    'duration': 7,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': 'ba86c401-bd3b-43fc-aae6-d58d59c3ef0a',\n",
       "    'name': 'treatment_1',\n",
       "    'interventionType': {'term': 'biologic intervention',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0001553'},\n",
       "    'duration': 7,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'bacteriophages',\n",
       "    'intensity': 9,\n",
       "    'intensityUnit': 'log10 (phages)'},\n",
       "   {'id': '27fdc605-2fdb-44eb-9b9b-d8210a1c2f71',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 60,\n",
       "    'durationUnit': 'days'}],\n",
       "  'events': [],\n",
       "  'arms': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'young',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "   {'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'young',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'middle-aged',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'middle-aged',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'elderly',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'observationalFactors': [{'name': 'age group',\n",
       "      'value': 'elderly',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "      'events': []},\n",
       "     {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "     {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "     {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "     {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]}]},\n",
       " 'selectedArms': [{'id': '#arm/0',\n",
       "   'name': 'Arm_0',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'young',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "  {'id': '#arm/1',\n",
       "   'name': 'Arm_1',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'young',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "  {'id': '#arm/2',\n",
       "   'name': 'Arm_2',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'middle-aged',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "  {'id': '#arm/3',\n",
       "   'name': 'Arm_3',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'middle-aged',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "  {'id': '#arm/4',\n",
       "   'name': 'Arm_4',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'elderly',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]},\n",
       "  {'id': '#arm/5',\n",
       "   'name': 'Arm_5',\n",
       "   'observationalFactors': [{'name': 'age group',\n",
       "     'value': 'elderly',\n",
       "     'unit': None,\n",
       "     'isQuantitative': False}],\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['273194f1-fa4a-45a0-9bcf-09e2dfee9cf5'],\n",
       "     'events': []},\n",
       "    {'elements': ['ba86c401-bd3b-43fc-aae6-d58d59c3ef0a'], 'events': []},\n",
       "    {'elements': ['6ee85545-0457-43f2-8fc7-812b3f63fca7'], 'events': []},\n",
       "    {'elements': ['23bbffef-3a20-4f80-a0be-9dda20987dac'], 'events': []},\n",
       "    {'elements': ['27fdc605-2fdb-44eb-9b9b-d8210a1c2f71'], 'events': []}]}],\n",
       " 'samplePlan': [{'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, True, False, True, True],\n",
       "    'Arm_1': [False, True, False, True, True],\n",
       "    'Arm_2': [False, True, False, True, True],\n",
       "    'Arm_3': [False, True, False, True, True],\n",
       "    'Arm_4': [False, True, False, True, True],\n",
       "    'Arm_5': [False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, 1, None, 1, 3],\n",
       "    'Arm_1': [None, 1, None, 1, 3],\n",
       "    'Arm_2': [None, 1, None, 1, 3],\n",
       "    'Arm_3': [None, 1, None, 1, 3],\n",
       "    'Arm_4': [None, 1, None, 1, 3],\n",
       "    'Arm_5': [None, 1, None, 1, 3]}},\n",
       "  {'sampleType': 'saliva',\n",
       "   'selectedCells': {'Arm_0': [True, False, False, False, True],\n",
       "    'Arm_1': [True, False, False, False, True],\n",
       "    'Arm_2': [True, False, False, False, True],\n",
       "    'Arm_3': [True, False, False, False, True],\n",
       "    'Arm_4': [True, False, False, False, True],\n",
       "    'Arm_5': [True, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, None, None, None, 3],\n",
       "    'Arm_1': [1, None, None, None, 3],\n",
       "    'Arm_2': [1, None, None, None, 3],\n",
       "    'Arm_3': [1, None, None, None, 3],\n",
       "    'Arm_4': [1, None, None, None, 3],\n",
       "    'Arm_5': [1, None, None, None, 3]}}],\n",
       " 'assayConfigs': [{'id': 0,\n",
       "   'name': 'metabolite profiling using mass spectrometry',\n",
       "   'icon': 'fas fa-chart-bar fas fa-bullseye',\n",
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'non-polar fraction'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labeling', {'#replicates': {'value': 1}}],\n",
       "    ['labeled extract',\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': 'labeled extract type',\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': []},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': []},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': []}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 1,\n",
       "   'name': 'metabolite profiling - mass spectrometry (ontology annotated)',\n",
       "   'icon': 'fas fa-chart-bar',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}],\n",
       "       'values': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 2}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': ['label_0']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None}]},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}],\n",
       "       'values': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}]},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}]}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': None},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [True, False, False, False, True],\n",
       "    'Arm_1': [True, False, False, False, True],\n",
       "    'Arm_2': [True, False, False, False, True],\n",
       "    'Arm_3': [True, False, False, False, True],\n",
       "    'Arm_4': [True, False, False, False, True],\n",
       "    'Arm_5': [True, False, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['saliva']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'metabolite profiling using NMR spectroscopy',\n",
       "   'icon': 'fas fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 2}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
       "       'values': ['supernatant', 'pellet']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nmr_spectroscopy',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Bruker Avance II 1 GHz'],\n",
       "       'values': ['Bruker Avance II 1 GHz']},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': ['1D 1H NMR']},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': ['CPMG', 'TOCSY']}}],\n",
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, True, False, True, True],\n",
       "    'Arm_1': [False, True, False, True, True],\n",
       "    'Arm_2': [False, True, False, True, True],\n",
       "    'Arm_3': [False, True, False, True, True],\n",
       "    'Arm_4': [False, True, False, True, True],\n",
       "    'Arm_5': [False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood']]}},\n",
       "  {'id': 3,\n",
       "   'name': 'transcription profiling by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'transcription profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['mRNA',\n",
       "        'total RNA',\n",
       "        'ncRNA',\n",
       "        'tRNA'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 4,\n",
       "   'name': 'DNA-protein binding site identification by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'DNA-protein binding site identification profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'cross linking': {'options': ['uv-light',\n",
       "        'formaldehyde',\n",
       "        'di-tert-butyl peroxide',\n",
       "        'formaldehyde and di-tert-butyl peroxyde'],\n",
       "       'values': []},\n",
       "      'DNA fragmentation': {'options': ['sonication',\n",
       "        'nebulization',\n",
       "        'micrococcal nuclease digestion',\n",
       "        'deoxyribonuclease digestion'],\n",
       "       'values': []},\n",
       "      'DNA fragment size': {'options': ['a', 'b'], 'values': []},\n",
       "      'immunoprecipitation antibody': {'options': ['c', 'd'], 'values': []}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []},\n",
       "      'library_strategy': {'options': ['Bisulfite-Seq',\n",
       "        'MRE-Seq',\n",
       "        'MeDIP-Seq',\n",
       "        'MBD-Seq',\n",
       "        'OTHER'],\n",
       "       'values': []},\n",
       "      'library_selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'HMPR',\n",
       "        'MF',\n",
       "        'MSLL',\n",
       "        '5-methylcytidine antibody',\n",
       "        'MBD2 protein methyl-CpG binding domain',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': []},\n",
       "      'multiplex identifier': {'options': ['e', 'f', 'g'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 5,\n",
       "   'name': 'DNA methylation profiling  by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'DNA methylation profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'cross linking': {'options': ['uv-light',\n",
       "        'formaldehyde',\n",
       "        'di-tert-butyl peroxide',\n",
       "        'formaldehyde and di-tert-butyl peroxyde'],\n",
       "       'values': []},\n",
       "      'DNA fragmentation': {'options': ['sonication',\n",
       "        'nebulization',\n",
       "        'micrococcal nuclease digestion',\n",
       "        'deoxyribonuclease digestion'],\n",
       "       'values': []},\n",
       "      'DNA fragment size': {'options': ['a', 'b'], 'values': []},\n",
       "      'immunoprecipitation antibody': {'options': ['c', 'd'], 'values': []}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []},\n",
       "      'library_strategy': {'options': ['Bisulfite-Seq',\n",
       "        'MRE-Seq',\n",
       "        'MeDIP-Seq',\n",
       "        'MBD-Seq',\n",
       "        'OTHER'],\n",
       "       'values': []},\n",
       "      'library_selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'HMPR',\n",
       "        'MF',\n",
       "        'MSLL',\n",
       "        '5-methylcytidine antibody',\n",
       "        'MBD2 protein methyl-CpG binding domain',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': []},\n",
       "      'multiplex identifier': {'options': ['e', 'f', 'g'], 'values': ''}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 6,\n",
       "   'name': 'histone modification profiling by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'histone modification profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'cross linking': {'options': ['uv-light',\n",
       "        'formaldehyde',\n",
       "        'di-tert-butyl peroxide',\n",
       "        'formaldehyde and di-tert-butyl peroxyde'],\n",
       "       'values': []},\n",
       "      'DNA fragmentation': {'options': ['sonication',\n",
       "        'nebulization',\n",
       "        'micrococcal nuclease digestion',\n",
       "        'deoxyribonuclease digestion'],\n",
       "       'values': []},\n",
       "      'DNA fragment size': {'options': ['a', 'b'], 'values': []},\n",
       "      'immunoprecipitation antibody': {'options': ['c', 'd'], 'values': []}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []},\n",
       "      'library_strategy': {'options': ['Bisulfite-Seq',\n",
       "        'MRE-Seq',\n",
       "        'MeDIP-Seq',\n",
       "        'MBD-Seq',\n",
       "        'OTHER'],\n",
       "       'values': []},\n",
       "      'library_selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'HMPR',\n",
       "        'MF',\n",
       "        'MSLL',\n",
       "        '5-methylcytidine antibody',\n",
       "        'MBD2 protein methyl-CpG binding domain',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': []},\n",
       "      'multiplex identifier': {'options': ['e', 'f', 'g'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 7,\n",
       "   'name': 'metagenome sequencing by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'metagenome sequencing',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []},\n",
       "      'library_strategy': {'options': ['AMPLICON', 'OTHER'], 'values': []},\n",
       "      'library_selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'Restriction Digest',\n",
       "        'size fractionation',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 8,\n",
       "   'name': 'genome sequencing by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'genome sequencing',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []},\n",
       "      'library_strategy': {'options': ['AMPLICON',\n",
       "        'CLONE',\n",
       "        'WGA',\n",
       "        'WGS',\n",
       "        'WXS',\n",
       "        'OTHER'],\n",
       "       'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 9,\n",
       "   'name': 'environmental gene survey by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'environmental gene survey',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['DNA'], 'values': ['DNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'target_taxon': {'options': ['Archeae', 'Bacteria', 'Eukaryota'],\n",
       "       'values': []},\n",
       "      'target_gene': {'options': ['16S rRNA',\n",
       "        '18S rRNA',\n",
       "        'RBCL',\n",
       "        'matK',\n",
       "        'COX1',\n",
       "        'ITS1-5.8S-ITS2'],\n",
       "       'values': []},\n",
       "      'target_subfragment': {'options': ['V6', 'V9', 'ITS'], 'values': []},\n",
       "      'pcr_primers': {'options': [''], 'values': []},\n",
       "      'pcr_cond': {'options': ['unspecified'], 'values': []},\n",
       "      'multiplex identifier': {'options': ['e', 'f', 'g'], 'values': []},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': []},\n",
       "      'library_orientation': {'options': ['single', 'paired'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 10,\n",
       "   'name': 'clinical chemistry by marker panel',\n",
       "   'icon': 'fa fa-flask',\n",
       "   'color': 'yellow',\n",
       "   'measurement_type': 'clinical chemistry',\n",
       "   'technology_type': 'marker panel',\n",
       "   'workflow': [['sample preparation', {'#replicates': {'value': 1}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 11,\n",
       "   'name': 'hematology by marker panel',\n",
       "   'icon': 'fa fa-flask',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'clinical chemistry',\n",
       "   'technology_type': 'marker panel',\n",
       "   'workflow': [['sample preparation', {'#replicates': {'value': 1}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 12,\n",
       "   'name': 'transcription profiling by DNA microarray',\n",
       "   'icon': 'fa fa-braille',\n",
       "   'color': 'blue-10',\n",
       "   'measurement_type': {'term': 'transcription profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000424',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'DNA microarray',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0400148',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'total RNA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/OBI_0000895',\n",
       "         'source': 'OBI'},\n",
       "        {'term': 'mRNA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/SO_0000234',\n",
       "         'source': 'SO'},\n",
       "        {'term': 'siRNA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/SO_0000646',\n",
       "         'source': 'SO'},\n",
       "        {'term': 'ncRNA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/SO_0000655',\n",
       "         'source': 'SO'},\n",
       "        {'term': 'other', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': 'CHMO'},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': 'OBI'},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'biotin',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/CHEBI_15956',\n",
       "         'source': 'CHEBI'},\n",
       "        {'term': 'cy-3',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/CHEBI_37987',\n",
       "         'source': 'CHEBI'},\n",
       "        {'term': 'cy-5',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/CHEBI_37989',\n",
       "         'source': 'CHEBI'}],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid hybridization',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'platform': {'options': [{'term': 'Agilent',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/OBI_0000765',\n",
       "         'source': 'OBI'},\n",
       "        {'term': 'Affymetrix',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/OBI_0000462',\n",
       "         'source': 'OBI'},\n",
       "        {'term': 'Illumina',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/OBI_0000759',\n",
       "         'source': 'OBI'},\n",
       "        {'term': 'other', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'array_design': {'options': ['GeneChip™ Human Transcriptome Assay 2.0',\n",
       "        'GeneChip™ Human Gene 2.0 ST Assay',\n",
       "        'GeneChip™ Rat Exon 1.0 ST Array',\n",
       "        'GeneChip™ Mouse Gene 1.0 ST Array',\n",
       "        'Clariom™ D Assay, human',\n",
       "        'GeneChip™ miRNA 4.0 Assay',\n",
       "        'SurePrint G3 Human Gene Expression 8x60K v2 Microarray Kit',\n",
       "        'SurePrint G3 Human Gene Expression v3 8x60K Microarray Kit',\n",
       "        'Human GE 4x44K v2 Microarray Kit'],\n",
       "       'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 13,\n",
       "   'name': 'copy number variation analysis by DNA microarray',\n",
       "   'icon': 'fa fa-braille',\n",
       "   'color': 'blue-10',\n",
       "   'measurement_type': 'copy number variation analysis',\n",
       "   'technology_type': 'DNA microarray',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['total RNA',\n",
       "        'mRNA',\n",
       "        'siRNA',\n",
       "        'ncRNA',\n",
       "        'other'],\n",
       "       'values': ['']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['Biotin', 'Cy-3', 'Cy-5'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid hybridization',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'platform': {'options': ['Illumina', 'Agilent', 'Affymetrix', 'Other'],\n",
       "       'values': []},\n",
       "      'array_design': {'options': ['Infinium Exome-24 Kit',\n",
       "        'Infinium CoreExome-24 Kit',\n",
       "        'Infinium Omni2.5Exome',\n",
       "        'Infinium Omni5Exome',\n",
       "        '',\n",
       "        'Genome-Wide Human SNP Array 6.0',\n",
       "        'Axiom™ Genome-Wide Human Origins 1 Array Plate ',\n",
       "        ''],\n",
       "       'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 14,\n",
       "   'name': 'cell sorting using flow cytometry',\n",
       "   'icon': 'fa fa-filter',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'cell sorting',\n",
       "   'technology_type': 'flow cytometry',\n",
       "   'workflow': [[{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['Biotin', 'Cy-3', 'Cy-5'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['data collection',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Becton Dickinson',\n",
       "        'Beckman Coulter',\n",
       "        'Other'],\n",
       "       'values': []},\n",
       "      'optical_path': {'options': ['', 'NA'], 'values': []},\n",
       "      'detector voltage': {'options': ['', 'NA'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}},\n",
       "  {'id': 15,\n",
       "   'name': 'cell counting using flow cytometry',\n",
       "   'icon': 'fa fa-filter',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'cell counting',\n",
       "   'technology_type': 'flow cytometry',\n",
       "   'workflow': [[{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['Biotin', 'Cy-3', 'Cy-5'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['data collection',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Becton Dickinson',\n",
       "        'Beckman Coulter',\n",
       "        'Other'],\n",
       "       'values': []},\n",
       "      'optical_path': {'options': ['', 'NA'], 'values': []},\n",
       "      'detector voltage': {'options': ['', 'NA'], 'values': []}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False],\n",
       "    'Arm_1': [False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False],\n",
       "    'Arm_5': [False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood', 'saliva']],\n",
       "    'Arm_1': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_2': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_3': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_4': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']],\n",
       "    'Arm_5': [['saliva'], ['blood'], [], ['blood'], ['blood', 'saliva']]}}],\n",
       " 'selectedAssayTypes': {'metabolite profiling using mass spectrometry': False,\n",
       "  'metabolite profiling - mass spectrometry (ontology annotated)': True,\n",
       "  'metabolite profiling using NMR spectroscopy': True,\n",
       "  'transcription profiling by nucleotide sequencing': False,\n",
       "  'DNA-protein binding site identification by nucleotide sequencing': False,\n",
       "  'DNA methylation profiling  by nucleotide sequencing': False,\n",
       "  'histone modification profiling by nucleotide sequencing': False,\n",
       "  'metagenome sequencing by nucleotide sequencing': False,\n",
       "  'genome sequencing by nucleotide sequencing': False,\n",
       "  'environmental gene survey by nucleotide sequencing': False,\n",
       "  'clinical chemistry by marker panel': False,\n",
       "  'hematology by marker panel': False,\n",
       "  'transcription profiling by DNA microarray': False,\n",
       "  'copy number variation analysis by DNA microarray': False,\n",
       "  'cell sorting using flow cytometry': False,\n",
       "  'cell counting using flow cytometry': False},\n",
       " 'assayInTab': 'metabolite profiling using NMR spectroscopy'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"study-design-crossover-onto-annotated-ms-and-nnmr.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design_from_config` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design_from_config(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 6.45 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 1.72 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output'))\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "with open(os.path.abspath(os.path.join('output','isa-investigation-2-arms-nmr-ms.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 96.57 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, os.path.abspath(os.path.join('output')))\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 1 assay file (is this correct? check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT2_metabolite-profiling_NMR-spectroscopy.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT1_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 10 subjects in the each of the four selected arms (Arm_0, Arm_2, Arm_3, Arm_4) and 24 samples have been collected (3 blood samples per subject during the follow-up epoch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 90 samples in the GRP0 arm (i.e. group)\n",
      "There are 90 samples in the GRP2 arm (i.e. group)\n",
      "There are 90 samples in the GRP3 arm (i.e. group)\n",
      "There are 90 samples in the GRP4 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])\n",
    "count_arm4_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP4' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))\n",
    "print(\"There are {} samples in the GRP4 arm (i.e. group)\".format(count_arm4_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Labeled Extract Name</th>\n",
       "      <th>Label</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Term Accession Number.1</th>\n",
       "      <th>Parameter Value[injection_mode]</th>\n",
       "      <th>Term Accession Number.2</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Term Accession Number.3</th>\n",
       "      <th>MS Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Spectral Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-saliva-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-LE3</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-6-&lt;acquisition&gt;10</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-6-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-saliva-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-LE3</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-6-&lt;acquisition&gt;9</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-6-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-saliva-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-LE4</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-6-&lt;acquisition&gt;16</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-6-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-saliva-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-LE4</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-6-&lt;acquisition&gt;15</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-6-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-saliva-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-Extract1</td>\n",
       "      <td>lipids</td>\n",
       "      <td>lipids</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-6-LE1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-6-&lt;acquisition&gt;2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3835</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-saliva-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-LE4</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-233-&lt;acquisition&gt;13</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-233-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3836</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-saliva-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-Extract1</td>\n",
       "      <td>lipids</td>\n",
       "      <td>lipids</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-LE2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-233-&lt;acquisition&gt;6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-233-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3837</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-saliva-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-LE4</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-233-&lt;acquisition&gt;16</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-233-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3838</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-saliva-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-Extract2</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-LE3</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-233-&lt;acquisition&gt;10</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-233-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3839</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-saliva-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-Extract1</td>\n",
       "      <td>lipids</td>\n",
       "      <td>lipids</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-233-LE2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>mass-spectrometry_AT1-233-&lt;acquisition&gt;7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw-spectral-data-file_AT1-233-7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3840 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sample Name Protocol REF Performer      Extract Name  \\\n",
       "0     GRP0_SBJ01_A0E0_SMP-saliva-1   extraction   Unknown    AT1-6-Extract2   \n",
       "1     GRP0_SBJ01_A0E0_SMP-saliva-1   extraction   Unknown    AT1-6-Extract2   \n",
       "2     GRP0_SBJ01_A0E0_SMP-saliva-1   extraction   Unknown    AT1-6-Extract2   \n",
       "3     GRP0_SBJ01_A0E0_SMP-saliva-1   extraction   Unknown    AT1-6-Extract2   \n",
       "4     GRP0_SBJ01_A0E0_SMP-saliva-1   extraction   Unknown    AT1-6-Extract1   \n",
       "...                            ...          ...       ...               ...   \n",
       "3835  GRP5_SBJ10_A5E4_SMP-saliva-3   extraction   Unknown  AT1-233-Extract2   \n",
       "3836  GRP5_SBJ10_A5E4_SMP-saliva-3   extraction   Unknown  AT1-233-Extract1   \n",
       "3837  GRP5_SBJ10_A5E4_SMP-saliva-3   extraction   Unknown  AT1-233-Extract2   \n",
       "3838  GRP5_SBJ10_A5E4_SMP-saliva-3   extraction   Unknown  AT1-233-Extract2   \n",
       "3839  GRP5_SBJ10_A5E4_SMP-saliva-3   extraction   Unknown  AT1-233-Extract1   \n",
       "\n",
       "     Characteristics[extract type] Term Accession Number Protocol REF.1  \\\n",
       "0                   polar fraction        polar fraction       labeling   \n",
       "1                   polar fraction        polar fraction       labeling   \n",
       "2                   polar fraction        polar fraction       labeling   \n",
       "3                   polar fraction        polar fraction       labeling   \n",
       "4                           lipids                lipids       labeling   \n",
       "...                            ...                   ...            ...   \n",
       "3835                polar fraction        polar fraction       labeling   \n",
       "3836                        lipids                lipids       labeling   \n",
       "3837                polar fraction        polar fraction       labeling   \n",
       "3838                polar fraction        polar fraction       labeling   \n",
       "3839                        lipids                lipids       labeling   \n",
       "\n",
       "     Performer.1 Labeled Extract Name    Label     Protocol REF.2  \\\n",
       "0        Unknown            AT1-6-LE3  label_0  mass spectrometry   \n",
       "1        Unknown            AT1-6-LE3  label_0  mass spectrometry   \n",
       "2        Unknown            AT1-6-LE4  label_0  mass spectrometry   \n",
       "3        Unknown            AT1-6-LE4  label_0  mass spectrometry   \n",
       "4        Unknown            AT1-6-LE1  label_0  mass spectrometry   \n",
       "...          ...                  ...      ...                ...   \n",
       "3835     Unknown          AT1-233-LE4  label_0  mass spectrometry   \n",
       "3836     Unknown          AT1-233-LE2  label_0  mass spectrometry   \n",
       "3837     Unknown          AT1-233-LE4  label_0  mass spectrometry   \n",
       "3838     Unknown          AT1-233-LE3  label_0  mass spectrometry   \n",
       "3839     Unknown          AT1-233-LE2  label_0  mass spectrometry   \n",
       "\n",
       "     Parameter Value[instrument]                    Term Accession Number.1  \\\n",
       "0              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "2              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "4              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "...                          ...                                        ...   \n",
       "3835           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3836           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3837           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3838           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3839           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "\n",
       "     Parameter Value[injection_mode]  \\\n",
       "0                                FIA   \n",
       "1                                FIA   \n",
       "2                                 LC   \n",
       "3                                 LC   \n",
       "4                                 LC   \n",
       "...                              ...   \n",
       "3835                             FIA   \n",
       "3836                              LC   \n",
       "3837                              LC   \n",
       "3838                             FIA   \n",
       "3839                             FIA   \n",
       "\n",
       "                        Term Accession Number.2  \\\n",
       "0     http://purl.obolibrary.org/obo/MS_1000058   \n",
       "1     http://purl.obolibrary.org/obo/MS_1000058   \n",
       "2                                                 \n",
       "3                                                 \n",
       "4                                                 \n",
       "...                                         ...   \n",
       "3835  http://purl.obolibrary.org/obo/MS_1000058   \n",
       "3836                                              \n",
       "3837                                              \n",
       "3838  http://purl.obolibrary.org/obo/MS_1000058   \n",
       "3839  http://purl.obolibrary.org/obo/MS_1000058   \n",
       "\n",
       "     Parameter Value[acquisition_mode]  \\\n",
       "0                        positive mode   \n",
       "1                        positive mode   \n",
       "2                        positive mode   \n",
       "3                        positive mode   \n",
       "4                        positive mode   \n",
       "...                                ...   \n",
       "3835                     positive mode   \n",
       "3836                     positive mode   \n",
       "3837                     positive mode   \n",
       "3838                     positive mode   \n",
       "3839                     positive mode   \n",
       "\n",
       "                        Term Accession Number.3  \\\n",
       "0     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "2     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "4     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "...                                         ...   \n",
       "3835  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3836  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3837  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3838  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3839  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "\n",
       "                                  MS Assay Name Performer.2  \\\n",
       "0       mass-spectrometry_AT1-6-<acquisition>10     Unknown   \n",
       "1        mass-spectrometry_AT1-6-<acquisition>9     Unknown   \n",
       "2       mass-spectrometry_AT1-6-<acquisition>16     Unknown   \n",
       "3       mass-spectrometry_AT1-6-<acquisition>15     Unknown   \n",
       "4        mass-spectrometry_AT1-6-<acquisition>2     Unknown   \n",
       "...                                         ...         ...   \n",
       "3835  mass-spectrometry_AT1-233-<acquisition>13     Unknown   \n",
       "3836   mass-spectrometry_AT1-233-<acquisition>6     Unknown   \n",
       "3837  mass-spectrometry_AT1-233-<acquisition>16     Unknown   \n",
       "3838  mass-spectrometry_AT1-233-<acquisition>10     Unknown   \n",
       "3839   mass-spectrometry_AT1-233-<acquisition>7     Unknown   \n",
       "\n",
       "                 Raw Spectral Data File  \n",
       "0       raw-spectral-data-file_AT1-6-10  \n",
       "1        raw-spectral-data-file_AT1-6-9  \n",
       "2       raw-spectral-data-file_AT1-6-16  \n",
       "3       raw-spectral-data-file_AT1-6-15  \n",
       "4        raw-spectral-data-file_AT1-6-2  \n",
       "...                                 ...  \n",
       "3835  raw-spectral-data-file_AT1-233-13  \n",
       "3836   raw-spectral-data-file_AT1-233-6  \n",
       "3837  raw-spectral-data-file_AT1-233-16  \n",
       "3838  raw-spectral-data-file_AT1-233-10  \n",
       "3839   raw-spectral-data-file_AT1-233-7  \n",
       "\n",
       "[3840 rows x 20 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Parameter Value[pulse_sequence]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Free Induction Decay Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-13-Extract1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-13-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-12-Extract2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-12-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-12-Extract1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-12-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-13-Extract2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-13-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-12-Extract2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-12-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4795</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-blood-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-587-Extract1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-587-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4796</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-blood-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-586-Extract2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-586-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4797</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-blood-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-586-Extract1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-586-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4798</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-blood-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-586-Extract1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-586-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4799</th>\n",
       "      <td>GRP5_SBJ10_A5E4_SMP-blood-3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-587-Extract2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 1H NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>raw_spectral_data_file_AT2-587-6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4800 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Protocol REF Performer      Extract Name  \\\n",
       "0     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown   AT2-13-Extract1   \n",
       "1     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown   AT2-12-Extract2   \n",
       "2     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown   AT2-12-Extract1   \n",
       "3     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown   AT2-13-Extract2   \n",
       "4     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown   AT2-12-Extract2   \n",
       "...                           ...          ...       ...               ...   \n",
       "4795  GRP5_SBJ10_A5E4_SMP-blood-3   extraction   Unknown  AT2-587-Extract1   \n",
       "4796  GRP5_SBJ10_A5E4_SMP-blood-3   extraction   Unknown  AT2-586-Extract2   \n",
       "4797  GRP5_SBJ10_A5E4_SMP-blood-3   extraction   Unknown  AT2-586-Extract1   \n",
       "4798  GRP5_SBJ10_A5E4_SMP-blood-3   extraction   Unknown  AT2-586-Extract1   \n",
       "4799  GRP5_SBJ10_A5E4_SMP-blood-3   extraction   Unknown  AT2-587-Extract2   \n",
       "\n",
       "     Characteristics[extract type]    Protocol REF.1  \\\n",
       "0                      supernatant  nmr_spectroscopy   \n",
       "1                           pellet  nmr_spectroscopy   \n",
       "2                      supernatant  nmr_spectroscopy   \n",
       "3                           pellet  nmr_spectroscopy   \n",
       "4                           pellet  nmr_spectroscopy   \n",
       "...                            ...               ...   \n",
       "4795                   supernatant  nmr_spectroscopy   \n",
       "4796                        pellet  nmr_spectroscopy   \n",
       "4797                   supernatant  nmr_spectroscopy   \n",
       "4798                   supernatant  nmr_spectroscopy   \n",
       "4799                        pellet  nmr_spectroscopy   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[acquisition_mode]  \\\n",
       "0         Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "1         Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "2         Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "3         Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "4         Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "...                          ...                               ...   \n",
       "4795      Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "4796      Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "4797      Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "4798      Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "4799      Bruker Avance II 1 GHz                         1D 1H NMR   \n",
       "\n",
       "     Parameter Value[pulse_sequence] Performer.1  \\\n",
       "0                              TOCSY     Unknown   \n",
       "1                               CPMG     Unknown   \n",
       "2                               CPMG     Unknown   \n",
       "3                              TOCSY     Unknown   \n",
       "4                               CPMG     Unknown   \n",
       "...                              ...         ...   \n",
       "4795                            CPMG     Unknown   \n",
       "4796                            CPMG     Unknown   \n",
       "4797                            CPMG     Unknown   \n",
       "4798                           TOCSY     Unknown   \n",
       "4799                           TOCSY     Unknown   \n",
       "\n",
       "        Free Induction Decay Data File  \n",
       "0      raw_spectral_data_file_AT2-13-2  \n",
       "1      raw_spectral_data_file_AT2-12-7  \n",
       "2      raw_spectral_data_file_AT2-12-4  \n",
       "3      raw_spectral_data_file_AT2-13-6  \n",
       "4      raw_spectral_data_file_AT2-12-8  \n",
       "...                                ...  \n",
       "4795  raw_spectral_data_file_AT2-587-3  \n",
       "4796  raw_spectral_data_file_AT2-586-7  \n",
       "4797  raw_spectral_data_file_AT2-586-3  \n",
       "4798  raw_spectral_data_file_AT2-586-1  \n",
       "4799  raw_spectral_data_file_AT2-587-6  \n",
       "\n",
       "[4800 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT2_metabolite-profiling_NMR-spectroscopy.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4800, 11)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT2_metabolite-profiling_NMR-spectroscopy.txt'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
