{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration with chained protocols in the assay workflow\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm repeated measures (crossover) study design\n",
    "- 4 subject groups: \"M\" and \"F\", \"healthy\" and \"diseased\"\n",
    "- 2 treatments in crossover: a drug treatment with ibuprofen and a biological treatment with injection of KpJH46Φ2 bacteriophage\n",
    "- 2 samples collected: blood and derma samples\n",
    "- three assays performed: mass. spec. on blood samples, Chip-seq and NMR on derma samples. Please note that the Chip-seq assay contains two chained protocols in its workflow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
<<<<<<< HEAD
    "from isatools.create.connectors import generate_study_design_from_config\n",
=======
    "from isatools.create.connectors import generate_study_design\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 207,
=======
   "execution_count": 4,
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 208,
=======
   "execution_count": 5,
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'Chained protocols study',\n",
       " 'subjectType': 'Homo sapiens',\n",
       " 'subjectSize': 10,\n",
       " 'designType': {'term': 'crossover design',\n",
       "  'id': 'OBI:0500003',\n",
       "  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',\n",
       "  'label': 'Study subjects receive repeated treatments',\n",
       "  'value': 'crossover'},\n",
       " 'observationalFactors': [{'name': 'sex',\n",
       "   'values': ['M', 'F'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None},\n",
       "  {'name': 'condition',\n",
       "   'values': ['healthy', 'diseased'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None}],\n",
       " 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_1',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_2',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_3',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]}],\n",
       "  'unselected': []},\n",
       " 'treatmentPlan': {'screen': {'selected': True,\n",
       "   'name': 'screen',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  'runIn': {'selected': True,\n",
       "   'name': 'run-in',\n",
       "   'duration': 2,\n",
       "   'durationUnit': 'days'},\n",
       "  'washout': {'selected': True,\n",
       "   'name': 'washout',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  'followUp': {'selected': True,\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'},\n",
       "  'treatments': [{'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'short_form': 'ERO_0000600',\n",
       "     'obo_id': 'ERO:0000600',\n",
       "     'label': 'drug intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg/day',\n",
       "    'duration': 12,\n",
       "    'durationUnit': 'days'},\n",
<<<<<<< HEAD
       "   {'agent': 'aspirin',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
=======
       "   {'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'short_form': 'ERO_0001553',\n",
       "     'obo_id': 'ERO:0001553',\n",
       "     'label': 'biologic intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'KpJH46Φ2 phage',\n",
       "    'intensity': 3,\n",
       "    'intensityUnit': 'injections/day',\n",
       "    'duration': 14,\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'interventionType': None,\n",
       "   'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
<<<<<<< HEAD
       " 'generatedStudyDesign': {'name': 'test study',\n",
       "  'description': 'this is the verbose description',\n",
       "  'type': {'term': 'crossover design',\n",
       "   'id': 'OBI:0500003',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/OBI_0500003'},\n",
       "  'elements': [{'id': '83978887-be4c-4cac-afb6-0faad3d07c86',\n",
       "    'name': 'treatment_0',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': 'e283a7c5-955f-4d3b-b99e-0053a2af5a61',\n",
       "    'name': 'washout',\n",
       "    'duration': 5,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': '696a2da4-329a-4b39-a681-e402afb4e89f',\n",
       "    'name': 'treatment_1',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'aspirin',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '6bf4662d-852f-44ad-9b46-acff5291c47f',\n",
       "    'name': 'treatment_2',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 60,\n",
       "    'durationUnit': 'days'}],\n",
       "  'events': [],\n",
       "  'arms': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
=======
       " 'elements': [{'id': 'd9c13049-a235-41e2-a3be-f929e54ee2ee',\n",
       "   'name': 'screen',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '39e80594-b9ee-47f2-baf1-b954d7c1b124',\n",
       "   'name': 'run-in',\n",
       "   'duration': 2,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': 'ca5c2ebf-facf-46af-b698-99b81a28d88b',\n",
       "   'name': 'treatment_0',\n",
       "   'interventionType': {'term': 'drug intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000600'},\n",
       "   'duration': 12,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'ibuprofen',\n",
       "   'intensity': 8,\n",
       "   'intensityUnit': 'mg/day'},\n",
       "  {'id': 'b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7',\n",
       "   'name': 'washout',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '200e4d48-4ab2-492c-abd8-7b59a01b39d5',\n",
       "   'name': 'treatment_1',\n",
       "   'interventionType': {'term': 'biologic intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0001553'},\n",
       "   'duration': 14,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'KpJH46Φ2 phage',\n",
       "   'intensity': 3,\n",
       "   'intensityUnit': 'injections/day'},\n",
       "  {'id': 'e5664aeb-182f-49e2-849b-82e4c5e269e3',\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'}],\n",
       " 'arms': {'selected': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/7',\n",
       "    'name': 'Arm_7',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]}],\n",
       "  'unselected': [{'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
<<<<<<< HEAD
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
=======
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
<<<<<<< HEAD
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}]},\n",
       " 'selectedArms': [{'id': '#arm/0',\n",
       "   'name': 'Arm_0',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/2',\n",
       "   'name': 'Arm_2',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/4',\n",
       "   'name': 'Arm_4',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/3',\n",
       "   'name': 'Arm_3',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}],\n",
       " 'samplePlan': [{'sampleType': 'saliva',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, None, 1, None, 1, 1],\n",
       "    'Arm_2': [1, None, 1, None, 1, 1],\n",
       "    'Arm_4': [1, None, 1, None, 1, 1],\n",
       "    'Arm_3': [1, None, 1, None, 1, 1]}},\n",
       "  {'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_4': [False, False, False, False, False, True],\n",
       "    'Arm_3': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 1],\n",
       "    'Arm_2': [None, None, None, None, None, 1],\n",
       "    'Arm_4': [None, None, None, None, None, 1],\n",
       "    'Arm_3': [None, None, None, None, None, 2]}}],\n",
       " 'assayConfigs': [{'id': 0,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec.',\n",
       "   'icon': 'fa fa-dna',\n",
=======
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/6',\n",
       "    'name': 'Arm_6',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]}]},\n",
       " 'samplePlan': [{'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, True, True, False, True, True],\n",
       "    'Arm_2': [False, True, True, False, True, True],\n",
       "    'Arm_5': [False, True, True, False, True, True],\n",
       "    'Arm_7': [False, True, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_2': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_5': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_7': [None, 1, 1, None, 1, 6]}},\n",
       "  {'sampleType': 'derma',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 2],\n",
       "    'Arm_2': [None, None, None, None, None, 2],\n",
       "    'Arm_5': [None, None, None, None, None, 2],\n",
       "    'Arm_7': [None, None, None, None, None, 2]}}],\n",
       " 'assayPlan': [{'id': 0,\n",
       "   'name': 'metabolite profiling using mass spectrometry',\n",
       "   'icon': 'fas fa-chart-bar fas fa-bullseye',\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 2}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'non-polar fraction'],\n",
       "       'values': ['polar fraction', 'non-polar fraction']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labeling', {'#replicates': {'value': 1}}],\n",
       "    ['labeled extract',\n",
       "     {'node_type': 'labeled extract',\n",
<<<<<<< HEAD
       "      'characteristics_category': 'labelled extract type',\n",
       "      'characteristics_value': {'options': ['biotin', 'cy-5', 'cy-3'],\n",
       "       'values': ['biotin']},\n",
=======
       "      'characteristics_category': 'labeled extract type',\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': ['label_0']},\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': ['Agilent QTQF 6510']},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': ['LC']},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': ['positive mode', 'negative mode']}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
<<<<<<< HEAD
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 1,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec. (Annotated)',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}],\n",
       "       'values': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labelled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'biotin',\n",
       "         'iri': 'biotin',\n",
       "         'source': None},\n",
       "        {'term': 'cy-5', 'iri': 'cy-5', 'source': None},\n",
       "        {'term': 'cy-3', 'iri': 'cy-3', 'source': None}],\n",
       "       'values': [{'term': 'cy-3', 'iri': 'cy-3', 'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}]},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': 'LC', 'source': None}],\n",
       "       'values': []},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': 'MS'}],\n",
       "       'values': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': 'MS'}]}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': 'MS'},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
=======
       "   'selectedCells': {'Arm_0': [False, True, True, False, False, True],\n",
       "    'Arm_2': [False, True, True, False, False, True],\n",
       "    'Arm_5': [False, True, True, False, False, True],\n",
       "    'Arm_7': [False, True, True, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['blood']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'metabolite profiling using NMR spectroscopy',\n",
       "   'icon': 'fas fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
<<<<<<< HEAD
       "       'values': ['pellet']},\n",
=======
       "       'values': ['supernatant', 'pellet']},\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'nmr sample', 'iri': None, 'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'NMR tube type': {'description': '',\n",
       "       'options': [{'term': '4 mm standard MAS rotor',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000379',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000393',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '2.5 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000375',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000372',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 7 in (Wilmad-LabGlass)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000387',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm Shigemi D2O matched',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000390',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm short',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000391',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000373',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000377',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000378',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4 mm Zirconia MAS rotor',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000380',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 541-pp',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000386',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '10 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000394',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4.25 mm (Bruker BioSpin)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000381',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000389',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (Norell Inc.)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000384',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (Wilmad-LabGlass)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000385',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4.25 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000382',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 7 in 507-grade',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000388',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (New Era)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000383',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm ST500-7 (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000392',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000376',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000374',\n",
       "         'source': 'Metabolights.owl'}],\n",
       "       'values': ['1.7 mm (Norell)']},\n",
       "      'NMR solvent': {'description': '',\n",
       "       'options': [{'term': 'CD3OD + phosphate sodium buffer D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000363',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.01 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000348',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 5 mM maleic acid',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000369',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000352',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000350',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate-buffered saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001163',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'H20 + D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001157',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001154',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 0.128 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001160',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.05 mM phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001145',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001143',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001155',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001150',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001164',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.154 M saline D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001156',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001149',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001153',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001165',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.01 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001147',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD + phosphate sodium buffer D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001159',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '10% D2O/0.04 mM sodium phosphate buffer/0.4 mM TMSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001162',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001151',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate sodium buffer',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001144',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.04 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001148',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 5 mM maleic acid ',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001158',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.9% NaCl in D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001142',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CDCL3',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001141',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.15 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001146',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.3 M saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001152',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.5 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001161',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000359',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.9% NaCl in D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000366',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate-buffered saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000368',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000354',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.15 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000351',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'H20 + D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000362',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 0.128 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000364',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CDCL3',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000361',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000371',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000356',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.154 M saline D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000370',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.05 mM phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000347',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000358',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000355',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.3 M saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000367',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.5 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000353',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.04 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000357',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate sodium buffer',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000360',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000349',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000365',\n",
       "         'source': 'Metabolights.owl'}],\n",
       "       'values': [{'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000365',\n",
       "         'source': 'Metabolights.owl'}]},\n",
       "      'NMR Sample pH': {'description': '', 'options': [], 'values': ['7.0']},\n",
       "      'NMR Temperature': {'description': '',\n",
       "       'options': [],\n",
       "       'values': ['65 degree C']}}],\n",
       "    ['nmr_spectroscopy',\n",
<<<<<<< HEAD
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Bruker AvanceII 1 GHz'],\n",
       "       'values': ['Bruker AvanceII 1 GHz']},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': ['1D 13C NMR']},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': ['CPMG']}}],\n",
=======
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Bruker Avance II 1 GHz'],\n",
       "       'values': ['Bruker Avance II 1 GHz']},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': ['1D 13C NMR', '2D 13C-13C NMR']},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': ['TOCSY', 'watergate']}}],\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
<<<<<<< HEAD
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 14,\n",
       "   'name': 'RNA-Seq',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'transcriptomics profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['mRNA',\n",
       "        'total RNA',\n",
       "        'ncRNA',\n",
       "        'tRNA'],\n",
       "       'values': ['mRNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina Genome Analyzer']},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 16,\n",
       "   'name': 'DNA Methylation profiling by sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'DNA methylation profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['GENOMIC',\n",
       "        'GENOMIC SINGLE CELL',\n",
       "        'METAGENOMIC',\n",
       "        'OTHER'],\n",
       "       'values': ['GENOMIC']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina Genome Analyzer']},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']},\n",
       "      'library selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'HMPR',\n",
       "        'MF',\n",
       "        'MSLL',\n",
       "        '5-methylcytidine antibody',\n",
       "        'MBD2 protein methyl-CpG binding domain',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': ['5-methylcytidine antibody']},\n",
       "      'library strategy': {'options': ['Bisulfite-Seq',\n",
       "        'MRE-Seq',\n",
       "        'MeDIP-Seq',\n",
       "        'MBD-Seq',\n",
       "        'OTHER'],\n",
       "       'values': ['MeDIP-Seq']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
=======
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
       "     [],\n",
       "     ['blood'],\n",
       "     ['derma']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['derma']]}},\n",
       "  {'id': 16,\n",
       "   'name': 'Chip-Seq',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'protein-DNA binding site identification',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'cross linking': {'options': ['uv-light',\n",
       "        'formaldehyde',\n",
       "        'di-tert-butyl peroxide',\n",
       "        'formaldehyde and di-tert-butyl peroxyde'],\n",
       "       'values': ['uv-light', 'formaldehyde']},\n",
       "      'DNA fragmentation': {'options': ['sonication',\n",
       "        'nebulization',\n",
       "        'micrococcal nuclease digestion',\n",
       "        'deoxyribonuclease digestion'],\n",
       "       'values': ['nebulization']},\n",
       "      'DNA fragment size': {'newValues': True,\n",
       "       'isQuantitative': True,\n",
       "       'options': ['bbd', '100 nm'],\n",
       "       'values': ['100 nm']},\n",
       "      'immunoprecipitation antibody': {'newValues': True,\n",
       "       'isQuantitative': True,\n",
       "       'options': ['monoclonal'],\n",
       "       'values': ['monoclonal']}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['GENOMIC',\n",
       "        'GENOMIC SINGLE CELL',\n",
       "        'OTHER'],\n",
       "       'values': ['GENOMIC']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library preparation',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']},\n",
       "      'library selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'ChIP',\n",
       "        'MSLL',\n",
       "        'Restriction Digest',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': ['PCR', 'ChIP']},\n",
       "      'library strategy': {'options': ['Chip-Seq',\n",
       "        'Hi-C',\n",
       "        'ATAC-seq',\n",
       "        'ChIA-PET',\n",
       "        'Tethered Chromatin Conformation Capture',\n",
       "        'FAIRE-seq',\n",
       "        'OTHER'],\n",
       "       'values': ['Chip-Seq']}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina HiSeq 4000', 'Ion Torrent S5 XL']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
       "     [],\n",
<<<<<<< HEAD
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}}],\n",
       " 'selectedAssayTypes': {'Metabolite Profiling - Mass. Spec.': False,\n",
       "  'Metabolite Profiling - Mass. Spec. (Annotated)': True,\n",
       "  'Metabolite Profiling - NMR': True,\n",
       "  'RNA-Seq': True,\n",
       "  'DNA Methylation profiling by sequencing': True}}"
      ]
     },
     "execution_count": 208,
=======
       "     ['blood'],\n",
       "     ['derma']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['derma']]}}]}"
      ]
     },
     "execution_count": 5,
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
<<<<<<< HEAD
    "    \"ds-study-design-config\", \"study-design-3-repeated-treatments-datascriptor.json\"\n",
=======
    "    \"config\", \"crossover-study-design-4-arms-blood-derma-nmr-ms-chipseq.json\"\n",
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 209,
=======
   "execution_count": 6,
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "The link to be added is not valid. Link that can be created are ProductNode->ProtocolNode or ProtocolNode->ProductNode.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-209-8344ddcaf4e1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mstudy_design\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenerate_study_design_from_config\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstudy_design_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstudy_design\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mStudyDesign\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/connectors.py\u001b[0m in \u001b[0;36mgenerate_study_design_from_config\u001b[0;34m(study_design_config)\u001b[0m\n\u001b[1;32m    295\u001b[0m             \u001b[0msa_plan_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'SAP_A{}E{}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marm_ix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepoch_ix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    296\u001b[0m             \u001b[0;31m# TODO this method will probably need some rework to bind a sample type to a specific assay plan\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 297\u001b[0;31m             sa_plan = SampleAndAssayPlan.from_sample_and_assay_plan_dict(\n\u001b[0m\u001b[1;32m    298\u001b[0m                 \u001b[0msa_plan_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_type_dicts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0massay_ord_dicts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m             )\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36mfrom_sample_and_assay_plan_dict\u001b[0;34m(cls, name, sample_type_dicts, validation_template, use_guids, quality_controls, *assay_plan_dicts)\u001b[0m\n\u001b[1;32m   1436\u001b[0m         \u001b[0massay_map\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1437\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massay_plan_dict\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massay_plan_dicts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1438\u001b[0;31m             assay_graph = AssayGraph.generate_assay_plan_from_dict(\n\u001b[0m\u001b[1;32m   1439\u001b[0m                 \u001b[0massay_plan_dict\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1440\u001b[0m                 \u001b[0;31m# FIXME: this id cannot work as it is\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36mgenerate_assay_plan_from_dict\u001b[0;34m(cls, assay_plan_dict, validation_template, quality_control, use_guids, **kwargs)\u001b[0m\n\u001b[1;32m   1094\u001b[0m                             )\n\u001b[1;32m   1095\u001b[0m                             \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_node\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1096\u001b[0;31m                             \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_link\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprev_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1097\u001b[0m                             \u001b[0mcurrent_nodes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0mprevious_nodes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcurrent_nodes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36madd_link\u001b[0;34m(self, start_node, target_node)\u001b[0m\n\u001b[1;32m   1164\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProductNode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProtocolNode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1165\u001b[0m                 \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProtocolNode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProductNode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1166\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mINVALID_LINK_ERROR\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1167\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstart_node\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__graph_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mtarget_node\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__graph_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1168\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMISSING_NODE_ERROR\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: The link to be added is not valid. Link that can be created are ProductNode->ProtocolNode or ProtocolNode->ProductNode."
     ]
    }
   ],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 6.92 s.\n"
     ]
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 1.47 s.\n"
     ]
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
=======
   "execution_count": 9,
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'crossover-bio+drug-treatment'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory,'isa-investigationn-crossover-bio+drug.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab\n",
    "\n",
    "Expect this to take up to a few minutes, depending on your machine (on mine, with an 8-th gen i7 and 16 GB of RAM it took 90s)."
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 91.08 s.\n"
     ]
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
<<<<<<< HEAD
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%debug"
   ]
  },
  {
=======
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 3 assay files, one per assay type in our assay plan:"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT16_protein-DNA-binding-site-identification_nucleic-acid-sequencing.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT2_metabolite-profiling_NMR-spectroscopy.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT0_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 10 subjects in the each of the four selected arms (Arm_0, Arm_2, Arm_5, Arm_7) and 11 samples have been collected per subject (9 blood samples per subject, 6 of which during the follow-up epoch + 3 derma samples collected during the follow-up phase)"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
=======
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 110 samples in the GRP0 arm (i.e. group)\n",
      "There are 110 samples in the GRP2 arm (i.e. group)\n",
      "There are 110 samples in the GRP3 arm (i.e. group)\n",
      "There are 110 samples in the GRP4 arm (i.e. group)\n"
     ]
    }
   ],
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP5' in el)])\n",
    "count_arm4_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP7' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))\n",
    "print(\"There are {} samples in the GRP4 arm (i.e. group)\".format(count_arm4_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Chip-seq assay table\n",
    "\n",
    "The two chained protocols have been serialised successfully."
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT14_transcriptomics-profiling_nucleic-acid-sequencing.txt']"
=======
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Parameter Value[cross linking]</th>\n",
       "      <th>Parameter Value[DNA fragmentation]</th>\n",
       "      <th>Parameter Value[DNA fragment size]</th>\n",
       "      <th>Parameter Value[immunoprecipitation antibody]</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[library orientation]</th>\n",
       "      <th>Parameter Value[library selection]</th>\n",
       "      <th>Parameter Value[library strategy]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[sequencing instrument]</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S81-nucleic-acid-sequencing-Acquisition-R5</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-raw_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S81-nucleic-acid-sequencing-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S81-nucleic-acid-sequencing-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S81-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S1-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S1-nucleic-acid-sequencing-Acquisition-R6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S1-raw_data_file-R6.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S1-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S1-nucleic-acid-sequencing-Acquisition-R5</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S1-raw_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S68-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S68-nucleic-acid-sequencing-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S68-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S148-nucleic-acid-sequencing-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-raw_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S148-nucleic-acid-sequencing-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S68-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S68-nucleic-acid-sequencing-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S68-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S148-nucleic-acid-sequencing-Acquisition-R6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S148-raw_data_file-R6.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1280 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Protocol REF Parameter Value[cross linking]  \\\n",
       "0     GRP0_SBJ01_A0E5_SMP-derma-1   extraction                       uv-light   \n",
       "1     GRP0_SBJ01_A0E5_SMP-derma-1   extraction                       uv-light   \n",
       "2     GRP0_SBJ01_A0E5_SMP-derma-1   extraction                       uv-light   \n",
       "3     GRP0_SBJ01_A0E5_SMP-derma-1   extraction                   formaldehyde   \n",
       "4     GRP0_SBJ01_A0E5_SMP-derma-1   extraction                   formaldehyde   \n",
       "...                           ...          ...                            ...   \n",
       "1275  GRP7_SBJ10_A3E5_SMP-derma-2   extraction                   formaldehyde   \n",
       "1276  GRP7_SBJ10_A3E5_SMP-derma-2   extraction                       uv-light   \n",
       "1277  GRP7_SBJ10_A3E5_SMP-derma-2   extraction                       uv-light   \n",
       "1278  GRP7_SBJ10_A3E5_SMP-derma-2   extraction                   formaldehyde   \n",
       "1279  GRP7_SBJ10_A3E5_SMP-derma-2   extraction                       uv-light   \n",
       "\n",
       "     Parameter Value[DNA fragmentation] Parameter Value[DNA fragment size]  \\\n",
       "0                          nebulization                             100 nm   \n",
       "1                          nebulization                             100 nm   \n",
       "2                          nebulization                             100 nm   \n",
       "3                          nebulization                             100 nm   \n",
       "4                          nebulization                             100 nm   \n",
       "...                                 ...                                ...   \n",
       "1275                       nebulization                             100 nm   \n",
       "1276                       nebulization                             100 nm   \n",
       "1277                       nebulization                             100 nm   \n",
       "1278                       nebulization                             100 nm   \n",
       "1279                       nebulization                             100 nm   \n",
       "\n",
       "     Parameter Value[immunoprecipitation antibody] Performer  \\\n",
       "0                                       monoclonal   Unknown   \n",
       "1                                       monoclonal   Unknown   \n",
       "2                                       monoclonal   Unknown   \n",
       "3                                       monoclonal   Unknown   \n",
       "4                                       monoclonal   Unknown   \n",
       "...                                            ...       ...   \n",
       "1275                                    monoclonal   Unknown   \n",
       "1276                                    monoclonal   Unknown   \n",
       "1277                                    monoclonal   Unknown   \n",
       "1278                                    monoclonal   Unknown   \n",
       "1279                                    monoclonal   Unknown   \n",
       "\n",
       "              Extract Name Characteristics[extract type]       Protocol REF.1  \\\n",
       "0      AT16-S81-Extract-R1                       GENOMIC  library preparation   \n",
       "1      AT16-S81-Extract-R1                       GENOMIC  library preparation   \n",
       "2      AT16-S81-Extract-R1                       GENOMIC  library preparation   \n",
       "3       AT16-S1-Extract-R1                       GENOMIC  library preparation   \n",
       "4       AT16-S1-Extract-R1                       GENOMIC  library preparation   \n",
       "...                    ...                           ...                  ...   \n",
       "1275   AT16-S68-Extract-R1                       GENOMIC  library preparation   \n",
       "1276  AT16-S148-Extract-R1                       GENOMIC  library preparation   \n",
       "1277  AT16-S148-Extract-R1                       GENOMIC  library preparation   \n",
       "1278   AT16-S68-Extract-R1                       GENOMIC  library preparation   \n",
       "1279  AT16-S148-Extract-R1                       GENOMIC  library preparation   \n",
       "\n",
       "     Parameter Value[library orientation] Parameter Value[library selection]  \\\n",
       "0                              single-end                                PCR   \n",
       "1                              single-end                               ChIP   \n",
       "2                              single-end                               ChIP   \n",
       "3                              single-end                               ChIP   \n",
       "4                              single-end                               ChIP   \n",
       "...                                   ...                                ...   \n",
       "1275                           single-end                                PCR   \n",
       "1276                           single-end                                PCR   \n",
       "1277                           single-end                               ChIP   \n",
       "1278                           single-end                               ChIP   \n",
       "1279                           single-end                                PCR   \n",
       "\n",
       "     Parameter Value[library strategy] Performer.1           Protocol REF.2  \\\n",
       "0                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "2                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "3                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "4                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "...                                ...         ...                      ...   \n",
       "1275                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1276                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1277                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1278                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1279                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "\n",
       "     Parameter Value[sequencing instrument]  \\\n",
       "0                         Ion Torrent S5 XL   \n",
       "1                       Illumina HiSeq 4000   \n",
       "2                         Ion Torrent S5 XL   \n",
       "3                       Illumina HiSeq 4000   \n",
       "4                         Ion Torrent S5 XL   \n",
       "...                                     ...   \n",
       "1275                    Illumina HiSeq 4000   \n",
       "1276                      Ion Torrent S5 XL   \n",
       "1277                    Illumina HiSeq 4000   \n",
       "1278                    Illumina HiSeq 4000   \n",
       "1279                    Illumina HiSeq 4000   \n",
       "\n",
       "                                            Assay Name Performer.2  \\\n",
       "0      AT16-S81-nucleic-acid-sequencing-Acquisition-R5     Unknown   \n",
       "1      AT16-S81-nucleic-acid-sequencing-Acquisition-R1     Unknown   \n",
       "2      AT16-S81-nucleic-acid-sequencing-Acquisition-R4     Unknown   \n",
       "3       AT16-S1-nucleic-acid-sequencing-Acquisition-R6     Unknown   \n",
       "4       AT16-S1-nucleic-acid-sequencing-Acquisition-R5     Unknown   \n",
       "...                                                ...         ...   \n",
       "1275   AT16-S68-nucleic-acid-sequencing-Acquisition-R3     Unknown   \n",
       "1276  AT16-S148-nucleic-acid-sequencing-Acquisition-R7     Unknown   \n",
       "1277  AT16-S148-nucleic-acid-sequencing-Acquisition-R3     Unknown   \n",
       "1278   AT16-S68-nucleic-acid-sequencing-Acquisition-R8     Unknown   \n",
       "1279  AT16-S148-nucleic-acid-sequencing-Acquisition-R6     Unknown   \n",
       "\n",
       "                       Raw Data File  \n",
       "0      AT16-S81-raw_data_file-R5.raw  \n",
       "1      AT16-S81-raw_data_file-R1.raw  \n",
       "2      AT16-S81-raw_data_file-R4.raw  \n",
       "3       AT16-S1-raw_data_file-R6.raw  \n",
       "4       AT16-S1-raw_data_file-R5.raw  \n",
       "...                              ...  \n",
       "1275   AT16-S68-raw_data_file-R3.raw  \n",
       "1276  AT16-S148-raw_data_file-R7.raw  \n",
       "1277  AT16-S148-raw_data_file-R3.raw  \n",
       "1278   AT16-S68-raw_data_file-R8.raw  \n",
       "1279  AT16-S148-raw_data_file-R6.raw  \n",
       "\n",
       "[1280 rows x 19 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT16_protein-DNA-binding-site-identification_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.2 NMR assay table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Parameter Value[pulse_sequence]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Free Induction Decay Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-raw_spectral_data_file-R12.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-Extract-R1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-raw_spectral_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-Extract-R1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-raw_spectral_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-raw_spectral_data_file-R14.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S1-raw_spectral_data_file-R13.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-Extract-R1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-raw_spectral_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-Extract-R1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-raw_spectral_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-raw_spectral_data_file-R16.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-raw_spectral_data_file-R15.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-Extract-R2</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S68-raw_spectral_data_file-R12.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1280 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Protocol REF Performer        Extract Name  \\\n",
       "0     GRP0_SBJ01_A0E5_SMP-derma-1   extraction   Unknown   AT2-S1-Extract-R2   \n",
       "1     GRP0_SBJ01_A0E5_SMP-derma-1   extraction   Unknown   AT2-S1-Extract-R1   \n",
       "2     GRP0_SBJ01_A0E5_SMP-derma-1   extraction   Unknown   AT2-S1-Extract-R1   \n",
       "3     GRP0_SBJ01_A0E5_SMP-derma-1   extraction   Unknown   AT2-S1-Extract-R2   \n",
       "4     GRP0_SBJ01_A0E5_SMP-derma-1   extraction   Unknown   AT2-S1-Extract-R2   \n",
       "...                           ...          ...       ...                 ...   \n",
       "1275  GRP7_SBJ10_A3E5_SMP-derma-2   extraction   Unknown  AT2-S68-Extract-R1   \n",
       "1276  GRP7_SBJ10_A3E5_SMP-derma-2   extraction   Unknown  AT2-S68-Extract-R1   \n",
       "1277  GRP7_SBJ10_A3E5_SMP-derma-2   extraction   Unknown  AT2-S68-Extract-R2   \n",
       "1278  GRP7_SBJ10_A3E5_SMP-derma-2   extraction   Unknown  AT2-S68-Extract-R2   \n",
       "1279  GRP7_SBJ10_A3E5_SMP-derma-2   extraction   Unknown  AT2-S68-Extract-R2   \n",
       "\n",
       "     Characteristics[extract type]    Protocol REF.1  \\\n",
       "0                           pellet  nmr_spectroscopy   \n",
       "1                      supernatant  nmr_spectroscopy   \n",
       "2                      supernatant  nmr_spectroscopy   \n",
       "3                           pellet  nmr_spectroscopy   \n",
       "4                           pellet  nmr_spectroscopy   \n",
       "...                            ...               ...   \n",
       "1275                   supernatant  nmr_spectroscopy   \n",
       "1276                   supernatant  nmr_spectroscopy   \n",
       "1277                        pellet  nmr_spectroscopy   \n",
       "1278                        pellet  nmr_spectroscopy   \n",
       "1279                        pellet  nmr_spectroscopy   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[acquisition_mode]  \\\n",
       "0         Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "1         Bruker Avance II 1 GHz                    2D 13C-13C NMR   \n",
       "2         Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "3         Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "4         Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "...                          ...                               ...   \n",
       "1275      Bruker Avance II 1 GHz                    2D 13C-13C NMR   \n",
       "1276      Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "1277      Bruker Avance II 1 GHz                    2D 13C-13C NMR   \n",
       "1278      Bruker Avance II 1 GHz                    2D 13C-13C NMR   \n",
       "1279      Bruker Avance II 1 GHz                        1D 13C NMR   \n",
       "\n",
       "     Parameter Value[pulse_sequence] Performer.1  \\\n",
       "0                              TOCSY     Unknown   \n",
       "1                              TOCSY     Unknown   \n",
       "2                          watergate     Unknown   \n",
       "3                          watergate     Unknown   \n",
       "4                          watergate     Unknown   \n",
       "...                              ...         ...   \n",
       "1275                           TOCSY     Unknown   \n",
       "1276                           TOCSY     Unknown   \n",
       "1277                           TOCSY     Unknown   \n",
       "1278                           TOCSY     Unknown   \n",
       "1279                           TOCSY     Unknown   \n",
       "\n",
       "              Free Induction Decay Data File  \n",
       "0      AT2-S1-raw_spectral_data_file-R12.raw  \n",
       "1       AT2-S1-raw_spectral_data_file-R7.raw  \n",
       "2       AT2-S1-raw_spectral_data_file-R5.raw  \n",
       "3      AT2-S1-raw_spectral_data_file-R14.raw  \n",
       "4      AT2-S1-raw_spectral_data_file-R13.raw  \n",
       "...                                      ...  \n",
       "1275   AT2-S68-raw_spectral_data_file-R7.raw  \n",
       "1276   AT2-S68-raw_spectral_data_file-R3.raw  \n",
       "1277  AT2-S68-raw_spectral_data_file-R16.raw  \n",
       "1278  AT2-S68-raw_spectral_data_file-R15.raw  \n",
       "1279  AT2-S68-raw_spectral_data_file-R12.raw  \n",
       "\n",
       "[1280 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT2_metabolite-profiling_NMR-spectroscopy.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.3 Mass Spectrometry assay table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Labeled Extract Name</th>\n",
       "      <th>Label</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[injection_mode]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>MS Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Spectral Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S2-mass-spectrometry-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-raw-spectral-data-file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S1-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S1-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S1-mass-spectrometry-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S1-raw-spectral-data-file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S2-mass-spectrometry-Acquisition-R2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-raw-spectral-data-file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S2-mass-spectrometry-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-raw-spectral-data-file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S2-mass-spectrometry-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S2-raw-spectral-data-file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5115</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S568-mass-spectrometry-Acquisition-R2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-raw-spectral-data-file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5116</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S568-mass-spectrometry-Acquisition-R5</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S568-raw-spectral-data-file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5117</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S567-mass-spectrometry-Acquisition-R6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-raw-spectral-data-file-R6.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S567-mass-spectrometry-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-raw-spectral-data-file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5119</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S567-mass-spectrometry-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S567-raw-spectral-data-file-R8.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5120 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Protocol REF Performer         Extract Name  \\\n",
       "0     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown    AT0-S2-Extract-R1   \n",
       "1     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown    AT0-S1-Extract-R2   \n",
       "2     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown    AT0-S2-Extract-R1   \n",
       "3     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown    AT0-S2-Extract-R1   \n",
       "4     GRP0_SBJ01_A0E1_SMP-blood-1   extraction   Unknown    AT0-S2-Extract-R1   \n",
       "...                           ...          ...       ...                  ...   \n",
       "5115  GRP7_SBJ10_A3E5_SMP-blood-6   extraction   Unknown  AT0-S568-Extract-R1   \n",
       "5116  GRP7_SBJ10_A3E5_SMP-blood-6   extraction   Unknown  AT0-S568-Extract-R2   \n",
       "5117  GRP7_SBJ10_A3E5_SMP-blood-6   extraction   Unknown  AT0-S567-Extract-R2   \n",
       "5118  GRP7_SBJ10_A3E5_SMP-blood-6   extraction   Unknown  AT0-S567-Extract-R1   \n",
       "5119  GRP7_SBJ10_A3E5_SMP-blood-6   extraction   Unknown  AT0-S567-Extract-R2   \n",
       "\n",
       "     Characteristics[extract type] Protocol REF.1 Performer.1  \\\n",
       "0                   polar fraction       labeling     Unknown   \n",
       "1               non-polar fraction       labeling     Unknown   \n",
       "2                   polar fraction       labeling     Unknown   \n",
       "3                   polar fraction       labeling     Unknown   \n",
       "4                   polar fraction       labeling     Unknown   \n",
       "...                            ...            ...         ...   \n",
       "5115                polar fraction       labeling     Unknown   \n",
       "5116            non-polar fraction       labeling     Unknown   \n",
       "5117            non-polar fraction       labeling     Unknown   \n",
       "5118                polar fraction       labeling     Unknown   \n",
       "5119            non-polar fraction       labeling     Unknown   \n",
       "\n",
       "     Labeled Extract Name    Label     Protocol REF.2  \\\n",
       "0            AT0-S2-LE-R1  label_0  mass spectrometry   \n",
       "1            AT0-S1-LE-R2  label_0  mass spectrometry   \n",
       "2            AT0-S2-LE-R1  label_0  mass spectrometry   \n",
       "3            AT0-S2-LE-R1  label_0  mass spectrometry   \n",
       "4            AT0-S2-LE-R1  label_0  mass spectrometry   \n",
       "...                   ...      ...                ...   \n",
       "5115       AT0-S568-LE-R1  label_0  mass spectrometry   \n",
       "5116       AT0-S568-LE-R2  label_0  mass spectrometry   \n",
       "5117       AT0-S567-LE-R2  label_0  mass spectrometry   \n",
       "5118       AT0-S567-LE-R1  label_0  mass spectrometry   \n",
       "5119       AT0-S567-LE-R2  label_0  mass spectrometry   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[injection_mode]  \\\n",
       "0              Agilent QTQF 6510                             FIA   \n",
       "1              Agilent QTQF 6510                              LC   \n",
       "2              Agilent QTQF 6510                              LC   \n",
       "3              Agilent QTQF 6510                              LC   \n",
       "4              Agilent QTQF 6510                             FIA   \n",
       "...                          ...                             ...   \n",
       "5115           Agilent QTQF 6510                              LC   \n",
       "5116           Agilent QTQF 6510                             FIA   \n",
       "5117           Agilent QTQF 6510                             FIA   \n",
       "5118           Agilent QTQF 6510                              LC   \n",
       "5119           Agilent QTQF 6510                              LC   \n",
       "\n",
       "     Parameter Value[acquisition_mode]  \\\n",
       "0                        positive mode   \n",
       "1                        positive mode   \n",
       "2                        positive mode   \n",
       "3                        positive mode   \n",
       "4                        positive mode   \n",
       "...                                ...   \n",
       "5115                     positive mode   \n",
       "5116                     positive mode   \n",
       "5117                     positive mode   \n",
       "5118                     positive mode   \n",
       "5119                     positive mode   \n",
       "\n",
       "                                  MS Assay Name Performer.2  \\\n",
       "0       AT0-S2-mass-spectrometry-Acquisition-R3     Unknown   \n",
       "1       AT0-S1-mass-spectrometry-Acquisition-R8     Unknown   \n",
       "2       AT0-S2-mass-spectrometry-Acquisition-R2     Unknown   \n",
       "3       AT0-S2-mass-spectrometry-Acquisition-R1     Unknown   \n",
       "4       AT0-S2-mass-spectrometry-Acquisition-R4     Unknown   \n",
       "...                                         ...         ...   \n",
       "5115  AT0-S568-mass-spectrometry-Acquisition-R2     Unknown   \n",
       "5116  AT0-S568-mass-spectrometry-Acquisition-R5     Unknown   \n",
       "5117  AT0-S567-mass-spectrometry-Acquisition-R6     Unknown   \n",
       "5118  AT0-S567-mass-spectrometry-Acquisition-R1     Unknown   \n",
       "5119  AT0-S567-mass-spectrometry-Acquisition-R8     Unknown   \n",
       "\n",
       "                      Raw Spectral Data File  \n",
       "0       AT0-S2-raw-spectral-data-file-R3.raw  \n",
       "1       AT0-S1-raw-spectral-data-file-R8.raw  \n",
       "2       AT0-S2-raw-spectral-data-file-R2.raw  \n",
       "3       AT0-S2-raw-spectral-data-file-R1.raw  \n",
       "4       AT0-S2-raw-spectral-data-file-R4.raw  \n",
       "...                                      ...  \n",
       "5115  AT0-S568-raw-spectral-data-file-R2.raw  \n",
       "5116  AT0-S568-raw-spectral-data-file-R5.raw  \n",
       "5117  AT0-S567-raw-spectral-data-file-R6.raw  \n",
       "5118  AT0-S567-raw-spectral-data-file-R1.raw  \n",
       "5119  AT0-S567-raw-spectral-data-file-R8.raw  \n",
       "\n",
       "[5120 rows x 16 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT0_metabolite-profiling_mass-spectrometry.txt']"
>>>>>>> 0775c3ef2bb426997059973e0034466b2d66a4e3
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isa-api-py38",
   "language": "python",
   "name": "isa-api-py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
