{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "[missing]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json\n",
    "\n",
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study\n",
    "\n",
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design_from_config\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'treatmentPlan': {'screen': {'name': 'screen',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'runIn': {'name': 'run-in', 'duration': None, 'durationUnit': ''},\n",
       "  'washout': {'name': 'washout', 'duration': '5', 'durationUnit': 'days'},\n",
       "  'followUp': {'name': 'follow-up', 'duration': 60, 'durationUnit': 'days'},\n",
       "  'treatments': [{'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'aspirin',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'generatedStudyDesign': {'name': 'test study',\n",
       "  'description': 'this is the verbose description',\n",
       "  'type': {'term': 'crossover design',\n",
       "   'id': 'OBI:0500003',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/OBI_0500003'},\n",
       "  'elements': [{'id': '83978887-be4c-4cac-afb6-0faad3d07c86',\n",
       "    'name': 'treatment_0',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': 'e283a7c5-955f-4d3b-b99e-0053a2af5a61',\n",
       "    'name': 'washout',\n",
       "    'duration': 5,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': '696a2da4-329a-4b39-a681-e402afb4e89f',\n",
       "    'name': 'treatment_1',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'aspirin',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '6bf4662d-852f-44ad-9b46-acff5291c47f',\n",
       "    'name': 'treatment_2',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 60,\n",
       "    'durationUnit': 'days'}],\n",
       "  'events': [],\n",
       "  'arms': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'subjectType': 'human',\n",
       "    'size': 5,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}]},\n",
       " 'selectedArms': [{'id': '#arm/0',\n",
       "   'name': 'Arm_0',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/2',\n",
       "   'name': 'Arm_2',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/4',\n",
       "   'name': 'Arm_4',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/3',\n",
       "   'name': 'Arm_3',\n",
       "   'subjectType': 'human',\n",
       "   'size': 5,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}],\n",
       " 'samplePlan': [{'sampleType': 'saliva',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, None, 1, None, 1, 1],\n",
       "    'Arm_2': [1, None, 1, None, 1, 1],\n",
       "    'Arm_4': [1, None, 1, None, 1, 1],\n",
       "    'Arm_3': [1, None, 1, None, 1, 1]}},\n",
       "  {'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_4': [False, False, False, False, False, True],\n",
       "    'Arm_3': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 1],\n",
       "    'Arm_2': [None, None, None, None, None, 1],\n",
       "    'Arm_4': [None, None, None, None, None, 1],\n",
       "    'Arm_3': [None, None, None, None, None, 2]}}],\n",
       " 'assayConfigs': [{'id': 0,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec.',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'nonpolar fraction'],\n",
       "       'values': ['polar fraction']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labeling', {'#replicates': {'value': 1}}],\n",
       "    ['labeled extract',\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': 'labelled extract type',\n",
       "      'characteristics_value': {'options': ['biotin', 'cy-5', 'cy-3'],\n",
       "       'values': ['biotin']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': ['Agilent QTQF 6510']},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': ['LC']},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': ['positive mode', 'negative mode']}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 1,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec. (Annotated)',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}],\n",
       "       'values': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labelled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'biotin',\n",
       "         'iri': 'biotin',\n",
       "         'source': None},\n",
       "        {'term': 'cy-5', 'iri': 'cy-5', 'source': None},\n",
       "        {'term': 'cy-3', 'iri': 'cy-3', 'source': None}],\n",
       "       'values': [{'term': 'cy-3', 'iri': 'cy-3', 'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}]},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': 'LC', 'source': None}],\n",
       "       'values': []},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': 'MS'}],\n",
       "       'values': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': 'MS'}]}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': 'MS'},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'Metabolite Profiling - NMR',\n",
       "   'icon': 'fa fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
       "       'values': ['pellet']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'nmr sample', 'iri': None, 'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'NMR tube type': {'description': '',\n",
       "       'options': [{'term': '4 mm standard MAS rotor',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000379',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000393',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '2.5 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000375',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000372',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 7 in (Wilmad-LabGlass)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000387',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm Shigemi D2O matched',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000390',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm short',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000391',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000373',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000377',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000378',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4 mm Zirconia MAS rotor',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000380',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 541-pp',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000386',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '10 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000394',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4.25 mm (Bruker BioSpin)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000381',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm SampleJet',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000389',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (Norell Inc.)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000384',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (Wilmad-LabGlass)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000385',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '4.25 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000382',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm 7 in 507-grade',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000388',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm (New Era)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000383',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '5 mm ST500-7 (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000392',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '3 mm standard',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000376',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1.7 mm (Norell)',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000374',\n",
       "         'source': 'Metabolights.owl'}],\n",
       "       'values': ['1.7 mm (Norell)']},\n",
       "      'NMR solvent': {'description': '',\n",
       "       'options': [{'term': 'CD3OD + phosphate sodium buffer D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000363',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.01 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000348',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 5 mM maleic acid',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000369',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000352',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000350',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate-buffered saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001163',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'H20 + D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001157',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001154',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 0.128 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001160',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.05 mM phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001145',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001143',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001155',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001150',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001164',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.154 M saline D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001156',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001149',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001153',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001165',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.01 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001147',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD + phosphate sodium buffer D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001159',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '10% D2O/0.04 mM sodium phosphate buffer/0.4 mM TMSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001162',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001151',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate sodium buffer',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001144',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.04 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001148',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 5 mM maleic acid ',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001158',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.9% NaCl in D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001142',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CDCL3',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001141',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.15 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001146',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.3 M saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001152',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.5 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_001161',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000359',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.9% NaCl in D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000366',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate-buffered saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000368',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '1 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000354',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.15 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000351',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'H20 + D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000362',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.2 M potassium phosphate buffered D2O + 0.128 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000364',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CDCL3',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000361',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000371',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000356',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.154 M saline D20',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000370',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.05 mM phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000347',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000358',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'CD3OD',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000355',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.3 M saline D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000367',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.5 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000353',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.04 M phosphate buffered D2O/H2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000357',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'phosphate sodium buffer',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000360',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': '0.07 M phosphate buffered D2O',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000349',\n",
       "         'source': 'Metabolights.owl'},\n",
       "        {'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000365',\n",
       "         'source': 'Metabolights.owl'}],\n",
       "       'values': [{'term': 'D2O + 0.35 mM TSP',\n",
       "         'iri': 'http://www.ebi.ac.uk/metabolights/ontology/MTBLS_000365',\n",
       "         'source': 'Metabolights.owl'}]},\n",
       "      'NMR Sample pH': {'description': '', 'options': [], 'values': ['7.0']},\n",
       "      'NMR Temperature': {'description': '',\n",
       "       'options': [],\n",
       "       'values': ['65 degree C']}}],\n",
       "    ['nmr_spectroscopy',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Bruker AvanceII 1 GHz'],\n",
       "       'values': ['Bruker AvanceII 1 GHz']},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': ['1D 13C NMR']},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': ['CPMG']}}],\n",
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 14,\n",
       "   'name': 'RNA-Seq',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'transcriptomics profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['mRNA',\n",
       "        'total RNA',\n",
       "        'ncRNA',\n",
       "        'tRNA'],\n",
       "       'values': ['mRNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina Genome Analyzer']},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 16,\n",
       "   'name': 'DNA Methylation profiling by sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'DNA methylation profiling',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['GENOMIC',\n",
       "        'GENOMIC SINGLE CELL',\n",
       "        'METAGENOMIC',\n",
       "        'OTHER'],\n",
       "       'values': ['GENOMIC']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina Genome Analyzer']},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']},\n",
       "      'library selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'HMPR',\n",
       "        'MF',\n",
       "        'MSLL',\n",
       "        '5-methylcytidine antibody',\n",
       "        'MBD2 protein methyl-CpG binding domain',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': ['5-methylcytidine antibody']},\n",
       "      'library strategy': {'options': ['Bisulfite-Seq',\n",
       "        'MRE-Seq',\n",
       "        'MeDIP-Seq',\n",
       "        'MBD-Seq',\n",
       "        'OTHER'],\n",
       "       'values': ['MeDIP-Seq']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}}],\n",
       " 'selectedAssayTypes': {'Metabolite Profiling - Mass. Spec.': False,\n",
       "  'Metabolite Profiling - Mass. Spec. (Annotated)': True,\n",
       "  'Metabolite Profiling - NMR': True,\n",
       "  'RNA-Seq': True,\n",
       "  'DNA Methylation profiling by sequencing': True}}"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"ds-study-design-config\", \"study-design-3-repeated-treatments-datascriptor.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design_from_config` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "The link to be added is not valid. Link that can be created are ProductNode->ProtocolNode or ProtocolNode->ProductNode.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-209-8344ddcaf4e1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mstudy_design\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenerate_study_design_from_config\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstudy_design_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstudy_design\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mStudyDesign\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/connectors.py\u001b[0m in \u001b[0;36mgenerate_study_design_from_config\u001b[0;34m(study_design_config)\u001b[0m\n\u001b[1;32m    295\u001b[0m             \u001b[0msa_plan_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'SAP_A{}E{}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marm_ix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepoch_ix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    296\u001b[0m             \u001b[0;31m# TODO this method will probably need some rework to bind a sample type to a specific assay plan\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 297\u001b[0;31m             sa_plan = SampleAndAssayPlan.from_sample_and_assay_plan_dict(\n\u001b[0m\u001b[1;32m    298\u001b[0m                 \u001b[0msa_plan_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_type_dicts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0massay_ord_dicts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m             )\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36mfrom_sample_and_assay_plan_dict\u001b[0;34m(cls, name, sample_type_dicts, validation_template, use_guids, quality_controls, *assay_plan_dicts)\u001b[0m\n\u001b[1;32m   1436\u001b[0m         \u001b[0massay_map\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1437\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massay_plan_dict\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massay_plan_dicts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1438\u001b[0;31m             assay_graph = AssayGraph.generate_assay_plan_from_dict(\n\u001b[0m\u001b[1;32m   1439\u001b[0m                 \u001b[0massay_plan_dict\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1440\u001b[0m                 \u001b[0;31m# FIXME: this id cannot work as it is\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36mgenerate_assay_plan_from_dict\u001b[0;34m(cls, assay_plan_dict, validation_template, quality_control, use_guids, **kwargs)\u001b[0m\n\u001b[1;32m   1094\u001b[0m                             )\n\u001b[1;32m   1095\u001b[0m                             \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_node\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1096\u001b[0;31m                             \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_link\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprev_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1097\u001b[0m                             \u001b[0mcurrent_nodes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0mprevious_nodes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcurrent_nodes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.0/envs/isa-api-py38/src/isatools/isatools/create/model.py\u001b[0m in \u001b[0;36madd_link\u001b[0;34m(self, start_node, target_node)\u001b[0m\n\u001b[1;32m   1164\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProductNode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProtocolNode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1165\u001b[0m                 \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProtocolNode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_node\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mProductNode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1166\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mINVALID_LINK_ERROR\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1167\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstart_node\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__graph_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mtarget_node\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__graph_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1168\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMISSING_NODE_ERROR\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: The link to be added is not valid. Link that can be created are ProductNode->ProtocolNode or ProtocolNode->ProductNode."
     ]
    }
   ],
   "source": [
    "study_design = generate_study_design_from_config(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output'))\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "with open(os.path.abspath(os.path.join('output','isa-investigation-2-arms-nmr-ms.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, os.path.abspath(os.path.join('output')))\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%debug"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 5 assay files:\n",
    "* 1 assay file for Mass. Spec. (treatmennt arm, third epoch: surgery)\n",
    "* 2 assay files for Mass. Spec. (both arms, fourth epoch: follow-up)\n",
    "* 2 assay files for NMR (both arms, fourth epoch: follow-up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 8 subjects in the control arm and 24 samples have been collected (3 blood samples per subject during the follow-up epoch)\n",
    "\n",
    "We have 10 subjects in the control arm and 40 samples have been collected (3 blood samples per subject during the follow-up epoch and 1 liver sample per subject during the surgery epoch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_control_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'control' in el)])\n",
    "count_treatment_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'treatment' in el)])\n",
    "print(\"There are {} samples in the control arm (i.e. group)\".format(count_control_samples))\n",
    "print(\"There are {} samples in the treatment arm (i.e. group)\".format(count_treatment_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each control samples is fractioned and 2 labelled extracts are produced (i.e. biological replicates).\n",
    "\n",
    "```\n",
    "[\n",
    "    \"labelling\",\n",
    "    {\n",
    "        \"#replicates\": 2\n",
    "    }\n",
    "]\n",
    "```\n",
    "\n",
    "There are 2 possible combinations of Mass Spec Assay specified in our configuration template and 2 techincal replicate are produced for each combinations:\n",
    "\n",
    "```\n",
    "[\n",
    "    \"mass spectrometry\",\n",
    "    {\n",
    "        \"#replicates\": 2,\n",
    "        \"instrument\": [\n",
    "            \"Agilent QTQF 6510\"\n",
    "        ],\n",
    "        \"injection_mode\": [\n",
    "            \"FIA\",\n",
    "            \"LC\"\n",
    "        ],\n",
    "        \"acquisition_mode\": [\n",
    "            \"positive mode\"\n",
    "        ]\n",
    "    }\n",
    "]\n",
    "```\n",
    "\n",
    "Two output raw spectral files are produced as the result of reach run\n",
    "\n",
    "```\n",
    "[\n",
    "    \"raw spectral data file\",\n",
    "    [\n",
    "        {\n",
    "            \"node_type\": \"data file\",\n",
    "            \"size\": 2,\n",
    "            \"is_input_to_next_protocols\": false\n",
    "        }\n",
    "    ]\n",
    "]\n",
    "```\n",
    "\n",
    "As a total we expect\n",
    "\n",
    "$$ N_{rows} = (N_{subjects} \\times N_{samples}) \\times (N_{biorepl} \\times N_{combinations} \\times N_{techrepl}) = (8 \\times 3) \\times (2 \\times 2 \\times 2 \\times 2) = 24 \\times 16 = 384 $$\n",
    "\n",
    "Which we can verify for the mass spectrometry assay file of the control group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT14_transcriptomics-profiling_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isa-api-py38",
   "language": "python",
   "name": "isa-api-py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
