{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "# Crossover Study with radiological and drug treatment on Mice\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm study design. Each arm has 10 subjects\n",
    "- subjects are mice. There is an observational factor, named \"sex\" with two values: \"M\" and \"F\"\n",
    "- a crossover of two treatments, a drug treatment (cyclophosphamide 500 mg for 20 days) and a radiological treatment (\"strontium chloride sr 89\" 2 MBq/Kg forr 40 days)\n",
    "- four non-treatment phases: screen, run-in, washout and follow-up\n",
    "- three sample types colllected: tissue, blood, and urine\n",
    "- three assay types: \n",
    "    - metabolite profiling through mass spectrometry on the urine sample.\n",
    "    - clinical chemistry with marker on blood samples\n",
    "    - genomic sequencing on tissue samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator with all the specs defined above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'Test study',\n",
       " 'description': 'This design describes a study with repeated treatments on the same arms. We will be collecting three samples and these will undergo three different assays.',\n",
       " 'subjectType': 'Mouse',\n",
       " 'subjectSize': 10,\n",
       " 'designType': {'term': 'crossover design',\n",
       "  'id': 'OBI:0500003',\n",
       "  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',\n",
       "  'label': 'Study subjects receive repeated treatments',\n",
       "  'value': 'crossover'},\n",
       " 'observationalFactors': [{'name': 'sex',\n",
       "   'values': ['M', 'F'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None}],\n",
       " 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "    'type': 'Mouse',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_1',\n",
       "    'type': 'Mouse',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]}],\n",
       "  'unselected': []},\n",
       " 'treatmentPlan': {'screen': {'selected': True,\n",
       "   'name': 'screen',\n",
       "   'duration': 10,\n",
       "   'durationUnit': 'days'},\n",
       "  'runIn': {'selected': True,\n",
       "   'name': 'run-in',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  'washout': {'selected': True,\n",
       "   'name': 'washout',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  'followUp': {'selected': True,\n",
       "   'name': 'follow-up',\n",
       "   'duration': 3,\n",
       "   'durationUnit': 'months'},\n",
       "  'treatments': [{'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000599',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000599',\n",
       "     'short_form': 'ERO_0000599',\n",
       "     'obo_id': 'ERO:0000599',\n",
       "     'label': 'dietary intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'strontium chloride sr 89',\n",
       "    'intensity': 2,\n",
       "    'intensityUnit': 'MBq/Kg',\n",
       "    'duration': 40,\n",
       "    'durationUnit': 'days'},\n",
       "   {'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'short_form': 'ERO_0000600',\n",
       "     'obo_id': 'ERO:0000600',\n",
       "     'label': 'drug intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'cyclophosphamide',\n",
       "    'intensity': 500,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 20,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'interventionType': None,\n",
       "   'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'elements': [{'id': 'f2f3b834-0555-4d2c-8aa9-835df184d188',\n",
       "   'name': 'screen',\n",
       "   'duration': 10,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '482fe6ce-0ddb-4c9e-9da1-063805f534b7',\n",
       "   'name': 'run-in',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '59e2cfb5-0133-4f3f-af2c-000e2c991744',\n",
       "   'name': 'treatment_0',\n",
       "   'interventionType': {'term': 'dietary intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000599'},\n",
       "   'duration': 40,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'strontium chloride sr 89',\n",
       "   'intensity': 2,\n",
       "   'intensityUnit': 'MBq/Kg'},\n",
       "  {'id': '7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb',\n",
       "   'name': 'washout',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '2465e61b-7b64-4afb-8c14-edce966bffbe',\n",
       "   'name': 'treatment_1',\n",
       "   'interventionType': {'term': 'drug intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000600'},\n",
       "   'duration': 20,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'cyclophosphamide',\n",
       "   'intensity': 500,\n",
       "   'intensityUnit': 'mg'},\n",
       "  {'id': '4e6771bf-a4a5-46a9-a7c6-62d04ab3e126',\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'}],\n",
       " 'arms': {'selected': [{'id': 0,\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 1,\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 2,\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 3,\n",
       "    'name': 'Arm_3',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]}],\n",
       "  'unselected': []},\n",
       " 'samplePlan': [{'sampleType': 'Blood Sample',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_1': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, None, 2, None, 2, 3],\n",
       "    'Arm_1': [1, None, 2, None, 2, 3],\n",
       "    'Arm_2': [1, None, 2, None, 2, 3],\n",
       "    'Arm_3': [1, None, 2, None, 2, 3]}},\n",
       "  {'sampleType': 'Tissue Sample',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, True, False],\n",
       "    'Arm_1': [False, False, True, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, True, False],\n",
       "    'Arm_3': [False, False, True, False, False, False]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, 1, None],\n",
       "    'Arm_1': [None, None, 1, None, None, None],\n",
       "    'Arm_2': [None, None, None, None, 1, None],\n",
       "    'Arm_3': [None, None, 1, None, None, None]}},\n",
       "  {'sampleType': 'Urine',\n",
       "   'selectedCells': {'Arm_0': [True, True, True, True, True, True],\n",
       "    'Arm_1': [True, True, True, True, True, True],\n",
       "    'Arm_2': [True, True, True, True, True, True],\n",
       "    'Arm_3': [True, True, True, True, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_1': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_2': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_3': [2, 1, 1, None, 1, 2]}}],\n",
       " 'assayPlan': [{'id': 1,\n",
       "   'name': 'metabolite profiling - mass spectrometry (ontology annotated)',\n",
       "   'icon': 'fas fa-chart-bar',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}],\n",
       "       'values': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': ['label_0']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None}]},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}],\n",
       "       'values': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}]},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}]}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': None},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [True, True, True, False, True, True],\n",
       "    'Arm_1': [True, True, True, False, True, True],\n",
       "    'Arm_2': [True, True, True, False, True, True],\n",
       "    'Arm_3': [True, True, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Urine'],\n",
       "     ['Urine'],\n",
       "     ['Urine'],\n",
       "     [],\n",
       "     ['Urine'],\n",
       "     ['Urine']],\n",
       "    'Arm_1': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']],\n",
       "    'Arm_2': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']],\n",
       "    'Arm_3': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']]}},\n",
       "  {'id': 8,\n",
       "   'name': 'genome sequencing by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'genome sequencing',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 2}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'],\n",
       "       'values': ['gDNA', 'DNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': ['Illumina NovaSeq 6000']},\n",
       "      'library_orientation': {'options': ['single', 'paired'],\n",
       "       'values': ['single']},\n",
       "      'library_strategy': {'options': ['AMPLICON',\n",
       "        'CLONE',\n",
       "        'WGA',\n",
       "        'WGS',\n",
       "        'WXS',\n",
       "        'OTHER'],\n",
       "       'values': ['AMPLICON', 'WGS']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, True, False],\n",
       "    'Arm_1': [False, False, True, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, True, False],\n",
       "    'Arm_3': [False, False, True, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     [],\n",
       "     ['Tissue Sample'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_1': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Tissue Sample'],\n",
       "     [],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_2': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     [],\n",
       "     ['Tissue Sample'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_3': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Tissue Sample'],\n",
       "     [],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     ['Blood Sample', 'Urine']]}},\n",
       "  {'id': 11,\n",
       "   'name': 'hematology by marker panel',\n",
       "   'icon': 'fa fa-flask',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'clinical chemistry',\n",
       "   'technology_type': 'marker panel',\n",
       "   'workflow': [['sample preparation', {'#replicates': {'value': 3}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_1': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_1': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_2': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_3': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']]}}]}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"crossover-study-design-2-treatments-mice.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 1.50 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(identifier='inv01', studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 0.67 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'crossover-2-treatments-mice'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory, 'isa-investigation-crossover-2-treatments-mice.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 32.09 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT11_clinical-chemistry_marker-panel.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT1_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Count of subjects and samples\n",
    "\n",
    "We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the \"mass spcetrometry\" assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 160 samples in the GRP0 arm (i.e. group)\n",
      "There are 160 samples in the GRP1 arm (i.e. group)\n",
      "There are 160 samples in the GRP2 arm (i.e. group)\n",
      "There are 160 samples in the GRP3 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP1 arm (i.e. group)\".format(count_arm1_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Characteristics[Study Subject]</th>\n",
       "      <th>Characteristics[sex]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Parameter Value[Sampling order]</th>\n",
       "      <th>Parameter Value[Study cell]</th>\n",
       "      <th>Date</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Factor Value[Sequence Order]</th>\n",
       "      <th>Factor Value[AGENT]</th>\n",
       "      <th>Factor Value[DURATION]</th>\n",
       "      <th>Unit</th>\n",
       "      <th>Factor Value[INTENSITY]</th>\n",
       "      <th>Unit.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>033</td>\n",
       "      <td>A0E1</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-Urine-1</td>\n",
       "      <td>Urine</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>056</td>\n",
       "      <td>A0E2</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ01_A0E2_SMP-Blood-Sample-2</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>YES</td>\n",
       "      <td>2</td>\n",
       "      <td>strontium chloride sr 89</td>\n",
       "      <td>40</td>\n",
       "      <td>days</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MBq/Kg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>015</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-1</td>\n",
       "      <td>Urine</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>117</td>\n",
       "      <td>A0E5</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-Blood-Sample-1</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>90</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>118</td>\n",
       "      <td>A0E5</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-Blood-Sample-2</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>90</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>GRP3_SBJ10</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>582</td>\n",
       "      <td>A3E4</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP3_SBJ10_A3E4_SMP-Blood-Sample-2</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>YES</td>\n",
       "      <td>4</td>\n",
       "      <td>strontium chloride sr 89</td>\n",
       "      <td>40</td>\n",
       "      <td>days</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MBq/Kg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636</th>\n",
       "      <td>GRP3_SBJ10</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>516</td>\n",
       "      <td>A3E1</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP3_SBJ10_A3E1_SMP-Urine-1</td>\n",
       "      <td>Urine</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>637</th>\n",
       "      <td>GRP3_SBJ10</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>541</td>\n",
       "      <td>A3E2</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Blood-Sample-1</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>YES</td>\n",
       "      <td>2</td>\n",
       "      <td>cyclophosphamide</td>\n",
       "      <td>20</td>\n",
       "      <td>days</td>\n",
       "      <td>500.0</td>\n",
       "      <td>mg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>GRP3_SBJ10</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>502</td>\n",
       "      <td>A3E0</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP3_SBJ10_A3E0_SMP-Urine-2</td>\n",
       "      <td>Urine</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>days</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>GRP3_SBJ10</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>581</td>\n",
       "      <td>A3E4</td>\n",
       "      <td>2021-04-14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP3_SBJ10_A3E4_SMP-Blood-Sample-1</td>\n",
       "      <td>Blood Sample</td>\n",
       "      <td>YES</td>\n",
       "      <td>4</td>\n",
       "      <td>strontium chloride sr 89</td>\n",
       "      <td>40</td>\n",
       "      <td>days</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MBq/Kg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>640 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Source Name Characteristics[Study Subject] Characteristics[sex]  \\\n",
       "0    GRP0_SBJ01                          Mouse                    M   \n",
       "1    GRP0_SBJ01                          Mouse                    M   \n",
       "2    GRP0_SBJ01                          Mouse                    M   \n",
       "3    GRP0_SBJ01                          Mouse                    M   \n",
       "4    GRP0_SBJ01                          Mouse                    M   \n",
       "..          ...                            ...                  ...   \n",
       "635  GRP3_SBJ10                          Mouse                    F   \n",
       "636  GRP3_SBJ10                          Mouse                    F   \n",
       "637  GRP3_SBJ10                          Mouse                    F   \n",
       "638  GRP3_SBJ10                          Mouse                    F   \n",
       "639  GRP3_SBJ10                          Mouse                    F   \n",
       "\n",
       "          Protocol REF Parameter Value[Sampling order]  \\\n",
       "0    sample collection                             033   \n",
       "1    sample collection                             056   \n",
       "2    sample collection                             015   \n",
       "3    sample collection                             117   \n",
       "4    sample collection                             118   \n",
       "..                 ...                             ...   \n",
       "635  sample collection                             582   \n",
       "636  sample collection                             516   \n",
       "637  sample collection                             541   \n",
       "638  sample collection                             502   \n",
       "639  sample collection                             581   \n",
       "\n",
       "    Parameter Value[Study cell]        Date Performer  \\\n",
       "0                          A0E1  2021-04-14   Unknown   \n",
       "1                          A0E2  2021-04-14   Unknown   \n",
       "2                          A0E0  2021-04-14   Unknown   \n",
       "3                          A0E5  2021-04-14   Unknown   \n",
       "4                          A0E5  2021-04-14   Unknown   \n",
       "..                          ...         ...       ...   \n",
       "635                        A3E4  2021-04-14   Unknown   \n",
       "636                        A3E1  2021-04-14   Unknown   \n",
       "637                        A3E2  2021-04-14   Unknown   \n",
       "638                        A3E0  2021-04-14   Unknown   \n",
       "639                        A3E4  2021-04-14   Unknown   \n",
       "\n",
       "                            Sample Name Characteristics[organism part]  \\\n",
       "0           GRP0_SBJ01_A0E1_SMP-Urine-1                          Urine   \n",
       "1    GRP0_SBJ01_A0E2_SMP-Blood-Sample-2                   Blood Sample   \n",
       "2           GRP0_SBJ01_A0E0_SMP-Urine-1                          Urine   \n",
       "3    GRP0_SBJ01_A0E5_SMP-Blood-Sample-1                   Blood Sample   \n",
       "4    GRP0_SBJ01_A0E5_SMP-Blood-Sample-2                   Blood Sample   \n",
       "..                                  ...                            ...   \n",
       "635  GRP3_SBJ10_A3E4_SMP-Blood-Sample-2                   Blood Sample   \n",
       "636         GRP3_SBJ10_A3E1_SMP-Urine-1                          Urine   \n",
       "637  GRP3_SBJ10_A3E2_SMP-Blood-Sample-1                   Blood Sample   \n",
       "638         GRP3_SBJ10_A3E0_SMP-Urine-2                          Urine   \n",
       "639  GRP3_SBJ10_A3E4_SMP-Blood-Sample-1                   Blood Sample   \n",
       "\n",
       "    Comment[study step with treatment] Factor Value[Sequence Order]  \\\n",
       "0                                   NO                            1   \n",
       "1                                  YES                            2   \n",
       "2                                   NO                            0   \n",
       "3                                   NO                            5   \n",
       "4                                   NO                            5   \n",
       "..                                 ...                          ...   \n",
       "635                                YES                            4   \n",
       "636                                 NO                            1   \n",
       "637                                YES                            2   \n",
       "638                                 NO                            0   \n",
       "639                                YES                            4   \n",
       "\n",
       "          Factor Value[AGENT] Factor Value[DURATION]  Unit  \\\n",
       "0                                                  5  days   \n",
       "1    strontium chloride sr 89                     40  days   \n",
       "2                                                 10  days   \n",
       "3                                                 90  days   \n",
       "4                                                 90  days   \n",
       "..                        ...                    ...   ...   \n",
       "635  strontium chloride sr 89                     40  days   \n",
       "636                                                5  days   \n",
       "637          cyclophosphamide                     20  days   \n",
       "638                                               10  days   \n",
       "639  strontium chloride sr 89                     40  days   \n",
       "\n",
       "    Factor Value[INTENSITY]  Unit.1  \n",
       "0                                    \n",
       "1                       2.0  MBq/Kg  \n",
       "2                                    \n",
       "3                                    \n",
       "4                                    \n",
       "..                      ...     ...  \n",
       "635                     2.0  MBq/Kg  \n",
       "636                                  \n",
       "637                   500.0      mg  \n",
       "638                                  \n",
       "639                     2.0  MBq/Kg  \n",
       "\n",
       "[640 rows x 17 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Term Accession Number</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Labeled Extract Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Term Accession Number.1</th>\n",
       "      <th>Parameter Value[injection_mode]</th>\n",
       "      <th>Term Accession Number.2</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Term Accession Number.3</th>\n",
       "      <th>MS Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Spectral Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S5-mass-spectrometry-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-raw-spectral-data-file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S5-mass-spectrometry-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-raw-spectral-data-file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S5-mass-spectrometry-Acquisition-R2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-raw-spectral-data-file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S5-mass-spectrometry-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S5-raw-spectral-data-file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Urine-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S6-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S6-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S6-mass-spectrometry-Acquisition-R2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S6-raw-spectral-data-file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Urine-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S271-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S271-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S271-mass-spectrometry-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S271-raw-spectral-data-file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1116</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Urine-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S272-mass-spectrometry-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-raw-spectral-data-file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Urine-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S272-mass-spectrometry-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-raw-spectral-data-file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1118</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Urine-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>LC</td>\n",
       "      <td></td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S272-mass-spectrometry-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-raw-spectral-data-file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Urine-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-LE-R1</td>\n",
       "      <td>...</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000676</td>\n",
       "      <td>FIA</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1000058</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1002807</td>\n",
       "      <td>AT1-S272-mass-spectrometry-Acquisition-R2</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT1-S272-raw-spectral-data-file-R2.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1120 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   \n",
       "1     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   \n",
       "2     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   \n",
       "3     GRP0_SBJ01_A0E0_SMP-Urine-1                                 NO   \n",
       "4     GRP0_SBJ01_A0E0_SMP-Urine-2                                 NO   \n",
       "...                           ...                                ...   \n",
       "1115  GRP3_SBJ10_A3E5_SMP-Urine-1                                 NO   \n",
       "1116  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   \n",
       "1117  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   \n",
       "1118  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   \n",
       "1119  GRP3_SBJ10_A3E5_SMP-Urine-2                                 NO   \n",
       "\n",
       "     Protocol REF Performer         Extract Name  \\\n",
       "0      extraction   Unknown    AT1-S5-Extract-R1   \n",
       "1      extraction   Unknown    AT1-S5-Extract-R1   \n",
       "2      extraction   Unknown    AT1-S5-Extract-R1   \n",
       "3      extraction   Unknown    AT1-S5-Extract-R1   \n",
       "4      extraction   Unknown    AT1-S6-Extract-R1   \n",
       "...           ...       ...                  ...   \n",
       "1115   extraction   Unknown  AT1-S271-Extract-R1   \n",
       "1116   extraction   Unknown  AT1-S272-Extract-R1   \n",
       "1117   extraction   Unknown  AT1-S272-Extract-R1   \n",
       "1118   extraction   Unknown  AT1-S272-Extract-R1   \n",
       "1119   extraction   Unknown  AT1-S272-Extract-R1   \n",
       "\n",
       "     Characteristics[extract type] Term Accession Number Protocol REF.1  \\\n",
       "0                   polar fraction        polar fraction       labeling   \n",
       "1                   polar fraction        polar fraction       labeling   \n",
       "2                   polar fraction        polar fraction       labeling   \n",
       "3                   polar fraction        polar fraction       labeling   \n",
       "4                   polar fraction        polar fraction       labeling   \n",
       "...                            ...                   ...            ...   \n",
       "1115                polar fraction        polar fraction       labeling   \n",
       "1116                polar fraction        polar fraction       labeling   \n",
       "1117                polar fraction        polar fraction       labeling   \n",
       "1118                polar fraction        polar fraction       labeling   \n",
       "1119                polar fraction        polar fraction       labeling   \n",
       "\n",
       "     Performer.1 Labeled Extract Name  ...     Protocol REF.2  \\\n",
       "0        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   \n",
       "1        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   \n",
       "2        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   \n",
       "3        Unknown         AT1-S5-LE-R1  ...  mass spectrometry   \n",
       "4        Unknown         AT1-S6-LE-R1  ...  mass spectrometry   \n",
       "...          ...                  ...  ...                ...   \n",
       "1115     Unknown       AT1-S271-LE-R1  ...  mass spectrometry   \n",
       "1116     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   \n",
       "1117     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   \n",
       "1118     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   \n",
       "1119     Unknown       AT1-S272-LE-R1  ...  mass spectrometry   \n",
       "\n",
       "     Parameter Value[instrument]                    Term Accession Number.1  \\\n",
       "0              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "2              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "3              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "4              Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "...                          ...                                        ...   \n",
       "1115           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1116           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1117           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1118           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "1119           Agilent QTQF 6510  http://purl.obolibrary.org/obo/MS_1000676   \n",
       "\n",
       "     Parameter Value[injection_mode]  \\\n",
       "0                                 LC   \n",
       "1                                FIA   \n",
       "2                                FIA   \n",
       "3                                 LC   \n",
       "4                                FIA   \n",
       "...                              ...   \n",
       "1115                              LC   \n",
       "1116                              LC   \n",
       "1117                             FIA   \n",
       "1118                              LC   \n",
       "1119                             FIA   \n",
       "\n",
       "                        Term Accession Number.2  \\\n",
       "0                                                 \n",
       "1     http://purl.obolibrary.org/obo/MS_1000058   \n",
       "2     http://purl.obolibrary.org/obo/MS_1000058   \n",
       "3                                                 \n",
       "4     http://purl.obolibrary.org/obo/MS_1000058   \n",
       "...                                         ...   \n",
       "1115                                              \n",
       "1116                                              \n",
       "1117  http://purl.obolibrary.org/obo/MS_1000058   \n",
       "1118                                              \n",
       "1119  http://purl.obolibrary.org/obo/MS_1000058   \n",
       "\n",
       "     Parameter Value[acquisition_mode]  \\\n",
       "0                        positive mode   \n",
       "1                        positive mode   \n",
       "2                        positive mode   \n",
       "3                        positive mode   \n",
       "4                        positive mode   \n",
       "...                                ...   \n",
       "1115                     positive mode   \n",
       "1116                     positive mode   \n",
       "1117                     positive mode   \n",
       "1118                     positive mode   \n",
       "1119                     positive mode   \n",
       "\n",
       "                        Term Accession Number.3  \\\n",
       "0     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "2     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "3     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "4     http://purl.obolibrary.org/obo/MS_1002807   \n",
       "...                                         ...   \n",
       "1115  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1116  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1117  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1118  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "1119  http://purl.obolibrary.org/obo/MS_1002807   \n",
       "\n",
       "                                  MS Assay Name Performer.2  \\\n",
       "0       AT1-S5-mass-spectrometry-Acquisition-R3     Unknown   \n",
       "1       AT1-S5-mass-spectrometry-Acquisition-R1     Unknown   \n",
       "2       AT1-S5-mass-spectrometry-Acquisition-R2     Unknown   \n",
       "3       AT1-S5-mass-spectrometry-Acquisition-R4     Unknown   \n",
       "4       AT1-S6-mass-spectrometry-Acquisition-R2     Unknown   \n",
       "...                                         ...         ...   \n",
       "1115  AT1-S271-mass-spectrometry-Acquisition-R4     Unknown   \n",
       "1116  AT1-S272-mass-spectrometry-Acquisition-R3     Unknown   \n",
       "1117  AT1-S272-mass-spectrometry-Acquisition-R1     Unknown   \n",
       "1118  AT1-S272-mass-spectrometry-Acquisition-R4     Unknown   \n",
       "1119  AT1-S272-mass-spectrometry-Acquisition-R2     Unknown   \n",
       "\n",
       "                      Raw Spectral Data File  \n",
       "0       AT1-S5-raw-spectral-data-file-R3.raw  \n",
       "1       AT1-S5-raw-spectral-data-file-R1.raw  \n",
       "2       AT1-S5-raw-spectral-data-file-R2.raw  \n",
       "3       AT1-S5-raw-spectral-data-file-R4.raw  \n",
       "4       AT1-S6-raw-spectral-data-file-R2.raw  \n",
       "...                                      ...  \n",
       "1115  AT1-S271-raw-spectral-data-file-R4.raw  \n",
       "1116  AT1-S272-raw-spectral-data-file-R3.raw  \n",
       "1117  AT1-S272-raw-spectral-data-file-R1.raw  \n",
       "1118  AT1-S272-raw-spectral-data-file-R4.raw  \n",
       "1119  AT1-S272-raw-spectral-data-file-R2.raw  \n",
       "\n",
       "[1120 rows x 21 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Blood-Sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S9-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Blood-Sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S7-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E0_SMP-Blood-Sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S8-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E2_SMP-Blood-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S43-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E2_SMP-Blood-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S45-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Blood-Sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S921-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Blood-Sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S919-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Blood-Sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S923-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Blood-Sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S924-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>GRP3_SBJ10_A3E5_SMP-Blood-Sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S922-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>960 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Sample Name Comment[study step with treatment]  \\\n",
       "0    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   \n",
       "1    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   \n",
       "2    GRP0_SBJ01_A0E0_SMP-Blood-Sample-1                                 NO   \n",
       "3    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1                                YES   \n",
       "4    GRP0_SBJ01_A0E2_SMP-Blood-Sample-1                                YES   \n",
       "..                                  ...                                ...   \n",
       "955  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2                                 NO   \n",
       "956  GRP3_SBJ10_A3E5_SMP-Blood-Sample-2                                 NO   \n",
       "957  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   \n",
       "958  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   \n",
       "959  GRP3_SBJ10_A3E5_SMP-Blood-Sample-3                                 NO   \n",
       "\n",
       "           Protocol REF Performer               Raw Data File  \n",
       "0    sample preparation   Unknown    AT11-S9-raw_data_file-R1  \n",
       "1    sample preparation   Unknown    AT11-S7-raw_data_file-R1  \n",
       "2    sample preparation   Unknown    AT11-S8-raw_data_file-R1  \n",
       "3    sample preparation   Unknown   AT11-S43-raw_data_file-R1  \n",
       "4    sample preparation   Unknown   AT11-S45-raw_data_file-R1  \n",
       "..                  ...       ...                         ...  \n",
       "955  sample preparation   Unknown  AT11-S921-raw_data_file-R1  \n",
       "956  sample preparation   Unknown  AT11-S919-raw_data_file-R1  \n",
       "957  sample preparation   Unknown  AT11-S923-raw_data_file-R1  \n",
       "958  sample preparation   Unknown  AT11-S924-raw_data_file-R1  \n",
       "959  sample preparation   Unknown  AT11-S922-raw_data_file-R1  \n",
       "\n",
       "[960 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT11_clinical-chemistry_marker-panel.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[library_orientation]</th>\n",
       "      <th>Parameter Value[library_strategy]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-Extract-R1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S5-Extract-R2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S5-raw_data_file-R6.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-Extract-R1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-Extract-R2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-Extract-R1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S6-raw_data_file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S71-Extract-R2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S71-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636</th>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S71-Extract-R2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S71-raw_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>637</th>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-Extract-R2</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-Extract-R1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1</td>\n",
       "      <td>YES</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-Extract-R1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S72-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>640 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample Name Comment[study step with treatment]  \\\n",
       "0    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   \n",
       "1    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   \n",
       "2    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   \n",
       "3    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   \n",
       "4    GRP0_SBJ01_A0E4_SMP-Tissue-Sample-1                                YES   \n",
       "..                                   ...                                ...   \n",
       "635  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   \n",
       "636  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   \n",
       "637  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   \n",
       "638  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   \n",
       "639  GRP3_SBJ10_A3E2_SMP-Tissue-Sample-1                                YES   \n",
       "\n",
       "    Protocol REF Performer        Extract Name Characteristics[extract type]  \\\n",
       "0     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   \n",
       "1     extraction   Unknown   AT8-S5-Extract-R2                           DNA   \n",
       "2     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   \n",
       "3     extraction   Unknown   AT8-S6-Extract-R2                           DNA   \n",
       "4     extraction   Unknown   AT8-S6-Extract-R1                          gDNA   \n",
       "..           ...       ...                 ...                           ...   \n",
       "635   extraction   Unknown  AT8-S71-Extract-R2                           DNA   \n",
       "636   extraction   Unknown  AT8-S71-Extract-R2                           DNA   \n",
       "637   extraction   Unknown  AT8-S72-Extract-R2                           DNA   \n",
       "638   extraction   Unknown  AT8-S72-Extract-R1                          gDNA   \n",
       "639   extraction   Unknown  AT8-S72-Extract-R1                          gDNA   \n",
       "\n",
       "          Protocol REF.1 Parameter Value[instrument]  \\\n",
       "0    library_preparation       Illumina NovaSeq 6000   \n",
       "1    library_preparation       Illumina NovaSeq 6000   \n",
       "2    library_preparation       Illumina NovaSeq 6000   \n",
       "3    library_preparation       Illumina NovaSeq 6000   \n",
       "4    library_preparation       Illumina NovaSeq 6000   \n",
       "..                   ...                         ...   \n",
       "635  library_preparation       Illumina NovaSeq 6000   \n",
       "636  library_preparation       Illumina NovaSeq 6000   \n",
       "637  library_preparation       Illumina NovaSeq 6000   \n",
       "638  library_preparation       Illumina NovaSeq 6000   \n",
       "639  library_preparation       Illumina NovaSeq 6000   \n",
       "\n",
       "    Parameter Value[library_orientation] Parameter Value[library_strategy]  \\\n",
       "0                                 single                               WGS   \n",
       "1                                 single                               WGS   \n",
       "2                                 single                          AMPLICON   \n",
       "3                                 single                          AMPLICON   \n",
       "4                                 single                               WGS   \n",
       "..                                   ...                               ...   \n",
       "635                               single                          AMPLICON   \n",
       "636                               single                               WGS   \n",
       "637                               single                          AMPLICON   \n",
       "638                               single                               WGS   \n",
       "639                               single                          AMPLICON   \n",
       "\n",
       "    Performer.1                 Raw Data File  \n",
       "0       Unknown   AT8-S6-raw_data_file-R1.raw  \n",
       "1       Unknown   AT8-S5-raw_data_file-R6.raw  \n",
       "2       Unknown   AT8-S6-raw_data_file-R3.raw  \n",
       "3       Unknown   AT8-S6-raw_data_file-R8.raw  \n",
       "4       Unknown   AT8-S6-raw_data_file-R2.raw  \n",
       "..          ...                           ...  \n",
       "635     Unknown  AT8-S71-raw_data_file-R8.raw  \n",
       "636     Unknown  AT8-S71-raw_data_file-R5.raw  \n",
       "637     Unknown  AT8-S72-raw_data_file-R8.raw  \n",
       "638     Unknown  AT8-S72-raw_data_file-R1.raw  \n",
       "639     Unknown  AT8-S72-raw_data_file-R3.raw  \n",
       "\n",
       "[640 rows x 12 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7.1 Overview of the Mass Spec assay table\n",
    "\n",
    "For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, \"polar\" fraction), 320 labeled extracts (2 per extract, as \"#replicates\" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `[\"Agilent QTQF 6510\", \"FIA\", \"positive mode\"]` and `[\"Agilent QTQF 6510\", \"LC\", \"positive mode\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                            280\n",
       "Comment[study step with treatment]       2\n",
       "Protocol REF                             1\n",
       "Performer                                1\n",
       "Extract Name                           280\n",
       "Characteristics[extract type]            1\n",
       "Term Accession Number                    1\n",
       "Protocol REF.1                           1\n",
       "Performer.1                              1\n",
       "Labeled Extract Name                   280\n",
       "Label                                    1\n",
       "Protocol REF.2                           1\n",
       "Parameter Value[instrument]              1\n",
       "Term Accession Number.1                  1\n",
       "Parameter Value[injection_mode]          2\n",
       "Term Accession Number.2                  2\n",
       "Parameter Value[acquisition_mode]        1\n",
       "Term Accession Number.3                  1\n",
       "MS Assay Name                         1120\n",
       "Performer.2                              1\n",
       "Raw Spectral Data File                1120\n",
       "dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Genome Sequencing assay table\n",
    "\n",
    "For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the \"supernatant\" and \"pellet\" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"CPGM\"]` and `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"TOCSY\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                              40\n",
       "Comment[study step with treatment]        1\n",
       "Protocol REF                              1\n",
       "Performer                                 1\n",
       "Extract Name                            160\n",
       "Characteristics[extract type]             2\n",
       "Protocol REF.1                            1\n",
       "Parameter Value[instrument]               1\n",
       "Parameter Value[library_orientation]      1\n",
       "Parameter Value[library_strategy]         2\n",
       "Performer.1                               1\n",
       "Raw Data File                           640\n",
       "dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
