{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "# Observational study on Human students\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm study design. Each arm has 10 subjects\n",
    "- subjects are human, male and female, of three age groups in years (6-10, 10-15, 15-18) \n",
    "- two phases: the observational phase and follow-up\n",
    "- one sample type colllected: blood\n",
    "- two assay types: \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator with all the specs defined above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_id': '606454b9d071027feb44ad9b',\n",
       " 'name': \"Student's study\",\n",
       " 'description': 'In an observational study, the investigator simply records observations and analyzes data, without administration of an intervention or alteration of the care people receive. These studies may focus on risk factors, natural history, variations in disease progression or disease treatment without delivering an intervention. They often assess specific health characteristics of the enrolled human subjects by collecting medical/dental history, exposure, or clinical data; obtaining biospecimens (e.g., for biomarker or genomic analyses); or obtaining photographic, radiographic or other images from research subjects.',\n",
       " 'design': {'isPending': False,\n",
       "  'subjectType': 'Homo sapiens',\n",
       "  'subjectSize': 100,\n",
       "  'designType': {'term': 'observational design',\n",
       "   'id': 'NCIT:C147138',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/NCIT_C147138',\n",
       "   'label': 'Study subjects do not receive treatments',\n",
       "   'value': 'observational'},\n",
       "  'observationalFactors': [{'name': 'sex',\n",
       "    'values': ['M', 'F'],\n",
       "    'isQuantitative': False,\n",
       "    'unit': None},\n",
       "   {'name': 'age group',\n",
       "    'values': ['6-10', '11-15', '15-18'],\n",
       "    'isQuantitative': True,\n",
       "    'unit': 'years'}],\n",
       "  'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'M',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '15-18',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]},\n",
       "    {'name': 'SubjectGroup_1',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'M',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '11-15',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]},\n",
       "    {'name': 'SubjectGroup_2',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'M',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '6-10',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]},\n",
       "    {'name': 'SubjectGroup_3',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'F',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '15-18',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]},\n",
       "    {'name': 'SubjectGroup_4',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'F',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '11-15',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]},\n",
       "    {'name': 'SubjectGroup_5',\n",
       "     'type': 'Homo sapiens',\n",
       "     'characteristics': [{'name': 'sex',\n",
       "       'value': 'F',\n",
       "       'unit': None,\n",
       "       'isQuantitative': False},\n",
       "      {'name': 'age group',\n",
       "       'value': '6-10',\n",
       "       'unit': 'years',\n",
       "       'isQuantitative': True}]}],\n",
       "   'unselected': []},\n",
       "  'treatmentPlan': {'observationPeriod': {'name': 'observation period',\n",
       "    'duration': 180,\n",
       "    'durationUnit': 'days'},\n",
       "   'screen': {'selected': False,\n",
       "    'name': 'screen',\n",
       "    'duration': None,\n",
       "    'durationUnit': ''},\n",
       "   'runIn': {'selected': False,\n",
       "    'name': 'run-in',\n",
       "    'duration': None,\n",
       "    'durationUnit': ''},\n",
       "   'washout': {'selected': False,\n",
       "    'name': 'washout',\n",
       "    'duration': None,\n",
       "    'durationUnit': ''},\n",
       "   'followUp': {'selected': True,\n",
       "    'name': 'follow-up',\n",
       "    'duration': 300,\n",
       "    'durationUnit': 'days'},\n",
       "   'elementParams': {'agents': [],\n",
       "    'intensities': [],\n",
       "    'durations': [],\n",
       "    'interventionType': None,\n",
       "    'intensityUnit': '',\n",
       "    'durationUnit': ''},\n",
       "   'treatments': [{'_id': '606454b9d071027feb44ad9c',\n",
       "     'interventionType': None,\n",
       "     'agent': None,\n",
       "     'intensity': None,\n",
       "     'intensityUnit': '',\n",
       "     'duration': None,\n",
       "     'durationUnit': ''},\n",
       "    {'_id': '606454b9d071027feb44ad9d',\n",
       "     'interventionType': None,\n",
       "     'agent': None,\n",
       "     'intensity': None,\n",
       "     'intensityUnit': '',\n",
       "     'duration': None,\n",
       "     'durationUnit': ''}]},\n",
       "  'elements': [{'_id': '606454b9d071027feb44ad9e',\n",
       "    'id': '608e698d-208d-415f-81f7-32d7cbf4b18b',\n",
       "    'name': 'observation period',\n",
       "    'duration': 180,\n",
       "    'durationUnit': 'days'},\n",
       "   {'_id': '606454b9d071027feb44ad9f',\n",
       "    'id': 'c43550b1-77ee-41f1-b54e-f22092cc6ded',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 300,\n",
       "    'durationUnit': 'days'}],\n",
       "  'arms': {'unselected': [],\n",
       "   'selected': [{'_id': '606454b9d071027feb44ada0',\n",
       "     'id': '0',\n",
       "     'name': 'Arm_0',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada1'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada2'}],\n",
       "     'observationalFactors': []},\n",
       "    {'_id': '606454b9d071027feb44ada3',\n",
       "     'id': '1',\n",
       "     'name': 'Arm_1',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada4'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada5'}],\n",
       "     'observationalFactors': []},\n",
       "    {'_id': '606454b9d071027feb44ada6',\n",
       "     'id': '2',\n",
       "     'name': 'Arm_2',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada7'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44ada8'}],\n",
       "     'observationalFactors': []},\n",
       "    {'_id': '606454b9d071027feb44ada9',\n",
       "     'id': '3',\n",
       "     'name': 'Arm_3',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adaa'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adab'}],\n",
       "     'observationalFactors': []},\n",
       "    {'_id': '606454b9d071027feb44adac',\n",
       "     'id': '4',\n",
       "     'name': 'Arm_4',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adad'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adae'}],\n",
       "     'observationalFactors': []},\n",
       "    {'_id': '606454b9d071027feb44adaf',\n",
       "     'id': '5',\n",
       "     'name': 'Arm_5',\n",
       "     'size': 100,\n",
       "     'epochs': [{'elements': ['608e698d-208d-415f-81f7-32d7cbf4b18b'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adb0'},\n",
       "      {'elements': ['c43550b1-77ee-41f1-b54e-f22092cc6ded'],\n",
       "       'events': [],\n",
       "       '_id': '606454b9d071027feb44adb1'}],\n",
       "     'observationalFactors': []}]},\n",
       "  'samplePlan': [{'sampleType': 'blood sample',\n",
       "    'selectedCells': {'Arm_0': [True, True],\n",
       "     'Arm_1': [True, True],\n",
       "     'Arm_2': [True, True],\n",
       "     'Arm_3': [True, True],\n",
       "     'Arm_4': [True, True],\n",
       "     'Arm_5': [True, True]},\n",
       "    'sampleTypeSizes': {'Arm_0': [3, 1],\n",
       "     'Arm_1': [3, 1],\n",
       "     'Arm_2': [3, 1],\n",
       "     'Arm_3': [3, 1],\n",
       "     'Arm_4': [3, 1],\n",
       "     'Arm_5': [3, 1]}}],\n",
       "  'assayPlan': [{'id': 11,\n",
       "    'name': 'hematology by marker panel',\n",
       "    'icon': 'fa fa-flask',\n",
       "    'color': 'red',\n",
       "    'measurement_type': 'clinical chemistry',\n",
       "    'technology_type': 'marker panel',\n",
       "    'workflow': [['sample preparation', {'#replicates': {'value': 2}}],\n",
       "     ['raw_data_file',\n",
       "      {'node_type': 'data file',\n",
       "       'size': 1,\n",
       "       'is_input_to_next_protocols': {'value': False}}]],\n",
       "    'selectedCells': {'Arm_0': [True, True],\n",
       "     'Arm_1': [True, True],\n",
       "     'Arm_2': [True, True],\n",
       "     'Arm_3': [True, True],\n",
       "     'Arm_4': [True, True],\n",
       "     'Arm_5': [False, False]},\n",
       "    'selectedSampleTypes': {'Arm_0': [['blood sample'], ['blood sample']],\n",
       "     'Arm_1': [['blood sample'], ['blood sample']],\n",
       "     'Arm_2': [['blood sample'], ['blood sample']],\n",
       "     'Arm_3': [['blood sample'], ['blood sample']],\n",
       "     'Arm_4': [['blood sample'], ['blood sample']],\n",
       "     'Arm_5': [['blood sample'], ['blood sample']]}},\n",
       "   {'id': 8,\n",
       "    'name': 'genome sequencing by nucleotide sequencing',\n",
       "    'icon': 'fa fa-dna',\n",
       "    'color': 'purple',\n",
       "    'measurement_type': 'genome sequencing',\n",
       "    'technology_type': 'nucleic acid sequencing',\n",
       "    'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "     ['extract',\n",
       "      {'node_type': 'extract',\n",
       "       'characteristics_category': 'extract type',\n",
       "       'characteristics_value': {'options': ['gDNA', 'DNA'],\n",
       "        'values': ['DNA']},\n",
       "       'is_input_to_next_protocols': {'value': True}}],\n",
       "     ['library_preparation',\n",
       "      {'#replicates': {'value': 2},\n",
       "       'instrument': {'options': ['MinION',\n",
       "         'GridION',\n",
       "         'PromethION',\n",
       "         'Illumina HiSeq 4000',\n",
       "         'Illumina NovaSeq 6000',\n",
       "         'AB 3730xL Genetic Analyzer'],\n",
       "        'values': ['Illumina NovaSeq 6000']},\n",
       "       'library_orientation': {'options': ['single', 'paired'],\n",
       "        'values': ['single', 'paired']},\n",
       "       'library_strategy': {'options': ['AMPLICON',\n",
       "         'CLONE',\n",
       "         'WGA',\n",
       "         'WGS',\n",
       "         'WXS',\n",
       "         'OTHER'],\n",
       "        'values': ['WGS']}}],\n",
       "     ['raw_data_file',\n",
       "      {'node_type': 'data file',\n",
       "       'size': 1,\n",
       "       'is_input_to_next_protocols': {'value': False}}]],\n",
       "    'selectedCells': {'Arm_0': [True, True],\n",
       "     'Arm_1': [True, True],\n",
       "     'Arm_2': [True, True],\n",
       "     'Arm_3': [True, True],\n",
       "     'Arm_4': [True, True],\n",
       "     'Arm_5': [False, False]},\n",
       "    'selectedSampleTypes': {'Arm_0': [['blood sample'], ['blood sample']],\n",
       "     'Arm_1': [['blood sample'], ['blood sample']],\n",
       "     'Arm_2': [['blood sample'], ['blood sample']],\n",
       "     'Arm_3': [['blood sample'], ['blood sample']],\n",
       "     'Arm_4': [['blood sample'], ['blood sample']],\n",
       "     'Arm_5': [['blood sample'], ['blood sample']]}}],\n",
       "  'events': []}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"observational-study-students.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study from the design took 3.94 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study from the design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(identifier='inv01', studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 2.11 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'observational-design-students'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory, 'isa-investigation-observational-design-students.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 66.29 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT11_clinical-chemistry_marker-panel.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Count of subjects and samples\n",
    "\n",
    "We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the \"mass spcetrometry\" assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 400 samples in the GRP0 arm (i.e. group)\n",
      "There are 400 samples in the GRP1 arm (i.e. group)\n",
      "There are 400 samples in the GRP2 arm (i.e. group)\n",
      "There are 400 samples in the GRP3 arm (i.e. group)\n",
      "There are 400 samples in the GRP4 arm (i.e. group)\n",
      "There are 400 samples in the GRP5 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])\n",
    "count_arm4_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP4' in el)])\n",
    "count_arm5_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP5' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP1 arm (i.e. group)\".format(count_arm1_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))\n",
    "print(\"There are {} samples in the GRP4 arm (i.e. group)\".format(count_arm4_samples))\n",
    "print(\"There are {} samples in the GRP5 arm (i.e. group)\".format(count_arm5_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Characteristics[Study Subject]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Parameter Value[Sampling order]</th>\n",
       "      <th>Parameter Value[Study cell]</th>\n",
       "      <th>Date</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Factor Value[Sequence Order]</th>\n",
       "      <th>Factor Value[DURATION]</th>\n",
       "      <th>Unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>343</td>\n",
       "      <td>A0E1</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>127</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>129</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-3</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>128</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-2</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>280</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ002_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>GRP5_SBJ099</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>2321</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ099_A5E1_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>2007</td>\n",
       "      <td>A5E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>2009</td>\n",
       "      <td>A5E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-sample-3</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2398</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>2303</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-sample-1</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>2008</td>\n",
       "      <td>A5E0</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-sample-2</td>\n",
       "      <td>blood sample</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2400 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Source Name Characteristics[Study Subject]       Protocol REF  \\\n",
       "0     GRP0_SBJ001                   Homo sapiens  sample collection   \n",
       "1     GRP0_SBJ001                   Homo sapiens  sample collection   \n",
       "2     GRP0_SBJ001                   Homo sapiens  sample collection   \n",
       "3     GRP0_SBJ001                   Homo sapiens  sample collection   \n",
       "4     GRP0_SBJ002                   Homo sapiens  sample collection   \n",
       "...           ...                            ...                ...   \n",
       "2395  GRP5_SBJ099                   Homo sapiens  sample collection   \n",
       "2396  GRP5_SBJ100                   Homo sapiens  sample collection   \n",
       "2397  GRP5_SBJ100                   Homo sapiens  sample collection   \n",
       "2398  GRP5_SBJ100                   Homo sapiens  sample collection   \n",
       "2399  GRP5_SBJ100                   Homo sapiens  sample collection   \n",
       "\n",
       "     Parameter Value[Sampling order] Parameter Value[Study cell]        Date  \\\n",
       "0                                343                        A0E1  2021-06-29   \n",
       "1                                127                        A0E0  2021-06-29   \n",
       "2                                129                        A0E0  2021-06-29   \n",
       "3                                128                        A0E0  2021-06-29   \n",
       "4                                280                        A0E0  2021-06-29   \n",
       "...                              ...                         ...         ...   \n",
       "2395                            2321                        A5E1  2021-06-29   \n",
       "2396                            2007                        A5E0  2021-06-29   \n",
       "2397                            2009                        A5E0  2021-06-29   \n",
       "2398                            2303                        A5E1  2021-06-29   \n",
       "2399                            2008                        A5E0  2021-06-29   \n",
       "\n",
       "     Performer                          Sample Name  \\\n",
       "0      Unknown  GRP0_SBJ001_A0E1_SMP-blood-sample-1   \n",
       "1      Unknown  GRP0_SBJ001_A0E0_SMP-blood-sample-1   \n",
       "2      Unknown  GRP0_SBJ001_A0E0_SMP-blood-sample-3   \n",
       "3      Unknown  GRP0_SBJ001_A0E0_SMP-blood-sample-2   \n",
       "4      Unknown  GRP0_SBJ002_A0E0_SMP-blood-sample-1   \n",
       "...        ...                                  ...   \n",
       "2395   Unknown  GRP5_SBJ099_A5E1_SMP-blood-sample-1   \n",
       "2396   Unknown  GRP5_SBJ100_A5E0_SMP-blood-sample-1   \n",
       "2397   Unknown  GRP5_SBJ100_A5E0_SMP-blood-sample-3   \n",
       "2398   Unknown  GRP5_SBJ100_A5E1_SMP-blood-sample-1   \n",
       "2399   Unknown  GRP5_SBJ100_A5E0_SMP-blood-sample-2   \n",
       "\n",
       "     Characteristics[organism part] Comment[study step with treatment]  \\\n",
       "0                      blood sample                                 NO   \n",
       "1                      blood sample                                 NO   \n",
       "2                      blood sample                                 NO   \n",
       "3                      blood sample                                 NO   \n",
       "4                      blood sample                                 NO   \n",
       "...                             ...                                ...   \n",
       "2395                   blood sample                                 NO   \n",
       "2396                   blood sample                                 NO   \n",
       "2397                   blood sample                                 NO   \n",
       "2398                   blood sample                                 NO   \n",
       "2399                   blood sample                                 NO   \n",
       "\n",
       "     Factor Value[Sequence Order] Factor Value[DURATION]  Unit  \n",
       "0                               1                    300  days  \n",
       "1                               0                    180  days  \n",
       "2                               0                    180  days  \n",
       "3                               0                    180  days  \n",
       "4                               0                    180  days  \n",
       "...                           ...                    ...   ...  \n",
       "2395                            1                    300  days  \n",
       "2396                            0                    180  days  \n",
       "2397                            0                    180  days  \n",
       "2398                            1                    300  days  \n",
       "2399                            0                    180  days  \n",
       "\n",
       "[2400 rows x 13 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S254-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S253-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S256-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S255-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S257-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3995</th>\n",
       "      <td>GRP4_SBJ100_A4E0_SMP-blood-sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S3419-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3996</th>\n",
       "      <td>GRP4_SBJ100_A4E0_SMP-blood-sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S3422-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>GRP4_SBJ100_A4E0_SMP-blood-sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S3421-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S3874-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT11-S3873-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "1     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "2     GRP0_SBJ001_A0E0_SMP-blood-sample-2                                 NO   \n",
       "3     GRP0_SBJ001_A0E0_SMP-blood-sample-2                                 NO   \n",
       "4     GRP0_SBJ001_A0E0_SMP-blood-sample-3                                 NO   \n",
       "...                                   ...                                ...   \n",
       "3995  GRP4_SBJ100_A4E0_SMP-blood-sample-2                                 NO   \n",
       "3996  GRP4_SBJ100_A4E0_SMP-blood-sample-3                                 NO   \n",
       "3997  GRP4_SBJ100_A4E0_SMP-blood-sample-3                                 NO   \n",
       "3998  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "3999  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "\n",
       "            Protocol REF Performer                Raw Data File  \n",
       "0     sample preparation   Unknown   AT11-S254-raw_data_file-R1  \n",
       "1     sample preparation   Unknown   AT11-S253-raw_data_file-R1  \n",
       "2     sample preparation   Unknown   AT11-S256-raw_data_file-R1  \n",
       "3     sample preparation   Unknown   AT11-S255-raw_data_file-R1  \n",
       "4     sample preparation   Unknown   AT11-S257-raw_data_file-R1  \n",
       "...                  ...       ...                          ...  \n",
       "3995  sample preparation   Unknown  AT11-S3419-raw_data_file-R1  \n",
       "3996  sample preparation   Unknown  AT11-S3422-raw_data_file-R1  \n",
       "3997  sample preparation   Unknown  AT11-S3421-raw_data_file-R1  \n",
       "3998  sample preparation   Unknown  AT11-S3874-raw_data_file-R1  \n",
       "3999  sample preparation   Unknown  AT11-S3873-raw_data_file-R1  \n",
       "\n",
       "[4000 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT11_clinical-chemistry_marker-panel.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[library_orientation]</th>\n",
       "      <th>Parameter Value[library_strategy]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-raw_data_file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S127-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-sample-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S128-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S128-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7995</th>\n",
       "      <td>GRP4_SBJ100_A4E0_SMP-blood-sample-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1711-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1711-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7996</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7997</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-raw_data_file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7998</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7999</th>\n",
       "      <td>GRP4_SBJ100_A4E1_SMP-blood-sample-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1937-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8000 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "1     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "2     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "3     GRP0_SBJ001_A0E0_SMP-blood-sample-1                                 NO   \n",
       "4     GRP0_SBJ001_A0E0_SMP-blood-sample-2                                 NO   \n",
       "...                                   ...                                ...   \n",
       "7995  GRP4_SBJ100_A4E0_SMP-blood-sample-3                                 NO   \n",
       "7996  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "7997  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "7998  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "7999  GRP4_SBJ100_A4E1_SMP-blood-sample-1                                 NO   \n",
       "\n",
       "     Protocol REF Performer          Extract Name  \\\n",
       "0      extraction   Unknown   AT8-S127-Extract-R1   \n",
       "1      extraction   Unknown   AT8-S127-Extract-R1   \n",
       "2      extraction   Unknown   AT8-S127-Extract-R1   \n",
       "3      extraction   Unknown   AT8-S127-Extract-R1   \n",
       "4      extraction   Unknown   AT8-S128-Extract-R1   \n",
       "...           ...       ...                   ...   \n",
       "7995   extraction   Unknown  AT8-S1711-Extract-R1   \n",
       "7996   extraction   Unknown  AT8-S1937-Extract-R1   \n",
       "7997   extraction   Unknown  AT8-S1937-Extract-R1   \n",
       "7998   extraction   Unknown  AT8-S1937-Extract-R1   \n",
       "7999   extraction   Unknown  AT8-S1937-Extract-R1   \n",
       "\n",
       "     Characteristics[extract type]       Protocol REF.1  \\\n",
       "0                              DNA  library_preparation   \n",
       "1                              DNA  library_preparation   \n",
       "2                              DNA  library_preparation   \n",
       "3                              DNA  library_preparation   \n",
       "4                              DNA  library_preparation   \n",
       "...                            ...                  ...   \n",
       "7995                           DNA  library_preparation   \n",
       "7996                           DNA  library_preparation   \n",
       "7997                           DNA  library_preparation   \n",
       "7998                           DNA  library_preparation   \n",
       "7999                           DNA  library_preparation   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[library_orientation]  \\\n",
       "0          Illumina NovaSeq 6000                               single   \n",
       "1          Illumina NovaSeq 6000                               paired   \n",
       "2          Illumina NovaSeq 6000                               single   \n",
       "3          Illumina NovaSeq 6000                               paired   \n",
       "4          Illumina NovaSeq 6000                               single   \n",
       "...                          ...                                  ...   \n",
       "7995       Illumina NovaSeq 6000                               paired   \n",
       "7996       Illumina NovaSeq 6000                               single   \n",
       "7997       Illumina NovaSeq 6000                               single   \n",
       "7998       Illumina NovaSeq 6000                               paired   \n",
       "7999       Illumina NovaSeq 6000                               paired   \n",
       "\n",
       "     Parameter Value[library_strategy] Performer.1  \\\n",
       "0                                  WGS     Unknown   \n",
       "1                                  WGS     Unknown   \n",
       "2                                  WGS     Unknown   \n",
       "3                                  WGS     Unknown   \n",
       "4                                  WGS     Unknown   \n",
       "...                                ...         ...   \n",
       "7995                               WGS     Unknown   \n",
       "7996                               WGS     Unknown   \n",
       "7997                               WGS     Unknown   \n",
       "7998                               WGS     Unknown   \n",
       "7999                               WGS     Unknown   \n",
       "\n",
       "                       Raw Data File  \n",
       "0      AT8-S127-raw_data_file-R1.raw  \n",
       "1      AT8-S127-raw_data_file-R3.raw  \n",
       "2      AT8-S127-raw_data_file-R2.raw  \n",
       "3      AT8-S127-raw_data_file-R4.raw  \n",
       "4      AT8-S128-raw_data_file-R1.raw  \n",
       "...                              ...  \n",
       "7995  AT8-S1711-raw_data_file-R4.raw  \n",
       "7996  AT8-S1937-raw_data_file-R1.raw  \n",
       "7997  AT8-S1937-raw_data_file-R2.raw  \n",
       "7998  AT8-S1937-raw_data_file-R4.raw  \n",
       "7999  AT8-S1937-raw_data_file-R3.raw  \n",
       "\n",
       "[8000 rows x 12 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Clinical Chemistry Marker Panel\n",
    "\n",
    "For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, \"polar\" fraction), 320 labeled extracts (2 per extract, as \"#replicates\" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `[\"Agilent QTQF 6510\", \"FIA\", \"positive mode\"]` and `[\"Agilent QTQF 6510\", \"LC\", \"positive mode\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                           2000\n",
       "Comment[study step with treatment]       1\n",
       "Protocol REF                             1\n",
       "Performer                                1\n",
       "Raw Data File                         4000\n",
       "dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT11_clinical-chemistry_marker-panel.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Genome Sequencing assay table\n",
    "\n",
    "For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the \"supernatant\" and \"pellet\" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"CPGM\"]` and `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"TOCSY\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                             2000\n",
       "Comment[study step with treatment]         1\n",
       "Protocol REF                               1\n",
       "Performer                                  1\n",
       "Extract Name                            2000\n",
       "Characteristics[extract type]              1\n",
       "Protocol REF.1                             1\n",
       "Parameter Value[instrument]                1\n",
       "Parameter Value[library_orientation]       2\n",
       "Parameter Value[library_strategy]          1\n",
       "Performer.1                                1\n",
       "Raw Data File                           8000\n",
       "dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
