{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration with chained protocols in the assay workflow\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm repeated measures (crossover) study design\n",
    "- 4 subject groups: \"M\" and \"F\", \"healthy\" and \"diseased\"\n",
    "- 2 treatments in crossover: a drug treatment with ibuprofen and a biological treatment with injection of KpJH46Φ2 bacteriophage\n",
    "- 2 samples collected: blood and derma samples\n",
    "- three assays performed: mass. spec. on blood samples, Chip-seq and NMR on derma samples. Please note that the Chip-seq assay contains two chained protocols in its workflow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'Chained protocols study',\n",
       " 'subjectType': 'Homo sapiens',\n",
       " 'subjectSize': 10,\n",
       " 'designType': {'term': 'crossover design',\n",
       "  'id': 'OBI:0500003',\n",
       "  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',\n",
       "  'label': 'Study subjects receive repeated treatments',\n",
       "  'value': 'crossover'},\n",
       " 'observationalFactors': [{'name': 'sex',\n",
       "   'values': ['M', 'F'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None},\n",
       "  {'name': 'condition',\n",
       "   'values': ['healthy', 'diseased'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None}],\n",
       " 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_1',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_2',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_3',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]}],\n",
       "  'unselected': []},\n",
       " 'treatmentPlan': {'screen': {'selected': True,\n",
       "   'name': 'screen',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  'runIn': {'selected': True,\n",
       "   'name': 'run-in',\n",
       "   'duration': 2,\n",
       "   'durationUnit': 'days'},\n",
       "  'washout': {'selected': True,\n",
       "   'name': 'washout',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  'followUp': {'selected': True,\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'},\n",
       "  'treatments': [{'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'short_form': 'ERO_0000600',\n",
       "     'obo_id': 'ERO:0000600',\n",
       "     'label': 'drug intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg/day',\n",
       "    'duration': 12,\n",
       "    'durationUnit': 'days'},\n",
       "   {'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0001553',\n",
       "     'short_form': 'ERO_0001553',\n",
       "     'obo_id': 'ERO:0001553',\n",
       "     'label': 'biologic intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'KpJH46Φ2 phage',\n",
       "    'intensity': 3,\n",
       "    'intensityUnit': 'injections/day',\n",
       "    'duration': 14,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'interventionType': None,\n",
       "   'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'elements': [{'id': 'd9c13049-a235-41e2-a3be-f929e54ee2ee',\n",
       "   'name': 'screen',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '39e80594-b9ee-47f2-baf1-b954d7c1b124',\n",
       "   'name': 'run-in',\n",
       "   'duration': 2,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': 'ca5c2ebf-facf-46af-b698-99b81a28d88b',\n",
       "   'name': 'treatment_0',\n",
       "   'interventionType': {'term': 'drug intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000600'},\n",
       "   'duration': 12,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'ibuprofen',\n",
       "   'intensity': 8,\n",
       "   'intensityUnit': 'mg/day'},\n",
       "  {'id': 'b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7',\n",
       "   'name': 'washout',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '200e4d48-4ab2-492c-abd8-7b59a01b39d5',\n",
       "   'name': 'treatment_1',\n",
       "   'interventionType': {'term': 'biologic intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0001553'},\n",
       "   'duration': 14,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'KpJH46Φ2 phage',\n",
       "   'intensity': 3,\n",
       "   'intensityUnit': 'injections/day'},\n",
       "  {'id': 'e5664aeb-182f-49e2-849b-82e4c5e269e3',\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'}],\n",
       " 'arms': {'selected': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/7',\n",
       "    'name': 'Arm_7',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]}],\n",
       "  'unselected': [{'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'diseased',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]},\n",
       "   {'id': '#arm/6',\n",
       "    'name': 'Arm_6',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'condition',\n",
       "      'value': 'healthy',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['d9c13049-a235-41e2-a3be-f929e54ee2ee'],\n",
       "      'events': []},\n",
       "     {'elements': ['39e80594-b9ee-47f2-baf1-b954d7c1b124'], 'events': []},\n",
       "     {'elements': ['ca5c2ebf-facf-46af-b698-99b81a28d88b'], 'events': []},\n",
       "     {'elements': ['b9ce4ff5-7dd4-42c0-bf6b-9e0cf3d605e7'], 'events': []},\n",
       "     {'elements': ['200e4d48-4ab2-492c-abd8-7b59a01b39d5'], 'events': []},\n",
       "     {'elements': ['e5664aeb-182f-49e2-849b-82e4c5e269e3'], 'events': []}]}]},\n",
       " 'samplePlan': [{'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, True, True, False, True, True],\n",
       "    'Arm_2': [False, True, True, False, True, True],\n",
       "    'Arm_5': [False, True, True, False, True, True],\n",
       "    'Arm_7': [False, True, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_2': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_5': [None, 1, 1, None, 1, 6],\n",
       "    'Arm_7': [None, 1, 1, None, 1, 6]}},\n",
       "  {'sampleType': 'derma',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 2],\n",
       "    'Arm_2': [None, None, None, None, None, 2],\n",
       "    'Arm_5': [None, None, None, None, None, 2],\n",
       "    'Arm_7': [None, None, None, None, None, 2]}}],\n",
       " 'assayPlan': [{'id': 0,\n",
       "   'name': 'metabolite profiling using mass spectrometry',\n",
       "   'icon': 'fas fa-chart-bar fas fa-bullseye',\n",
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 2}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'non-polar fraction'],\n",
       "       'values': ['polar fraction', 'non-polar fraction']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labeling', {'#replicates': {'value': 1}}],\n",
       "    ['labeled extract',\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': 'labeled extract type',\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': ['label_0']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': ['Agilent QTQF 6510']},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': ['FIA', 'LC']},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': ['positive mode']}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   'selectedCells': {'Arm_0': [False, True, True, False, False, True],\n",
       "    'Arm_2': [False, True, True, False, False, True],\n",
       "    'Arm_5': [False, True, True, False, False, True],\n",
       "    'Arm_7': [False, True, True, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['blood']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['blood']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['blood']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'metabolite profiling using NMR spectroscopy',\n",
       "   'icon': 'fas fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
       "       'values': ['supernatant', 'pellet']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nmr_spectroscopy',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Bruker Avance II 1 GHz'],\n",
       "       'values': ['Bruker Avance II 1 GHz']},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': ['1D 13C NMR', '2D 13C-13C NMR']},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': ['TOCSY', 'watergate']}}],\n",
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['derma']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['derma']]}},\n",
       "  {'id': 16,\n",
       "   'name': 'Chip-Seq',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'protein-DNA binding site identification',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'cross linking': {'options': ['uv-light',\n",
       "        'formaldehyde',\n",
       "        'di-tert-butyl peroxide',\n",
       "        'formaldehyde and di-tert-butyl peroxyde'],\n",
       "       'values': ['uv-light', 'formaldehyde']},\n",
       "      'DNA fragmentation': {'options': ['sonication',\n",
       "        'nebulization',\n",
       "        'micrococcal nuclease digestion',\n",
       "        'deoxyribonuclease digestion'],\n",
       "       'values': ['nebulization']},\n",
       "      'DNA fragment size': {'newValues': True,\n",
       "       'isQuantitative': True,\n",
       "       'options': ['bbd', '100 nm'],\n",
       "       'values': ['100 nm']},\n",
       "      'immunoprecipitation antibody': {'newValues': True,\n",
       "       'isQuantitative': True,\n",
       "       'options': ['monoclonal'],\n",
       "       'values': ['monoclonal']}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['GENOMIC',\n",
       "        'GENOMIC SINGLE CELL',\n",
       "        'OTHER'],\n",
       "       'values': ['GENOMIC']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library preparation',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'library orientation': {'options': ['single-end', 'paired-end'],\n",
       "       'values': ['single-end']},\n",
       "      'library selection': {'options': ['PCR',\n",
       "        'RANDOM-PCR',\n",
       "        'ChIP',\n",
       "        'MSLL',\n",
       "        'Restriction Digest',\n",
       "        'other',\n",
       "        'unspecified'],\n",
       "       'values': ['PCR', 'ChIP']},\n",
       "      'library strategy': {'options': ['Chip-Seq',\n",
       "        'Hi-C',\n",
       "        'ATAC-seq',\n",
       "        'ChIA-PET',\n",
       "        'Tethered Chromatin Conformation Capture',\n",
       "        'FAIRE-seq',\n",
       "        'OTHER'],\n",
       "       'values': ['Chip-Seq']}}],\n",
       "    ['nucleic acid sequencing',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'sequencing instrument': {'options': ['454 GS',\n",
       "        '454 GS 20',\n",
       "        '454 GS FLX',\n",
       "        '454 GS FLX+',\n",
       "        '454 GS FLX Titanium',\n",
       "        '454 GS Junior',\n",
       "        'HiSeq X Five',\n",
       "        'HiSeq X Ten',\n",
       "        'Illumina Genome Analyzer',\n",
       "        'Illumina Genome Analyzer II',\n",
       "        'Illumina Genome Analyzer IIx',\n",
       "        'Illumina HiScanSQ',\n",
       "        'Illumina HiSeq 1000',\n",
       "        'Illumina HiSeq 1500',\n",
       "        'Illumina HiSeq 2000',\n",
       "        'Illumina HiSeq 2500',\n",
       "        'Illumina HiSeq 3000',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina iSeq 100',\n",
       "        'Illumina MiSeq',\n",
       "        'Illumina MiniSeq',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'NextSeq 500',\n",
       "        'NextSeq 550',\n",
       "        'PacBio RS',\n",
       "        'PacBio RS II',\n",
       "        'Sequel',\n",
       "        'Ion Torrent PGM',\n",
       "        'Ion Torrent Proton',\n",
       "        'Ion Torrent S5',\n",
       "        'Ion Torrent S5 XL',\n",
       "        'AB 3730xL Genetic Analyzer',\n",
       "        'AB 3730 Genetic Analyzer',\n",
       "        'AB 3500xL Genetic Analyzer',\n",
       "        'AB 3500 Genetic Analyzer',\n",
       "        'AB 3130xL Genetic Analyzer',\n",
       "        'AB 3130 Genetic Analyzer',\n",
       "        'AB 310 Genetic Analyzer',\n",
       "        'MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'BGISEQ-500',\n",
       "        'DNBSEQ-T7',\n",
       "        'DNBSEQ-G400',\n",
       "        'DNBSEQ-G50',\n",
       "        'DNBSEQ-G400 FAST',\n",
       "        'unspecified'],\n",
       "       'values': ['Illumina HiSeq 4000', 'Ion Torrent S5 XL']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_5': [False, False, False, False, False, True],\n",
       "    'Arm_7': [False, False, False, False, False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [[],\n",
       "     ['blood'],\n",
       "     ['blood'],\n",
       "     [],\n",
       "     ['blood'],\n",
       "     ['derma']],\n",
       "    'Arm_2': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_5': [[], ['blood'], ['blood'], [], ['blood'], ['derma']],\n",
       "    'Arm_7': [[], ['blood'], ['blood'], [], ['blood'], ['derma']]}}]}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"crossover-study-design-4-arms-blood-derma-nmr-ms-chipseq.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 7.56 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 1.58 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'crossover-bio+drug-treatment'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory,'isa-investigationn-crossover-bio+drug.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab\n",
    "\n",
    "Expect this to take up to a few minutes, depending on your machine (on mine, with an 8-th gen i7 and 16 GB of RAM it took 90s)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 114.08 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 3 assay files, one per assay type in our assay plan:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT16_protein-DNA-binding-site-identification_nucleic-acid-sequencing.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT2_metabolite-profiling_NMR-spectroscopy.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT0_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 10 subjects in the each of the four selected arms (Arm_0, Arm_2, Arm_5, Arm_7) and 11 samples have been collected per subject (9 blood samples per subject, 6 of which during the follow-up epoch + 3 derma samples collected during the follow-up phase)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 110 samples in the GRP0 arm (i.e. group)\n",
      "There are 110 samples in the GRP2 arm (i.e. group)\n",
      "There are 110 samples in the GRP3 arm (i.e. group)\n",
      "There are 110 samples in the GRP4 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP5' in el)])\n",
    "count_arm4_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP7' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))\n",
    "print(\"There are {} samples in the GRP4 arm (i.e. group)\".format(count_arm4_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Chip-seq assay table\n",
    "\n",
    "The two chained protocols have been serialised successfully."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Parameter Value[cross linking]</th>\n",
       "      <th>Parameter Value[DNA fragmentation]</th>\n",
       "      <th>Parameter Value[DNA fragment size]</th>\n",
       "      <th>Parameter Value[immunoprecipitation antibody]</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[library orientation]</th>\n",
       "      <th>Parameter Value[library selection]</th>\n",
       "      <th>Parameter Value[library strategy]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[sequencing instrument]</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S19-nucleic-acid-sequencing-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-raw_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S19-nucleic-acid-sequencing-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S99-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S99-nucleic-acid-sequencing-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S99-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S19-nucleic-acid-sequencing-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S19-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>uv-light</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S99-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S99-nucleic-acid-sequencing-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S99-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S66-nucleic-acid-sequencing-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S66-nucleic-acid-sequencing-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-raw_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>PCR</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S66-nucleic-acid-sequencing-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>AT16-S66-nucleic-acid-sequencing-Acquisition-R5</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-raw_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>formaldehyde</td>\n",
       "      <td>nebulization</td>\n",
       "      <td>100 nm</td>\n",
       "      <td>monoclonal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-Extract-R1</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>library preparation</td>\n",
       "      <td>single-end</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>Chip-Seq</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>nucleic acid sequencing</td>\n",
       "      <td>Ion Torrent S5 XL</td>\n",
       "      <td>AT16-S66-nucleic-acid-sequencing-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT16-S66-raw_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1280 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "1     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "2     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "3     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "4     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "...                           ...                                ...   \n",
       "1275  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1276  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1277  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1278  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1279  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "\n",
       "     Protocol REF Parameter Value[cross linking]  \\\n",
       "0      extraction                   formaldehyde   \n",
       "1      extraction                   formaldehyde   \n",
       "2      extraction                       uv-light   \n",
       "3      extraction                   formaldehyde   \n",
       "4      extraction                       uv-light   \n",
       "...           ...                            ...   \n",
       "1275   extraction                   formaldehyde   \n",
       "1276   extraction                   formaldehyde   \n",
       "1277   extraction                   formaldehyde   \n",
       "1278   extraction                   formaldehyde   \n",
       "1279   extraction                   formaldehyde   \n",
       "\n",
       "     Parameter Value[DNA fragmentation] Parameter Value[DNA fragment size]  \\\n",
       "0                          nebulization                             100 nm   \n",
       "1                          nebulization                             100 nm   \n",
       "2                          nebulization                             100 nm   \n",
       "3                          nebulization                             100 nm   \n",
       "4                          nebulization                             100 nm   \n",
       "...                                 ...                                ...   \n",
       "1275                       nebulization                             100 nm   \n",
       "1276                       nebulization                             100 nm   \n",
       "1277                       nebulization                             100 nm   \n",
       "1278                       nebulization                             100 nm   \n",
       "1279                       nebulization                             100 nm   \n",
       "\n",
       "     Parameter Value[immunoprecipitation antibody] Performer  \\\n",
       "0                                       monoclonal   Unknown   \n",
       "1                                       monoclonal   Unknown   \n",
       "2                                       monoclonal   Unknown   \n",
       "3                                       monoclonal   Unknown   \n",
       "4                                       monoclonal   Unknown   \n",
       "...                                            ...       ...   \n",
       "1275                                    monoclonal   Unknown   \n",
       "1276                                    monoclonal   Unknown   \n",
       "1277                                    monoclonal   Unknown   \n",
       "1278                                    monoclonal   Unknown   \n",
       "1279                                    monoclonal   Unknown   \n",
       "\n",
       "             Extract Name Characteristics[extract type]       Protocol REF.1  \\\n",
       "0     AT16-S19-Extract-R1                       GENOMIC  library preparation   \n",
       "1     AT16-S19-Extract-R1                       GENOMIC  library preparation   \n",
       "2     AT16-S99-Extract-R1                       GENOMIC  library preparation   \n",
       "3     AT16-S19-Extract-R1                       GENOMIC  library preparation   \n",
       "4     AT16-S99-Extract-R1                       GENOMIC  library preparation   \n",
       "...                   ...                           ...                  ...   \n",
       "1275  AT16-S66-Extract-R1                       GENOMIC  library preparation   \n",
       "1276  AT16-S66-Extract-R1                       GENOMIC  library preparation   \n",
       "1277  AT16-S66-Extract-R1                       GENOMIC  library preparation   \n",
       "1278  AT16-S66-Extract-R1                       GENOMIC  library preparation   \n",
       "1279  AT16-S66-Extract-R1                       GENOMIC  library preparation   \n",
       "\n",
       "     Parameter Value[library orientation] Parameter Value[library selection]  \\\n",
       "0                              single-end                               ChIP   \n",
       "1                              single-end                                PCR   \n",
       "2                              single-end                                PCR   \n",
       "3                              single-end                                PCR   \n",
       "4                              single-end                               ChIP   \n",
       "...                                   ...                                ...   \n",
       "1275                           single-end                                PCR   \n",
       "1276                           single-end                               ChIP   \n",
       "1277                           single-end                                PCR   \n",
       "1278                           single-end                               ChIP   \n",
       "1279                           single-end                               ChIP   \n",
       "\n",
       "     Parameter Value[library strategy] Performer.1           Protocol REF.2  \\\n",
       "0                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "2                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "3                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "4                             Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "...                                ...         ...                      ...   \n",
       "1275                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1276                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1277                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1278                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "1279                          Chip-Seq     Unknown  nucleic acid sequencing   \n",
       "\n",
       "     Parameter Value[sequencing instrument]  \\\n",
       "0                       Illumina HiSeq 4000   \n",
       "1                       Illumina HiSeq 4000   \n",
       "2                         Ion Torrent S5 XL   \n",
       "3                         Ion Torrent S5 XL   \n",
       "4                       Illumina HiSeq 4000   \n",
       "...                                     ...   \n",
       "1275                      Ion Torrent S5 XL   \n",
       "1276                    Illumina HiSeq 4000   \n",
       "1277                    Illumina HiSeq 4000   \n",
       "1278                    Illumina HiSeq 4000   \n",
       "1279                      Ion Torrent S5 XL   \n",
       "\n",
       "                                           Assay Name Performer.2  \\\n",
       "0     AT16-S19-nucleic-acid-sequencing-Acquisition-R7     Unknown   \n",
       "1     AT16-S19-nucleic-acid-sequencing-Acquisition-R4     Unknown   \n",
       "2     AT16-S99-nucleic-acid-sequencing-Acquisition-R8     Unknown   \n",
       "3     AT16-S19-nucleic-acid-sequencing-Acquisition-R3     Unknown   \n",
       "4     AT16-S99-nucleic-acid-sequencing-Acquisition-R1     Unknown   \n",
       "...                                               ...         ...   \n",
       "1275  AT16-S66-nucleic-acid-sequencing-Acquisition-R1     Unknown   \n",
       "1276  AT16-S66-nucleic-acid-sequencing-Acquisition-R7     Unknown   \n",
       "1277  AT16-S66-nucleic-acid-sequencing-Acquisition-R4     Unknown   \n",
       "1278  AT16-S66-nucleic-acid-sequencing-Acquisition-R5     Unknown   \n",
       "1279  AT16-S66-nucleic-acid-sequencing-Acquisition-R8     Unknown   \n",
       "\n",
       "                      Raw Data File  \n",
       "0     AT16-S19-raw_data_file-R7.raw  \n",
       "1     AT16-S19-raw_data_file-R4.raw  \n",
       "2     AT16-S99-raw_data_file-R8.raw  \n",
       "3     AT16-S19-raw_data_file-R3.raw  \n",
       "4     AT16-S99-raw_data_file-R1.raw  \n",
       "...                             ...  \n",
       "1275  AT16-S66-raw_data_file-R1.raw  \n",
       "1276  AT16-S66-raw_data_file-R7.raw  \n",
       "1277  AT16-S66-raw_data_file-R4.raw  \n",
       "1278  AT16-S66-raw_data_file-R5.raw  \n",
       "1279  AT16-S66-raw_data_file-R8.raw  \n",
       "\n",
       "[1280 rows x 20 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT16_protein-DNA-binding-site-identification_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.2 NMR assay table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Parameter Value[pulse_sequence]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Free Induction Decay Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-raw_spectral_data_file-R12.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-raw_spectral_data_file-R11.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-Extract-R1</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-raw_spectral_data_file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-Extract-R1</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-raw_spectral_data_file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E5_SMP-derma-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S19-raw_spectral_data_file-R14.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-raw_spectral_data_file-R13.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-Extract-R1</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>watergate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-raw_spectral_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-raw_spectral_data_file-R12.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-Extract-R1</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-raw_spectral_data_file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-derma-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-Extract-R2</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr_spectroscopy</td>\n",
       "      <td>Bruker Avance II 1 GHz</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>TOCSY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT2-S66-raw_spectral_data_file-R9.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1280 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "1     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "2     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "3     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "4     GRP0_SBJ01_A0E5_SMP-derma-1                                 NO   \n",
       "...                           ...                                ...   \n",
       "1275  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1276  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1277  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1278  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "1279  GRP7_SBJ10_A3E5_SMP-derma-2                                 NO   \n",
       "\n",
       "     Protocol REF Performer        Extract Name Characteristics[extract type]  \\\n",
       "0      extraction   Unknown  AT2-S19-Extract-R2                   supernatant   \n",
       "1      extraction   Unknown  AT2-S19-Extract-R2                   supernatant   \n",
       "2      extraction   Unknown  AT2-S19-Extract-R1                        pellet   \n",
       "3      extraction   Unknown  AT2-S19-Extract-R1                        pellet   \n",
       "4      extraction   Unknown  AT2-S19-Extract-R2                   supernatant   \n",
       "...           ...       ...                 ...                           ...   \n",
       "1275   extraction   Unknown  AT2-S66-Extract-R2                   supernatant   \n",
       "1276   extraction   Unknown  AT2-S66-Extract-R1                        pellet   \n",
       "1277   extraction   Unknown  AT2-S66-Extract-R2                   supernatant   \n",
       "1278   extraction   Unknown  AT2-S66-Extract-R1                        pellet   \n",
       "1279   extraction   Unknown  AT2-S66-Extract-R2                   supernatant   \n",
       "\n",
       "        Protocol REF.1 Parameter Value[instrument]  \\\n",
       "0     nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "1     nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "2     nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "3     nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "4     nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "...                ...                         ...   \n",
       "1275  nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "1276  nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "1277  nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "1278  nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "1279  nmr_spectroscopy      Bruker Avance II 1 GHz   \n",
       "\n",
       "     Parameter Value[acquisition_mode] Parameter Value[pulse_sequence]  \\\n",
       "0                       2D 13C-13C NMR                           TOCSY   \n",
       "1                       2D 13C-13C NMR                           TOCSY   \n",
       "2                       2D 13C-13C NMR                           TOCSY   \n",
       "3                       2D 13C-13C NMR                           TOCSY   \n",
       "4                       2D 13C-13C NMR                       watergate   \n",
       "...                                ...                             ...   \n",
       "1275                    2D 13C-13C NMR                       watergate   \n",
       "1276                    2D 13C-13C NMR                       watergate   \n",
       "1277                    2D 13C-13C NMR                           TOCSY   \n",
       "1278                        1D 13C NMR                           TOCSY   \n",
       "1279                        1D 13C NMR                           TOCSY   \n",
       "\n",
       "     Performer.1          Free Induction Decay Data File  \n",
       "0        Unknown  AT2-S19-raw_spectral_data_file-R12.raw  \n",
       "1        Unknown  AT2-S19-raw_spectral_data_file-R11.raw  \n",
       "2        Unknown   AT2-S19-raw_spectral_data_file-R7.raw  \n",
       "3        Unknown   AT2-S19-raw_spectral_data_file-R8.raw  \n",
       "4        Unknown  AT2-S19-raw_spectral_data_file-R14.raw  \n",
       "...          ...                                     ...  \n",
       "1275     Unknown  AT2-S66-raw_spectral_data_file-R13.raw  \n",
       "1276     Unknown   AT2-S66-raw_spectral_data_file-R4.raw  \n",
       "1277     Unknown  AT2-S66-raw_spectral_data_file-R12.raw  \n",
       "1278     Unknown   AT2-S66-raw_spectral_data_file-R5.raw  \n",
       "1279     Unknown   AT2-S66-raw_spectral_data_file-R9.raw  \n",
       "\n",
       "[1280 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT2_metabolite-profiling_NMR-spectroscopy.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.3 Mass Spectrometry assay table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[study step with treatment]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Labeled Extract Name</th>\n",
       "      <th>Label</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[injection_mode]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>MS Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Spectral Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S20-mass-spectrometry-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-raw-spectral-data-file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S19-mass-spectrometry-Acquisition-R5</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-raw-spectral-data-file-R5.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S19-mass-spectrometry-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-raw-spectral-data-file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S20-mass-spectrometry-Acquisition-R6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S20-raw-spectral-data-file-R6.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ01_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S19-mass-spectrometry-Acquisition-R4</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S19-raw-spectral-data-file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5115</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S556-mass-spectrometry-Acquisition-R1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-raw-spectral-data-file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5116</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S555-mass-spectrometry-Acquisition-R7</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-raw-spectral-data-file-R7.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5117</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S556-mass-spectrometry-Acquisition-R6</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-raw-spectral-data-file-R6.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-Extract-R2</td>\n",
       "      <td>non-polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-LE-R2</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S555-mass-spectrometry-Acquisition-R8</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S555-raw-spectral-data-file-R8.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5119</th>\n",
       "      <td>GRP7_SBJ10_A3E5_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-Extract-R1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labeling</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-LE-R1</td>\n",
       "      <td>label_0</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>AT0-S556-mass-spectrometry-Acquisition-R3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT0-S556-raw-spectral-data-file-R3.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5120 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Sample Name Comment[study step with treatment]  \\\n",
       "0     GRP0_SBJ01_A0E1_SMP-blood-1                                 NO   \n",
       "1     GRP0_SBJ01_A0E1_SMP-blood-1                                 NO   \n",
       "2     GRP0_SBJ01_A0E1_SMP-blood-1                                 NO   \n",
       "3     GRP0_SBJ01_A0E1_SMP-blood-1                                 NO   \n",
       "4     GRP0_SBJ01_A0E1_SMP-blood-1                                 NO   \n",
       "...                           ...                                ...   \n",
       "5115  GRP7_SBJ10_A3E5_SMP-blood-6                                 NO   \n",
       "5116  GRP7_SBJ10_A3E5_SMP-blood-6                                 NO   \n",
       "5117  GRP7_SBJ10_A3E5_SMP-blood-6                                 NO   \n",
       "5118  GRP7_SBJ10_A3E5_SMP-blood-6                                 NO   \n",
       "5119  GRP7_SBJ10_A3E5_SMP-blood-6                                 NO   \n",
       "\n",
       "     Protocol REF Performer         Extract Name  \\\n",
       "0      extraction   Unknown   AT0-S20-Extract-R2   \n",
       "1      extraction   Unknown   AT0-S19-Extract-R2   \n",
       "2      extraction   Unknown   AT0-S19-Extract-R2   \n",
       "3      extraction   Unknown   AT0-S20-Extract-R2   \n",
       "4      extraction   Unknown   AT0-S19-Extract-R1   \n",
       "...           ...       ...                  ...   \n",
       "5115   extraction   Unknown  AT0-S556-Extract-R1   \n",
       "5116   extraction   Unknown  AT0-S555-Extract-R2   \n",
       "5117   extraction   Unknown  AT0-S556-Extract-R2   \n",
       "5118   extraction   Unknown  AT0-S555-Extract-R2   \n",
       "5119   extraction   Unknown  AT0-S556-Extract-R1   \n",
       "\n",
       "     Characteristics[extract type] Protocol REF.1 Performer.1  \\\n",
       "0               non-polar fraction       labeling     Unknown   \n",
       "1               non-polar fraction       labeling     Unknown   \n",
       "2               non-polar fraction       labeling     Unknown   \n",
       "3               non-polar fraction       labeling     Unknown   \n",
       "4                   polar fraction       labeling     Unknown   \n",
       "...                            ...            ...         ...   \n",
       "5115                polar fraction       labeling     Unknown   \n",
       "5116            non-polar fraction       labeling     Unknown   \n",
       "5117            non-polar fraction       labeling     Unknown   \n",
       "5118            non-polar fraction       labeling     Unknown   \n",
       "5119                polar fraction       labeling     Unknown   \n",
       "\n",
       "     Labeled Extract Name    Label     Protocol REF.2  \\\n",
       "0           AT0-S20-LE-R2  label_0  mass spectrometry   \n",
       "1           AT0-S19-LE-R2  label_0  mass spectrometry   \n",
       "2           AT0-S19-LE-R2  label_0  mass spectrometry   \n",
       "3           AT0-S20-LE-R2  label_0  mass spectrometry   \n",
       "4           AT0-S19-LE-R1  label_0  mass spectrometry   \n",
       "...                   ...      ...                ...   \n",
       "5115       AT0-S556-LE-R1  label_0  mass spectrometry   \n",
       "5116       AT0-S555-LE-R2  label_0  mass spectrometry   \n",
       "5117       AT0-S556-LE-R2  label_0  mass spectrometry   \n",
       "5118       AT0-S555-LE-R2  label_0  mass spectrometry   \n",
       "5119       AT0-S556-LE-R1  label_0  mass spectrometry   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[injection_mode]  \\\n",
       "0              Agilent QTQF 6510                              LC   \n",
       "1              Agilent QTQF 6510                             FIA   \n",
       "2              Agilent QTQF 6510                              LC   \n",
       "3              Agilent QTQF 6510                             FIA   \n",
       "4              Agilent QTQF 6510                              LC   \n",
       "...                          ...                             ...   \n",
       "5115           Agilent QTQF 6510                             FIA   \n",
       "5116           Agilent QTQF 6510                              LC   \n",
       "5117           Agilent QTQF 6510                             FIA   \n",
       "5118           Agilent QTQF 6510                              LC   \n",
       "5119           Agilent QTQF 6510                              LC   \n",
       "\n",
       "     Parameter Value[acquisition_mode]  \\\n",
       "0                        positive mode   \n",
       "1                        positive mode   \n",
       "2                        positive mode   \n",
       "3                        positive mode   \n",
       "4                        positive mode   \n",
       "...                                ...   \n",
       "5115                     positive mode   \n",
       "5116                     positive mode   \n",
       "5117                     positive mode   \n",
       "5118                     positive mode   \n",
       "5119                     positive mode   \n",
       "\n",
       "                                  MS Assay Name Performer.2  \\\n",
       "0      AT0-S20-mass-spectrometry-Acquisition-R7     Unknown   \n",
       "1      AT0-S19-mass-spectrometry-Acquisition-R5     Unknown   \n",
       "2      AT0-S19-mass-spectrometry-Acquisition-R7     Unknown   \n",
       "3      AT0-S20-mass-spectrometry-Acquisition-R6     Unknown   \n",
       "4      AT0-S19-mass-spectrometry-Acquisition-R4     Unknown   \n",
       "...                                         ...         ...   \n",
       "5115  AT0-S556-mass-spectrometry-Acquisition-R1     Unknown   \n",
       "5116  AT0-S555-mass-spectrometry-Acquisition-R7     Unknown   \n",
       "5117  AT0-S556-mass-spectrometry-Acquisition-R6     Unknown   \n",
       "5118  AT0-S555-mass-spectrometry-Acquisition-R8     Unknown   \n",
       "5119  AT0-S556-mass-spectrometry-Acquisition-R3     Unknown   \n",
       "\n",
       "                      Raw Spectral Data File  \n",
       "0      AT0-S20-raw-spectral-data-file-R7.raw  \n",
       "1      AT0-S19-raw-spectral-data-file-R5.raw  \n",
       "2      AT0-S19-raw-spectral-data-file-R7.raw  \n",
       "3      AT0-S20-raw-spectral-data-file-R6.raw  \n",
       "4      AT0-S19-raw-spectral-data-file-R4.raw  \n",
       "...                                      ...  \n",
       "5115  AT0-S556-raw-spectral-data-file-R1.raw  \n",
       "5116  AT0-S555-raw-spectral-data-file-R7.raw  \n",
       "5117  AT0-S556-raw-spectral-data-file-R6.raw  \n",
       "5118  AT0-S555-raw-spectral-data-file-R8.raw  \n",
       "5119  AT0-S556-raw-spectral-data-file-R3.raw  \n",
       "\n",
       "[5120 rows x 17 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT0_metabolite-profiling_mass-spectrometry.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
