{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "# Crossover Study with radiological and drug treatment on Mice\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm study design. Each arm has 10 subjects\n",
    "- subjects are mice. There is an observational factor, named \"sex\" with two values: \"M\" and \"F\"\n",
    "- a crossover of two treatments, a drug treatment (cyclophosphamide 500 mg for 20 days) and a radiological treatment (\"strontium chloride sr 89\" 2 MBq/Kg forr 40 days)\n",
    "- four non-treatment phases: screen, run-in, washout and follow-up\n",
    "- three sample types colllected: tissue, blood, and urine\n",
    "- three assay types: \n",
    "    - metabolite profiling through mass spectrometry on the urine sample.\n",
    "    - clinical chemistry with marker on blood samples\n",
    "    - genomic sequencing on tissue samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator with all the specs defined above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'Test study',\n",
       " 'description': 'This design describes a study with repeated treatments on the same arms. We will be collecting three samples and these will undergo three different assays.',\n",
       " 'subjectType': 'Mouse',\n",
       " 'subjectSize': 10,\n",
       " 'designType': {'term': 'crossover design',\n",
       "  'id': 'OBI:0500003',\n",
       "  'iri': 'http://purl.obolibrary.org/obo/OBI_0500003',\n",
       "  'label': 'Study subjects receive repeated treatments',\n",
       "  'value': 'crossover'},\n",
       " 'observationalFactors': [{'name': 'sex',\n",
       "   'values': ['M', 'F'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None}],\n",
       " 'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "    'type': 'Mouse',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]},\n",
       "   {'name': 'SubjectGroup_1',\n",
       "    'type': 'Mouse',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}]}],\n",
       "  'unselected': []},\n",
       " 'treatmentPlan': {'screen': {'selected': True,\n",
       "   'name': 'screen',\n",
       "   'duration': 10,\n",
       "   'durationUnit': 'days'},\n",
       "  'runIn': {'selected': True,\n",
       "   'name': 'run-in',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  'washout': {'selected': True,\n",
       "   'name': 'washout',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  'followUp': {'selected': True,\n",
       "   'name': 'follow-up',\n",
       "   'duration': 3,\n",
       "   'durationUnit': 'months'},\n",
       "  'treatments': [{'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000599',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000599',\n",
       "     'short_form': 'ERO_0000599',\n",
       "     'obo_id': 'ERO:0000599',\n",
       "     'label': 'dietary intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'strontium chloride sr 89',\n",
       "    'intensity': 2,\n",
       "    'intensityUnit': 'MBq/Kg',\n",
       "    'duration': 40,\n",
       "    'durationUnit': 'days'},\n",
       "   {'interventionType': {'id': 'ero:class:http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'iri': 'http://purl.obolibrary.org/obo/ERO_0000600',\n",
       "     'short_form': 'ERO_0000600',\n",
       "     'obo_id': 'ERO:0000600',\n",
       "     'label': 'drug intervention',\n",
       "     'ontology_name': 'ero',\n",
       "     'ontology_prefix': 'ERO',\n",
       "     'type': 'class'},\n",
       "    'agent': 'cyclophosphamide',\n",
       "    'intensity': 500,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 20,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'interventionType': None,\n",
       "   'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'elements': [{'id': 'f2f3b834-0555-4d2c-8aa9-835df184d188',\n",
       "   'name': 'screen',\n",
       "   'duration': 10,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '482fe6ce-0ddb-4c9e-9da1-063805f534b7',\n",
       "   'name': 'run-in',\n",
       "   'duration': 5,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '59e2cfb5-0133-4f3f-af2c-000e2c991744',\n",
       "   'name': 'treatment_0',\n",
       "   'interventionType': {'term': 'dietary intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000599'},\n",
       "   'duration': 40,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'strontium chloride sr 89',\n",
       "   'intensity': 2,\n",
       "   'intensityUnit': 'MBq/Kg'},\n",
       "  {'id': '7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb',\n",
       "   'name': 'washout',\n",
       "   'duration': 7,\n",
       "   'durationUnit': 'days'},\n",
       "  {'id': '2465e61b-7b64-4afb-8c14-edce966bffbe',\n",
       "   'name': 'treatment_1',\n",
       "   'interventionType': {'term': 'drug intervention',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/ERO_0000600'},\n",
       "   'duration': 20,\n",
       "   'durationUnit': 'days',\n",
       "   'agent': 'cyclophosphamide',\n",
       "   'intensity': 500,\n",
       "   'intensityUnit': 'mg'},\n",
       "  {'id': '4e6771bf-a4a5-46a9-a7c6-62d04ab3e126',\n",
       "   'name': 'follow-up',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'}],\n",
       " 'arms': {'selected': [{'id': 0,\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 1,\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 2,\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]},\n",
       "   {'id': 3,\n",
       "    'name': 'Arm_3',\n",
       "    'observationalFactors': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False}],\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['f2f3b834-0555-4d2c-8aa9-835df184d188'],\n",
       "      'events': []},\n",
       "     {'elements': ['482fe6ce-0ddb-4c9e-9da1-063805f534b7'], 'events': []},\n",
       "     {'elements': ['2465e61b-7b64-4afb-8c14-edce966bffbe'], 'events': []},\n",
       "     {'elements': ['7fb4d4c4-f7e5-4a16-86af-53c7e96eb5fb'], 'events': []},\n",
       "     {'elements': ['59e2cfb5-0133-4f3f-af2c-000e2c991744'], 'events': []},\n",
       "     {'elements': ['4e6771bf-a4a5-46a9-a7c6-62d04ab3e126'], 'events': []}]}],\n",
       "  'unselected': []},\n",
       " 'samplePlan': [{'sampleType': 'Blood Sample',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_1': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, None, 2, None, 2, 3],\n",
       "    'Arm_1': [1, None, 2, None, 2, 3],\n",
       "    'Arm_2': [1, None, 2, None, 2, 3],\n",
       "    'Arm_3': [1, None, 2, None, 2, 3]}},\n",
       "  {'sampleType': 'Tissue Sample',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, True, False],\n",
       "    'Arm_1': [False, False, True, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, True, False],\n",
       "    'Arm_3': [False, False, True, False, False, False]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, 1, None],\n",
       "    'Arm_1': [None, None, 1, None, None, None],\n",
       "    'Arm_2': [None, None, None, None, 1, None],\n",
       "    'Arm_3': [None, None, 1, None, None, None]}},\n",
       "  {'sampleType': 'Urine',\n",
       "   'selectedCells': {'Arm_0': [True, True, True, False, True, True],\n",
       "    'Arm_1': [True, True, True, False, True, True],\n",
       "    'Arm_2': [True, True, True, False, True, True],\n",
       "    'Arm_3': [True, True, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_1': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_2': [2, 1, 1, None, 1, 2],\n",
       "    'Arm_3': [2, 1, 1, None, 1, 2]}}],\n",
       " 'assayPlan': [{'id': 1,\n",
       "   'name': 'metabolite profiling - mass spectrometry (ontology annotated)',\n",
       "   'icon': 'fas fa-chart-bar',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000423',\n",
       "       'source': 'OBI'},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': 'lipids', 'source': None}],\n",
       "       'values': [{'term': 'polar fraction',\n",
       "         'iri': 'polar fraction',\n",
       "         'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labeling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labeled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labeled extract type',\n",
       "       'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': ['label_0', 'label_1'],\n",
       "       'values': ['label_0']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None}]},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}],\n",
       "       'values': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}]},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}],\n",
       "       'values': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}]}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': None},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [True, True, True, False, True, True],\n",
       "    'Arm_1': [True, True, True, False, True, True],\n",
       "    'Arm_2': [True, True, True, False, True, True],\n",
       "    'Arm_3': [True, True, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Urine'],\n",
       "     ['Urine'],\n",
       "     ['Urine'],\n",
       "     [],\n",
       "     ['Urine'],\n",
       "     ['Urine']],\n",
       "    'Arm_1': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']],\n",
       "    'Arm_2': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']],\n",
       "    'Arm_3': [['Urine'], ['Urine'], ['Urine'], [], ['Urine'], ['Urine']]}},\n",
       "  {'id': 8,\n",
       "   'name': 'genome sequencing by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'genome sequencing',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 2}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'],\n",
       "       'values': ['gDNA', 'DNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': ['Illumina NovaSeq 6000']},\n",
       "      'library_orientation': {'options': ['single', 'paired'],\n",
       "       'values': ['single']},\n",
       "      'library_strategy': {'options': ['AMPLICON',\n",
       "        'CLONE',\n",
       "        'WGA',\n",
       "        'WGS',\n",
       "        'WXS',\n",
       "        'OTHER'],\n",
       "       'values': ['AMPLICON', 'WGS']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, True, False],\n",
       "    'Arm_1': [False, False, True, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, True, False],\n",
       "    'Arm_3': [False, False, True, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     [],\n",
       "     ['Tissue Sample'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_1': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Tissue Sample'],\n",
       "     [],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_2': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     [],\n",
       "     ['Tissue Sample'],\n",
       "     ['Blood Sample', 'Urine']],\n",
       "    'Arm_3': [['Blood Sample', 'Urine'],\n",
       "     ['Urine'],\n",
       "     ['Tissue Sample'],\n",
       "     [],\n",
       "     ['Blood Sample', 'Urine'],\n",
       "     ['Blood Sample', 'Urine']]}},\n",
       "  {'id': 11,\n",
       "   'name': 'hematology by marker panel',\n",
       "   'icon': 'fa fa-flask',\n",
       "   'color': 'red',\n",
       "   'measurement_type': 'clinical chemistry',\n",
       "   'technology_type': 'marker panel',\n",
       "   'workflow': [['sample preparation', {'#replicates': {'value': 3}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_1': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_1': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_2': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']],\n",
       "    'Arm_3': [['Blood Sample'],\n",
       "     ['Urine'],\n",
       "     ['Blood Sample'],\n",
       "     [],\n",
       "     ['Blood Sample'],\n",
       "     ['Blood Sample']]}}]}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"crossover-study-design-2treatments-mice.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 1.70 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(identifier='inv01', studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 0.71 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'crossover-2treatments-mice'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory, 'isa-investigation-crossover-2treatments-mice.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1736, in get_pv_columns\n",
      "    label, pv.category.parameter_name.term)]\n",
      "AttributeError: 'NoneType' object has no attribute 'parameter_name'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/logging/__init__.py\", line 1081, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/logging/__init__.py\", line 925, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/logging/__init__.py\", line 664, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/logging/__init__.py\", line 369, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/runpy.py\", line 193, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/runpy.py\", line 86, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel_launcher.py\", line 16, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/traitlets/config/application.py\", line 664, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/kernelapp.py\", line 583, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/platform/asyncio.py\", line 149, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/asyncio/base_events.py\", line 570, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/asyncio/base_events.py\", line 1859, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/lib/python3.8/asyncio/events.py\", line 81, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/ioloop.py\", line 690, in <lambda>\n",
      "    lambda f: self._run_callback(functools.partial(callback, future))\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/ioloop.py\", line 743, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 787, in inner\n",
      "    self.run()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 748, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/kernelbase.py\", line 381, in dispatch_queue\n",
      "    yield self.process_one()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 225, in wrapper\n",
      "    runner = Runner(result, future, yielded)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 714, in __init__\n",
      "    self.run()\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 748, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/kernelbase.py\", line 365, in process_one\n",
      "    yield gen.maybe_future(dispatch(*args))\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 209, in wrapper\n",
      "    yielded = next(result)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/kernelbase.py\", line 268, in dispatch_shell\n",
      "    yield gen.maybe_future(handler(stream, idents, msg))\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 209, in wrapper\n",
      "    yielded = next(result)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/kernelbase.py\", line 543, in execute_request\n",
      "    self.do_execute(\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/tornado/gen.py\", line 209, in wrapper\n",
      "    yielded = next(result)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/ipkernel.py\", line 300, in do_execute\n",
      "    res = shell.run_cell(code, store_history=store_history, silent=silent)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/ipykernel/zmqshell.py\", line 536, in run_cell\n",
      "    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2857, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2886, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3062, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3254, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3331, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-10-c7e41802aff0>\", line 2, in <module>\n",
      "    isatab.dump(investigation, directory)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1049, in dump\n",
      "    write_study_table_files(investigation, output_path)\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1206, in write_study_table_files\n",
      "    columns += flatten(map(lambda x: get_pv_columns(olabel, x),\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1182, in flatten\n",
      "    def flatten(l): return [item for sublist in l for item in sublist]\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1182, in <listcomp>\n",
      "    def flatten(l): return [item for sublist in l for item in sublist]\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1206, in <lambda>\n",
      "    columns += flatten(map(lambda x: get_pv_columns(olabel, x),\n",
      "  File \"/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\", line 1738, in get_pv_columns\n",
      "    log.fatal(label, pv)\n",
      "Message: 'Protocol REF.sample collection'\n",
      "Arguments: (isatools.model.ParameterValue(category=None, value='041', unit=None, comments=[]),)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'extend'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-c7e41802aff0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0misatab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minvestigation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdirectory\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'The Tab serialisation of the ISA investigation took {:.2f} s.'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mend\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mdump\u001b[0;34m(isa_obj, output_path, i_file_name, skip_dump_tables, write_factor_values_in_assay_table)\u001b[0m\n\u001b[1;32m   1047\u001b[0m         \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1048\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1049\u001b[0;31m         \u001b[0mwrite_study_table_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minvestigation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1050\u001b[0m         write_assay_table_files(\n\u001b[1;32m   1051\u001b[0m             investigation, output_path, write_factor_values_in_assay_table)\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mwrite_study_table_files\u001b[0;34m(inv_obj, output_dir)\u001b[0m\n\u001b[1;32m   1204\u001b[0m                     \u001b[0mprotnames\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprotrefcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1205\u001b[0m                     \u001b[0mprotrefcount\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1206\u001b[0;31m                 columns += flatten(map(lambda x: get_pv_columns(olabel, x),\n\u001b[0m\u001b[1;32m   1207\u001b[0m                                        node.parameter_values))\n\u001b[1;32m   1208\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdate\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mflatten\u001b[0;34m(l)\u001b[0m\n\u001b[1;32m   1180\u001b[0m         \u001b[0mprotnames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1181\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1182\u001b[0;31m         \u001b[0;32mdef\u001b[0m \u001b[0mflatten\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msublist\u001b[0m \u001b[0;32min\u001b[0m \u001b[0ml\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msublist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1183\u001b[0m         \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1180\u001b[0m         \u001b[0mprotnames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1181\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1182\u001b[0;31m         \u001b[0;32mdef\u001b[0m \u001b[0mflatten\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msublist\u001b[0m \u001b[0;32min\u001b[0m \u001b[0ml\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msublist\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1183\u001b[0m         \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   1204\u001b[0m                     \u001b[0mprotnames\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprotrefcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1205\u001b[0m                     \u001b[0mprotrefcount\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1206\u001b[0;31m                 columns += flatten(map(lambda x: get_pv_columns(olabel, x),\n\u001b[0m\u001b[1;32m   1207\u001b[0m                                        node.parameter_values))\n\u001b[1;32m   1208\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdate\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mget_pv_columns\u001b[0;34m(label, pv)\u001b[0m\n\u001b[1;32m   1737\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m         \u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfatal\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1739\u001b[0;31m     \u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_value_columns\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1740\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1741\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'extend'"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Count of subjects and samples\n",
    "\n",
    "We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the \"mass spcetrometry\" assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP1 arm (i.e. group)\".format(count_arm1_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT11_clinical-chemistry_marker-panel.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7.1 Overview of the Mass Spec assay table\n",
    "\n",
    "For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, \"polar\" fraction), 320 labeled extracts (2 per extract, as \"#replicates\" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `[\"Agilent QTQF 6510\", \"FIA\", \"positive mode\"]` and `[\"Agilent QTQF 6510\", \"LC\", \"positive mode\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT1_metabolite-profiling_mass-spectrometry.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Genome Sequencing assay table\n",
    "\n",
    "For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the \"supernatant\" and \"pellet\" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"CPGM\"]` and `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"TOCSY\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
