{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "[missing]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "log_level: error\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/sampletab.py:147: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if term_source_value is not '':\n",
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/sampletab.py:156: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if term_accession_value is not '':\n",
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/sampletab.py:180: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if unit_term_source_value is not '':\n",
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/sampletab.py:191: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if term_accession_value is not '':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOG: <Logger isatools (DEBUG)>\n",
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/net/resources/saxon9/saxon9he.jar\n",
      "[{'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'targeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'untargeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'isotopomer distribution analysis', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'targeted metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'untargeted metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'isotologue distribution analysis', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': 'Derived Spectral Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'genotype profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'epigenome profling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'exome profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'DNA methylation profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'copy number variation profling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'transcription factor binding site identification', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'protein-DNA binding site identification', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'SNP analysis', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'transcription factor binding site identification', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'protein-DNA binding site identification', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'DNA methylation profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'histone modification profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'genome sequencing', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'metagenome sequencing', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'environmental gene survey', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell sorting', 'technology type': 'flow cytometry', 'protocol type': 'flow cytometry', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell counting', 'technology type': 'flow cytometry', 'protocol type': 'flow cytometry', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell migration assay', 'technology type': 'microscopy imaging', 'protocol type': 'migration assay', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'phenotyping', 'technology type': 'imaging', 'protocol type': 'migration assay', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'RT-pcr', 'protocol type': 'RT-pcr', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': None, 'technology type': None, 'protocol type': None, 'parameter-like file': None, 'node name': None, 'key in json_configuration': 'my_favourite_institution_tag (no white space allowed)', 'raw data file': None, 'derived data file': None}]\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json\n",
    "\n",
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study\n",
    "\n",
    "## ISA-API create mode related imports\n",
    "from isatools.create.models import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design_from_config\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import models\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'treatmentPlan': {'screen': {'name': 'screen',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'runIn': {'name': 'run-in', 'duration': None, 'durationUnit': ''},\n",
       "  'washout': {'name': 'washout', 'duration': '5', 'durationUnit': 'days'},\n",
       "  'followUp': {'name': 'follow-up', 'duration': 60, 'durationUnit': 'days'},\n",
       "  'treatments': [{'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'hydroxy',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'generatedStudyDesign': {'name': 'test study',\n",
       "  'description': 'this is the verbose description',\n",
       "  'type': {'term': 'crossover design',\n",
       "   'id': 'OBI:0500003',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/OBI_0500003'},\n",
       "  'elements': [{'id': '83978887-be4c-4cac-afb6-0faad3d07c86',\n",
       "    'name': 'treatment_0',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': 'e283a7c5-955f-4d3b-b99e-0053a2af5a61',\n",
       "    'name': 'washout',\n",
       "    'duration': 5,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': '696a2da4-329a-4b39-a681-e402afb4e89f',\n",
       "    'name': 'treatment_1',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'hydroxy',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '6bf4662d-852f-44ad-9b46-acff5291c47f',\n",
       "    'name': 'treatment_2',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 60,\n",
       "    'durationUnit': 'days'}],\n",
       "  'events': [],\n",
       "  'arms': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}]},\n",
       " 'selectedArms': [{'id': '#arm/0',\n",
       "   'name': 'Arm_0',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/2',\n",
       "   'name': 'Arm_2',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/4',\n",
       "   'name': 'Arm_4',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/3',\n",
       "   'name': 'Arm_3',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}],\n",
       " 'samplePlan': [{'sampleType': 'saliva',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, 1, 1, 1, 1, 3],\n",
       "    'Arm_2': [1, 1, 1, 1, 1, 3],\n",
       "    'Arm_4': [1, 1, 1, None, 1, 3],\n",
       "    'Arm_3': [1, 1, 1, 1, None, 3]}},\n",
       "  {'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_4': [False, False, False, False, False, True],\n",
       "    'Arm_3': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 1],\n",
       "    'Arm_2': [None, None, None, None, None, 1],\n",
       "    'Arm_4': [None, None, None, None, None, 1],\n",
       "    'Arm_3': [None, None, None, None, None, 1]}}],\n",
       " 'assayConfigs': [{'id': 0,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec.',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'nonpolar fraction'],\n",
       "       'values': ['polar fraction']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labelling', {'#replicates': {'value': 1}}],\n",
       "    ['labelled extract',\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': 'labelled extract type',\n",
       "      'characteristics_value': {'options': ['biotin', 'cy-5', 'cy-3'],\n",
       "       'values': ['biotin']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': ['Agilent QTQF 6510']},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': ['FIA', 'LC']},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': ['positive mode']}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, True, False, True, True],\n",
       "    'Arm_2': [False, False, True, False, True, True],\n",
       "    'Arm_4': [False, False, True, False, True, True],\n",
       "    'Arm_3': [False, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 1,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec. (Annotated)',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': None,\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labelling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labelled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labelled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'biotin',\n",
       "         'iri': None,\n",
       "         'source': None},\n",
       "        {'term': 'cy-5', 'iri': None, 'source': None},\n",
       "        {'term': 'cy-3', 'iri': None, 'source': None}],\n",
       "       'values': [{'term': 'cy-3', 'iri': None, 'source': None}]},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': []},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}],\n",
       "       'values': []}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': None},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'Metabolite Profiling - NMR',\n",
       "   'icon': 'fa fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nmr_spectroscopy',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Bruker AvanceII 1 GHz'], 'values': []},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': []},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': []}}],\n",
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}}],\n",
       " 'selectedAssayTypes': {'Metabolite Profiling - Mass. Spec.': True,\n",
       "  'Metabolite Profiling - Mass. Spec. (Annotated)': False,\n",
       "  'Metabolite Profiling - NMR': False}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"study-design-3-repeated-treatments-datascriptor.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design_from_config` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design_from_config(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 0.48 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 0.26 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output'))\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "with open(os.path.abspath(os.path.join('output','isa-investigation-2-arms-nmr-ms.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-08-21 13:04:46,572 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 270 paths!\n",
      "2020-08-21 13:04:46,586 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 270 paths\n",
      "2020-08-21 13:04:46,592 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 270 rows\n",
      "2020-08-21 13:04:46,836 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:47,070 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:47,956 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!\n",
      "2020-08-21 13:04:48,200 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:48,433 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:49,308 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!\n",
      "2020-08-21 13:04:49,554 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 4.52 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, os.path.abspath(os.path.join('output')))\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-08-21 13:04:53,304 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 270 paths!\n",
      "2020-08-21 13:04:53,314 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 270 paths\n",
      "2020-08-21 13:04:53,318 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 270 rows\n",
      "2020-08-21 13:04:53,543 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:53,773 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:54,635 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!\n",
      "2020-08-21 13:04:54,875 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:55,100 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n",
      "2020-08-21 13:04:55,956 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 160 paths!\n",
      "2020-08-21 13:04:56,189 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 40 paths!\n"
     ]
    }
   ],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 1 assay file (is this correct? check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_0_5_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_0_2_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_0_4_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_3_2_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_2_5_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_2_4_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_Arm_2_2_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 10 subjects in the each of the four selected arms (Arm_0, Arm_2, Arm_3, Arm_4) and 24 samples have been collected (3 blood samples per subject during the follow-up epoch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 70 samples in the Arm_0 arm (i.e. group)\n",
      "There are 70 samples in the Arm_2 arm (i.e. group)\n",
      "There are 60 samples in the Arm_3 arm (i.e. group)\n",
      "There are 70 samples in the Arm_4 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'Arm_0' in el)])\n",
    "count_arm2_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'Arm_2' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'Arm_3' in el)])\n",
    "count_arm4_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'Arm_4' in el)])\n",
    "print(\"There are {} samples in the Arm_0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the Arm_2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the Arm_3 arm (i.e. group)\".format(count_arm3_samples))\n",
    "print(\"There are {} samples in the Arm_4 arm (i.e. group)\".format(count_arm4_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Labeled Extract Name</th>\n",
       "      <th>Label</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[injection_mode]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>MS Assay Name</th>\n",
       "      <th>Performer.2</th>\n",
       "      <th>Raw Spectral Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_4-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_4-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_4-8</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_4-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_4-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_4-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_4-4</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_4-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_4-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_4-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_4-6</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_4-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_4-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_4-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_4-10</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_4-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.saliva.1</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_22-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_22-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_22-6</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_22-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.2</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_17-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_17-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_17-10</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_17-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_18-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_18-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_18-4</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_18-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_18-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_18-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_18-8</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_18-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_18-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_18-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>FIA</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_18-10</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_18-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_18-1</td>\n",
       "      <td>polar fraction</td>\n",
       "      <td>labelling</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>labelled extract_18-3</td>\n",
       "      <td>biotin</td>\n",
       "      <td>mass spectrometry</td>\n",
       "      <td>Agilent QTQF 6510</td>\n",
       "      <td>LC</td>\n",
       "      <td>positive mode</td>\n",
       "      <td>mass spectrometry_18-6</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_18-7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160 rows  16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Sample Name Protocol REF Performer  \\\n",
       "0     GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1   extraction  Ellipsis   \n",
       "1     GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1   extraction  Ellipsis   \n",
       "2     GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1   extraction  Ellipsis   \n",
       "3     GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.blood.1   extraction  Ellipsis   \n",
       "4    GRP-Arm_0.SBJ-001.CEL-CELL_Arm_0_5.SMP-.saliva.1   extraction  Ellipsis   \n",
       "..                                                ...          ...       ...   \n",
       "155  GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.2   extraction  Ellipsis   \n",
       "156  GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3   extraction  Ellipsis   \n",
       "157  GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3   extraction  Ellipsis   \n",
       "158  GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3   extraction  Ellipsis   \n",
       "159  GRP-Arm_0.SBJ-010.CEL-CELL_Arm_0_5.SMP-.saliva.3   extraction  Ellipsis   \n",
       "\n",
       "     Extract Name Characteristics[extract type] Protocol REF.1 Performer.1  \\\n",
       "0     extract_4-1                polar fraction      labelling    Ellipsis   \n",
       "1     extract_4-1                polar fraction      labelling    Ellipsis   \n",
       "2     extract_4-1                polar fraction      labelling    Ellipsis   \n",
       "3     extract_4-1                polar fraction      labelling    Ellipsis   \n",
       "4    extract_22-1                polar fraction      labelling    Ellipsis   \n",
       "..            ...                           ...            ...         ...   \n",
       "155  extract_17-1                polar fraction      labelling    Ellipsis   \n",
       "156  extract_18-1                polar fraction      labelling    Ellipsis   \n",
       "157  extract_18-1                polar fraction      labelling    Ellipsis   \n",
       "158  extract_18-1                polar fraction      labelling    Ellipsis   \n",
       "159  extract_18-1                polar fraction      labelling    Ellipsis   \n",
       "\n",
       "      Labeled Extract Name   Label     Protocol REF.2  \\\n",
       "0     labelled extract_4-3  biotin  mass spectrometry   \n",
       "1     labelled extract_4-3  biotin  mass spectrometry   \n",
       "2     labelled extract_4-3  biotin  mass spectrometry   \n",
       "3     labelled extract_4-3  biotin  mass spectrometry   \n",
       "4    labelled extract_22-3  biotin  mass spectrometry   \n",
       "..                     ...     ...                ...   \n",
       "155  labelled extract_17-3  biotin  mass spectrometry   \n",
       "156  labelled extract_18-3  biotin  mass spectrometry   \n",
       "157  labelled extract_18-3  biotin  mass spectrometry   \n",
       "158  labelled extract_18-3  biotin  mass spectrometry   \n",
       "159  labelled extract_18-3  biotin  mass spectrometry   \n",
       "\n",
       "    Parameter Value[instrument] Parameter Value[injection_mode]  \\\n",
       "0             Agilent QTQF 6510                             FIA   \n",
       "1             Agilent QTQF 6510                              LC   \n",
       "2             Agilent QTQF 6510                              LC   \n",
       "3             Agilent QTQF 6510                             FIA   \n",
       "4             Agilent QTQF 6510                              LC   \n",
       "..                          ...                             ...   \n",
       "155           Agilent QTQF 6510                             FIA   \n",
       "156           Agilent QTQF 6510                              LC   \n",
       "157           Agilent QTQF 6510                             FIA   \n",
       "158           Agilent QTQF 6510                             FIA   \n",
       "159           Agilent QTQF 6510                              LC   \n",
       "\n",
       "    Parameter Value[acquisition_mode]            MS Assay Name Performer.2  \\\n",
       "0                       positive mode    mass spectrometry_4-8    Ellipsis   \n",
       "1                       positive mode    mass spectrometry_4-4    Ellipsis   \n",
       "2                       positive mode    mass spectrometry_4-6    Ellipsis   \n",
       "3                       positive mode   mass spectrometry_4-10    Ellipsis   \n",
       "4                       positive mode   mass spectrometry_22-6    Ellipsis   \n",
       "..                                ...                      ...         ...   \n",
       "155                     positive mode  mass spectrometry_17-10    Ellipsis   \n",
       "156                     positive mode   mass spectrometry_18-4    Ellipsis   \n",
       "157                     positive mode   mass spectrometry_18-8    Ellipsis   \n",
       "158                     positive mode  mass spectrometry_18-10    Ellipsis   \n",
       "159                     positive mode   mass spectrometry_18-6    Ellipsis   \n",
       "\n",
       "           Raw Spectral Data File  \n",
       "0      raw spectral data file_4-9  \n",
       "1      raw spectral data file_4-5  \n",
       "2      raw spectral data file_4-7  \n",
       "3     raw spectral data file_4-11  \n",
       "4     raw spectral data file_22-7  \n",
       "..                            ...  \n",
       "155  raw spectral data file_17-11  \n",
       "156   raw spectral data file_18-5  \n",
       "157   raw spectral data file_18-9  \n",
       "158  raw spectral data file_18-11  \n",
       "159   raw spectral data file_18-7  \n",
       "\n",
       "[160 rows x 16 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_CELL_Arm_0_5_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
