{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "# Observational study on Human students\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "- a 4-arm study design. Each arm has 10 subjects\n",
    "- subjects are human, male and female, of three age groups in years (6-10, 10-15, 15-18) \n",
    "- two phases: the observational phase and follow-up\n",
    "- one sample type colllected: blood\n",
    "- two assay types: \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create.model import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import model\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator with all the specs defined above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'subjectGroups': {'selected': [{'name': 'SubjectGroup_0',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '15-18',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]},\n",
       "   {'name': 'SubjectGroup_1',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '10-15',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]},\n",
       "   {'name': 'SubjectGroup_2',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '6-10',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]},\n",
       "   {'name': 'SubjectGroup_3',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '15-18',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]},\n",
       "   {'name': 'SubjectGroup_4',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '10-15',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]},\n",
       "   {'name': 'SubjectGroup_5',\n",
       "    'type': 'Homo sapiens',\n",
       "    'characteristics': [{'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'name': 'age group',\n",
       "      'value': '6-10',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}]}],\n",
       "  'unselected': []},\n",
       " 'treatmentPlan': {'observationPeriod': {'name': 'observation period',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'},\n",
       "  'screen': {'selected': False,\n",
       "   'name': 'screen',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'runIn': {'selected': False,\n",
       "   'name': 'run-in',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'washout': {'selected': False,\n",
       "   'name': 'washout',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'followUp': {'selected': True,\n",
       "   'name': 'follow-up',\n",
       "   'duration': 180,\n",
       "   'durationUnit': 'days'},\n",
       "  'elementParams': {'agents': [],\n",
       "   'intensities': [],\n",
       "   'durations': [],\n",
       "   'interventionType': None,\n",
       "   'intensityUnit': '',\n",
       "   'durationUnit': ''},\n",
       "  'treatments': [{'_id': '60648e78f3eccab7f53da5a5',\n",
       "    'interventionType': None,\n",
       "    'agent': None,\n",
       "    'intensity': None,\n",
       "    'intensityUnit': '',\n",
       "    'duration': None,\n",
       "    'durationUnit': ''},\n",
       "   {'_id': '60648e78f3eccab7f53da5a6',\n",
       "    'interventionType': None,\n",
       "    'agent': None,\n",
       "    'intensity': None,\n",
       "    'intensityUnit': '',\n",
       "    'duration': None,\n",
       "    'durationUnit': ''}]},\n",
       " 'arms': {'unselected': [],\n",
       "  'selected': [{'_id': '60648e78f3eccab7f53da5a9',\n",
       "    'id': '0',\n",
       "    'name': 'Arm_0',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5aa',\n",
       "      'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5ab',\n",
       "      'name': 'age group',\n",
       "      'value': '15-18',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5ac'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5ad'}]},\n",
       "   {'_id': '60648e78f3eccab7f53da5ae',\n",
       "    'id': '1',\n",
       "    'name': 'Arm_1',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5af',\n",
       "      'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5b0',\n",
       "      'name': 'age group',\n",
       "      'value': '10-15',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5b1'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5b2'}]},\n",
       "   {'_id': '60648e78f3eccab7f53da5b3',\n",
       "    'id': '2',\n",
       "    'name': 'Arm_2',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5b4',\n",
       "      'name': 'sex',\n",
       "      'value': 'M',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5b5',\n",
       "      'name': 'age group',\n",
       "      'value': '6-10',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5b6'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5b7'}]},\n",
       "   {'_id': '60648e78f3eccab7f53da5b8',\n",
       "    'id': '3',\n",
       "    'name': 'Arm_3',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5b9',\n",
       "      'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5ba',\n",
       "      'name': 'age group',\n",
       "      'value': '15-18',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5bb'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5bc'}]},\n",
       "   {'_id': '60648e78f3eccab7f53da5bd',\n",
       "    'id': '4',\n",
       "    'name': 'Arm_4',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5be',\n",
       "      'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5bf',\n",
       "      'name': 'age group',\n",
       "      'value': '10-15',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5c0'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5c1'}]},\n",
       "   {'_id': '60648e78f3eccab7f53da5c2',\n",
       "    'id': '5',\n",
       "    'name': 'Arm_5',\n",
       "    'observationalFactors': [{'_id': '60648e78f3eccab7f53da5c3',\n",
       "      'name': 'sex',\n",
       "      'value': 'F',\n",
       "      'unit': None,\n",
       "      'isQuantitative': False},\n",
       "     {'_id': '60648e78f3eccab7f53da5c4',\n",
       "      'name': 'age group',\n",
       "      'value': '6-10',\n",
       "      'unit': 'years',\n",
       "      'isQuantitative': True}],\n",
       "    'size': 100,\n",
       "    'epochs': [{'elements': ['5de2c40f-9320-422f-8fc8-b1bb80ba5adf'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5c5'},\n",
       "     {'elements': ['007d1ff8-5e16-401c-b847-5eaf6ba61436'],\n",
       "      'events': [],\n",
       "      '_id': '60648e78f3eccab7f53da5c6'}]}]},\n",
       " 'observationalFactors': [{'name': 'sex',\n",
       "   'values': ['M', 'F'],\n",
       "   'isQuantitative': False,\n",
       "   'unit': None},\n",
       "  {'name': 'age group',\n",
       "   'values': ['6-10', '10-15', '15-18'],\n",
       "   'isQuantitative': True,\n",
       "   'unit': 'years'}],\n",
       " 'samplePlan': [{'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [True, True],\n",
       "    'Arm_1': [True, True],\n",
       "    'Arm_2': [True, True],\n",
       "    'Arm_3': [True, True],\n",
       "    'Arm_4': [True, True],\n",
       "    'Arm_5': [True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [3, 6],\n",
       "    'Arm_1': [3, 6],\n",
       "    'Arm_2': [3, 6],\n",
       "    'Arm_3': [3, 6],\n",
       "    'Arm_4': [3, 6],\n",
       "    'Arm_5': [3, 6]}}],\n",
       " 'assayPlan': [{'id': 8,\n",
       "   'name': 'genome sequencing by nucleotide sequencing',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'purple',\n",
       "   'measurement_type': 'genome sequencing',\n",
       "   'technology_type': 'nucleic acid sequencing',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['gDNA', 'DNA'], 'values': ['DNA']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['library_preparation',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['MinION',\n",
       "        'GridION',\n",
       "        'PromethION',\n",
       "        'Illumina HiSeq 4000',\n",
       "        'Illumina NovaSeq 6000',\n",
       "        'AB 3730xL Genetic Analyzer'],\n",
       "       'values': ['Illumina NovaSeq 6000']},\n",
       "      'library_orientation': {'options': ['single', 'paired'],\n",
       "       'values': ['single', 'paired']},\n",
       "      'library_strategy': {'options': ['AMPLICON',\n",
       "        'CLONE',\n",
       "        'WGA',\n",
       "        'WGS',\n",
       "        'WXS',\n",
       "        'OTHER'],\n",
       "       'values': ['WGA']}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [True, False],\n",
       "    'Arm_1': [True, False],\n",
       "    'Arm_2': [True, False],\n",
       "    'Arm_3': [True, False],\n",
       "    'Arm_4': [True, False],\n",
       "    'Arm_5': [True, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['blood'], ['blood']],\n",
       "    'Arm_1': [['blood'], ['blood']],\n",
       "    'Arm_2': [['blood'], ['blood']],\n",
       "    'Arm_3': [['blood'], ['blood']],\n",
       "    'Arm_4': [['blood'], ['blood']],\n",
       "    'Arm_5': [['blood'], ['blood']]}},\n",
       "  {'id': 10,\n",
       "   'name': 'clinical chemistry by marker panel',\n",
       "   'icon': 'fa fa-flask',\n",
       "   'color': 'yellow',\n",
       "   'measurement_type': 'clinical chemistry',\n",
       "   'technology_type': 'marker panel',\n",
       "   'workflow': [['sample preparation', {'#replicates': {'value': 1}}],\n",
       "    ['raw_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, True],\n",
       "    'Arm_1': [False, True],\n",
       "    'Arm_2': [False, True],\n",
       "    'Arm_3': [False, True],\n",
       "    'Arm_4': [False, True],\n",
       "    'Arm_5': [False, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['blood'], ['blood']],\n",
       "    'Arm_1': [['blood'], ['blood']],\n",
       "    'Arm_2': [['blood'], ['blood']],\n",
       "    'Arm_3': [['blood'], ['blood']],\n",
       "    'Arm_4': [['blood'], ['blood']],\n",
       "    'Arm_5': [['blood'], ['blood']]}}],\n",
       " 'subjectType': 'Homo sapiens',\n",
       " 'subjectSize': 100,\n",
       " 'designType': {'term': 'observational design',\n",
       "  'id': 'NCIT:C147138',\n",
       "  'iri': 'http://purl.obolibrary.org/obo/NCIT_C147138',\n",
       "  'label': 'Study subjects do not receive treatments',\n",
       "  'value': 'observational'},\n",
       " 'elements': [{'_id': '60648e78f3eccab7f53da5a7',\n",
       "   'id': '5de2c40f-9320-422f-8fc8-b1bb80ba5adf',\n",
       "   'name': 'observation period',\n",
       "   'duration': 90,\n",
       "   'durationUnit': 'days'},\n",
       "  {'_id': '60648e78f3eccab7f53da5a8',\n",
       "   'id': '007d1ff8-5e16-401c-b847-5eaf6ba61436',\n",
       "   'name': 'follow-up',\n",
       "   'duration': 180,\n",
       "   'durationUnit': 'days'}],\n",
       " 'events': [],\n",
       " '_id': '60648e78f3eccab7f53da5a4',\n",
       " 'name': \"Student's study v2\",\n",
       " 'description': 'In an observational study, the investigator simply records observations and analyzes data, without administration of an intervention or alteration of the care people receive. These studies may focus on risk factors, natural history, variations in disease progression or disease treatment without delivering an intervention. They often assess specific health characteristics of the enrolled human subjects by collecting medical/dental history, exposure, or clinical data; obtaining biospecimens (e.g., for biomarker or genomic analyses); or obtaining photographic, radiographic or other images from research subjects.',\n",
       " 'identifier': 's_dbd4a1e5-35d8-4c6c-aa8b-60c86f3388a7'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"observational-design-students.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design()` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study from the design took 3.56 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study from the design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(identifier='inv01', studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 2.21 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output', 'observational-design-students'))\n",
    "os.makedirs(directory, exist_ok=True)\n",
    "with open(os.path.abspath(os.path.join(directory, 'isa-investigation-observational-design-students.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Tab serialisation of the ISA investigation took 123.83 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, directory)\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 2 assay files (one for MS and one for NMR). Let's check the names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT10_clinical-chemistry_marker-panel.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Count of subjects and samples\n",
    "\n",
    "We have 10 subjects in the each of the six arms for a total of 60 subjects. 5 blood samples per subject are collected (1 in treatment 1 phase, 1 in treatment, and 3 in the follow-up phase) for a total of 300 blood samples. These will undergo the NMR assay. We have 4 saliva samples per subject (1 during screen and 3 during follow-up) for a total of 240 saliva samples. These will undergo the \"mass spcetrometry\" assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 900 samples in the GRP0 arm (i.e. group)\n",
      "There are 900 samples in the GRP1 arm (i.e. group)\n",
      "There are 900 samples in the GRP2 arm (i.e. group)\n",
      "There are 900 samples in the GRP3 arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_arm0_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP0' in el)])\n",
    "count_arm1_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP1' in el)])\n",
    "count_arm3_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'GRP3' in el)])\n",
    "print(\"There are {} samples in the GRP0 arm (i.e. group)\".format(count_arm0_samples))\n",
    "print(\"There are {} samples in the GRP1 arm (i.e. group)\".format(count_arm1_samples))\n",
    "print(\"There are {} samples in the GRP2 arm (i.e. group)\".format(count_arm2_samples))\n",
    "print(\"There are {} samples in the GRP3 arm (i.e. group)\".format(count_arm3_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Characteristics[Study Subject]</th>\n",
       "      <th>Characteristics[age group]</th>\n",
       "      <th>Characteristics[sex]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Parameter Value[Sampling order]</th>\n",
       "      <th>Parameter Value[Study cell]</th>\n",
       "      <th>Date</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Comment[Treatment?]</th>\n",
       "      <th>Factor Value[Sequence Order]</th>\n",
       "      <th>Factor Value[DURATION]</th>\n",
       "      <th>Unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>15-18</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>769</td>\n",
       "      <td>A0E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-1</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>15-18</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>237</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-3</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>15-18</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>235</td>\n",
       "      <td>A0E0</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-1</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>15-18</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>772</td>\n",
       "      <td>A0E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-4</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ001</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>15-18</td>\n",
       "      <td>M</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>773</td>\n",
       "      <td>A0E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-5</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5395</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>6-10</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>4891</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-1</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5396</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>6-10</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>4896</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-6</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5397</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>6-10</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>4893</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-3</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5398</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>6-10</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>4546</td>\n",
       "      <td>A5E0</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-1</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5399</th>\n",
       "      <td>GRP5_SBJ100</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>6-10</td>\n",
       "      <td>F</td>\n",
       "      <td>sample collection</td>\n",
       "      <td>4892</td>\n",
       "      <td>A5E1</td>\n",
       "      <td>2021-03-31</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-2</td>\n",
       "      <td>blood</td>\n",
       "      <td>NO</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>days</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5400 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Source Name Characteristics[Study Subject] Characteristics[age group]  \\\n",
       "0     GRP0_SBJ001                   Homo sapiens                      15-18   \n",
       "1     GRP0_SBJ001                   Homo sapiens                      15-18   \n",
       "2     GRP0_SBJ001                   Homo sapiens                      15-18   \n",
       "3     GRP0_SBJ001                   Homo sapiens                      15-18   \n",
       "4     GRP0_SBJ001                   Homo sapiens                      15-18   \n",
       "...           ...                            ...                        ...   \n",
       "5395  GRP5_SBJ100                   Homo sapiens                       6-10   \n",
       "5396  GRP5_SBJ100                   Homo sapiens                       6-10   \n",
       "5397  GRP5_SBJ100                   Homo sapiens                       6-10   \n",
       "5398  GRP5_SBJ100                   Homo sapiens                       6-10   \n",
       "5399  GRP5_SBJ100                   Homo sapiens                       6-10   \n",
       "\n",
       "     Characteristics[sex]       Protocol REF Parameter Value[Sampling order]  \\\n",
       "0                       M  sample collection                             769   \n",
       "1                       M  sample collection                             237   \n",
       "2                       M  sample collection                             235   \n",
       "3                       M  sample collection                             772   \n",
       "4                       M  sample collection                             773   \n",
       "...                   ...                ...                             ...   \n",
       "5395                    F  sample collection                            4891   \n",
       "5396                    F  sample collection                            4896   \n",
       "5397                    F  sample collection                            4893   \n",
       "5398                    F  sample collection                            4546   \n",
       "5399                    F  sample collection                            4892   \n",
       "\n",
       "     Parameter Value[Study cell]        Date Performer  \\\n",
       "0                           A0E1  2021-03-31   Unknown   \n",
       "1                           A0E0  2021-03-31   Unknown   \n",
       "2                           A0E0  2021-03-31   Unknown   \n",
       "3                           A0E1  2021-03-31   Unknown   \n",
       "4                           A0E1  2021-03-31   Unknown   \n",
       "...                          ...         ...       ...   \n",
       "5395                        A5E1  2021-03-31   Unknown   \n",
       "5396                        A5E1  2021-03-31   Unknown   \n",
       "5397                        A5E1  2021-03-31   Unknown   \n",
       "5398                        A5E0  2021-03-31   Unknown   \n",
       "5399                        A5E1  2021-03-31   Unknown   \n",
       "\n",
       "                       Sample Name Characteristics[organism part]  \\\n",
       "0     GRP0_SBJ001_A0E1_SMP-blood-1                          blood   \n",
       "1     GRP0_SBJ001_A0E0_SMP-blood-3                          blood   \n",
       "2     GRP0_SBJ001_A0E0_SMP-blood-1                          blood   \n",
       "3     GRP0_SBJ001_A0E1_SMP-blood-4                          blood   \n",
       "4     GRP0_SBJ001_A0E1_SMP-blood-5                          blood   \n",
       "...                            ...                            ...   \n",
       "5395  GRP5_SBJ100_A5E1_SMP-blood-1                          blood   \n",
       "5396  GRP5_SBJ100_A5E1_SMP-blood-6                          blood   \n",
       "5397  GRP5_SBJ100_A5E1_SMP-blood-3                          blood   \n",
       "5398  GRP5_SBJ100_A5E0_SMP-blood-1                          blood   \n",
       "5399  GRP5_SBJ100_A5E1_SMP-blood-2                          blood   \n",
       "\n",
       "     Comment[Treatment?] Factor Value[Sequence Order] Factor Value[DURATION]  \\\n",
       "0                     NO                            1                    180   \n",
       "1                     NO                            0                     90   \n",
       "2                     NO                            0                     90   \n",
       "3                     NO                            1                    180   \n",
       "4                     NO                            1                    180   \n",
       "...                  ...                          ...                    ...   \n",
       "5395                  NO                            1                    180   \n",
       "5396                  NO                            1                    180   \n",
       "5397                  NO                            1                    180   \n",
       "5398                  NO                            0                     90   \n",
       "5399                  NO                            1                    180   \n",
       "\n",
       "      Unit  \n",
       "0     days  \n",
       "1     days  \n",
       "2     days  \n",
       "3     days  \n",
       "4     days  \n",
       "...    ...  \n",
       "5395  days  \n",
       "5396  days  \n",
       "5397  days  \n",
       "5398  days  \n",
       "5399  days  \n",
       "\n",
       "[5400 rows x 15 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[Treatment?]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S469-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S470-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S471-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-4</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S472-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ001_A0E1_SMP-blood-5</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S473-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3595</th>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S3092-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3596</th>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S3093-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3597</th>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-4</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S3094-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3598</th>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-5</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S3095-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3599</th>\n",
       "      <td>GRP5_SBJ100_A5E1_SMP-blood-6</td>\n",
       "      <td>NO</td>\n",
       "      <td>sample preparation</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT10-S3096-raw_data_file-R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3600 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sample Name Comment[Treatment?]        Protocol REF  \\\n",
       "0     GRP0_SBJ001_A0E1_SMP-blood-1                  NO  sample preparation   \n",
       "1     GRP0_SBJ001_A0E1_SMP-blood-2                  NO  sample preparation   \n",
       "2     GRP0_SBJ001_A0E1_SMP-blood-3                  NO  sample preparation   \n",
       "3     GRP0_SBJ001_A0E1_SMP-blood-4                  NO  sample preparation   \n",
       "4     GRP0_SBJ001_A0E1_SMP-blood-5                  NO  sample preparation   \n",
       "...                            ...                 ...                 ...   \n",
       "3595  GRP5_SBJ100_A5E1_SMP-blood-2                  NO  sample preparation   \n",
       "3596  GRP5_SBJ100_A5E1_SMP-blood-3                  NO  sample preparation   \n",
       "3597  GRP5_SBJ100_A5E1_SMP-blood-4                  NO  sample preparation   \n",
       "3598  GRP5_SBJ100_A5E1_SMP-blood-5                  NO  sample preparation   \n",
       "3599  GRP5_SBJ100_A5E1_SMP-blood-6                  NO  sample preparation   \n",
       "\n",
       "     Performer                Raw Data File  \n",
       "0      Unknown   AT10-S469-raw_data_file-R1  \n",
       "1      Unknown   AT10-S470-raw_data_file-R1  \n",
       "2      Unknown   AT10-S471-raw_data_file-R1  \n",
       "3      Unknown   AT10-S472-raw_data_file-R1  \n",
       "4      Unknown   AT10-S473-raw_data_file-R1  \n",
       "...        ...                          ...  \n",
       "3595   Unknown  AT10-S3092-raw_data_file-R1  \n",
       "3596   Unknown  AT10-S3093-raw_data_file-R1  \n",
       "3597   Unknown  AT10-S3094-raw_data_file-R1  \n",
       "3598   Unknown  AT10-S3095-raw_data_file-R1  \n",
       "3599   Unknown  AT10-S3096-raw_data_file-R1  \n",
       "\n",
       "[3600 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT10_clinical-chemistry_marker-panel.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Comment[Treatment?]</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[library_orientation]</th>\n",
       "      <th>Parameter Value[library_strategy]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Raw Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-raw_data_file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S235-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP0_SBJ001_A0E0_SMP-blood-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S236-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S236-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7195</th>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-2</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1547-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1547-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7196</th>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-raw_data_file-R2.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7197</th>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-raw_data_file-R4.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198</th>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>single</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-raw_data_file-R3.raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7199</th>\n",
       "      <td>GRP5_SBJ100_A5E0_SMP-blood-3</td>\n",
       "      <td>NO</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-Extract-R1</td>\n",
       "      <td>DNA</td>\n",
       "      <td>library_preparation</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>paired</td>\n",
       "      <td>WGA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AT8-S1548-raw_data_file-R1.raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7200 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sample Name Comment[Treatment?] Protocol REF Performer  \\\n",
       "0     GRP0_SBJ001_A0E0_SMP-blood-1                  NO   extraction   Unknown   \n",
       "1     GRP0_SBJ001_A0E0_SMP-blood-1                  NO   extraction   Unknown   \n",
       "2     GRP0_SBJ001_A0E0_SMP-blood-1                  NO   extraction   Unknown   \n",
       "3     GRP0_SBJ001_A0E0_SMP-blood-1                  NO   extraction   Unknown   \n",
       "4     GRP0_SBJ001_A0E0_SMP-blood-2                  NO   extraction   Unknown   \n",
       "...                            ...                 ...          ...       ...   \n",
       "7195  GRP5_SBJ100_A5E0_SMP-blood-2                  NO   extraction   Unknown   \n",
       "7196  GRP5_SBJ100_A5E0_SMP-blood-3                  NO   extraction   Unknown   \n",
       "7197  GRP5_SBJ100_A5E0_SMP-blood-3                  NO   extraction   Unknown   \n",
       "7198  GRP5_SBJ100_A5E0_SMP-blood-3                  NO   extraction   Unknown   \n",
       "7199  GRP5_SBJ100_A5E0_SMP-blood-3                  NO   extraction   Unknown   \n",
       "\n",
       "              Extract Name Characteristics[extract type]       Protocol REF.1  \\\n",
       "0      AT8-S235-Extract-R1                           DNA  library_preparation   \n",
       "1      AT8-S235-Extract-R1                           DNA  library_preparation   \n",
       "2      AT8-S235-Extract-R1                           DNA  library_preparation   \n",
       "3      AT8-S235-Extract-R1                           DNA  library_preparation   \n",
       "4      AT8-S236-Extract-R1                           DNA  library_preparation   \n",
       "...                    ...                           ...                  ...   \n",
       "7195  AT8-S1547-Extract-R1                           DNA  library_preparation   \n",
       "7196  AT8-S1548-Extract-R1                           DNA  library_preparation   \n",
       "7197  AT8-S1548-Extract-R1                           DNA  library_preparation   \n",
       "7198  AT8-S1548-Extract-R1                           DNA  library_preparation   \n",
       "7199  AT8-S1548-Extract-R1                           DNA  library_preparation   \n",
       "\n",
       "     Parameter Value[instrument] Parameter Value[library_orientation]  \\\n",
       "0          Illumina NovaSeq 6000                               single   \n",
       "1          Illumina NovaSeq 6000                               paired   \n",
       "2          Illumina NovaSeq 6000                               paired   \n",
       "3          Illumina NovaSeq 6000                               single   \n",
       "4          Illumina NovaSeq 6000                               single   \n",
       "...                          ...                                  ...   \n",
       "7195       Illumina NovaSeq 6000                               single   \n",
       "7196       Illumina NovaSeq 6000                               paired   \n",
       "7197       Illumina NovaSeq 6000                               single   \n",
       "7198       Illumina NovaSeq 6000                               single   \n",
       "7199       Illumina NovaSeq 6000                               paired   \n",
       "\n",
       "     Parameter Value[library_strategy] Performer.1  \\\n",
       "0                                  WGA     Unknown   \n",
       "1                                  WGA     Unknown   \n",
       "2                                  WGA     Unknown   \n",
       "3                                  WGA     Unknown   \n",
       "4                                  WGA     Unknown   \n",
       "...                                ...         ...   \n",
       "7195                               WGA     Unknown   \n",
       "7196                               WGA     Unknown   \n",
       "7197                               WGA     Unknown   \n",
       "7198                               WGA     Unknown   \n",
       "7199                               WGA     Unknown   \n",
       "\n",
       "                       Raw Data File  \n",
       "0      AT8-S235-raw_data_file-R3.raw  \n",
       "1      AT8-S235-raw_data_file-R1.raw  \n",
       "2      AT8-S235-raw_data_file-R2.raw  \n",
       "3      AT8-S235-raw_data_file-R4.raw  \n",
       "4      AT8-S236-raw_data_file-R3.raw  \n",
       "...                              ...  \n",
       "7195  AT8-S1547-raw_data_file-R4.raw  \n",
       "7196  AT8-S1548-raw_data_file-R2.raw  \n",
       "7197  AT8-S1548-raw_data_file-R4.raw  \n",
       "7198  AT8-S1548-raw_data_file-R3.raw  \n",
       "7199  AT8-S1548-raw_data_file-R1.raw  \n",
       "\n",
       "[7200 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Clinical Chemistry Marker Panel\n",
    "\n",
    "For the mass. spec. assay table, we have 160 (derma) samples, 160 extracts (1 per  sample, \"polar\" fraction), 320 labeled extracts (2 per extract, as \"#replicates\" is  2) and 640 mass spec protocols + 640 output files (2 per labeled extract as we do 1 technical replicate with 2 protocol parameter combinations `[\"Agilent QTQF 6510\", \"FIA\", \"positive mode\"]` and `[\"Agilent QTQF 6510\", \"LC\", \"positive mode\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                             1800\n",
       "Comment[Treatment?]                        1\n",
       "Protocol REF                               1\n",
       "Performer                                  1\n",
       "Extract Name                            1800\n",
       "Characteristics[extract type]              1\n",
       "Protocol REF.1                             1\n",
       "Parameter Value[instrument]                1\n",
       "Parameter Value[library_orientation]       2\n",
       "Parameter Value[library_strategy]          1\n",
       "Performer.1                                1\n",
       "Raw Data File                           7200\n",
       "dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Genome Sequencing assay table\n",
    "\n",
    "For the NMR assay table, we have 360 (blood) samples, 720 extracts (2 per  sample, a single replicate of the \"supernatant\" and \"pellet\" fractions) and 2880 NMR protocols + 4800 output files (4 per extract as we do 2 technical replicates with 2 protocol parameter combinations `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"CPGM\"]` and `[\"Bruker Avance II 1 GHz\", \"1D 1H NMR\", \"TOCSY\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sample Name                             1800\n",
       "Comment[Treatment?]                        1\n",
       "Protocol REF                               1\n",
       "Performer                                  1\n",
       "Extract Name                            1800\n",
       "Characteristics[extract type]              1\n",
       "Protocol REF.1                             1\n",
       "Parameter Value[instrument]                1\n",
       "Parameter Value[library_orientation]       2\n",
       "Parameter Value[library_strategy]          1\n",
       "Performer.1                                1\n",
       "Raw Data File                           7200\n",
       "dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_AT8_genome-sequencing_nucleic-acid-sequencing.txt'].nunique(axis=0, dropna=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
