{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create ISA-API Investigation from Datascriptor Study Design configuration\n",
    "\n",
    "In this notebook I will show you how you can use a study design configuration is JSON format as produce by datascriptor (https://gitlab.com/datascriptor/datascriptor) to generate a single-study ISA investigation and how you can then serialise it in JSON and tabular (i.e. CSV) format.\n",
    "\n",
    "Or study design configuration consists of:\n",
    "[missing]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Setup\n",
    "\n",
    "Let's import all the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'measurement type': 'metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': 'Acquisition Parameter File', 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'targeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'untargeted metabolite profiling', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'isotopomer distribution analysis', 'technology type': 'NMR spectroscopy', 'protocol type': 'NMR spectroscopy', 'parameter-like file': None, 'node name': 'NMR Assay Name', 'key in json_configuration': 'nmr_raw_date_file', 'raw data file': 'Free Induction Decay Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'targeted metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'untargeted metabolite profiling', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': ['Derived Spectral Data File', 'Metabolite Assignment File']}, {'measurement type': 'isotologue distribution analysis', 'technology type': 'mass spectrometry', 'protocol type': 'mass spectrometry', 'parameter-like file': None, 'node name': 'MS Assay Name', 'key in json_configuration': 'ms_raw_date_file', 'raw data file': 'Raw Spectral Data File', 'derived data file': 'Derived Spectral Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'genotype profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'epigenome profling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'exome profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'DNA methylation profiling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'copy number variation profling', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'transcription factor binding site identification', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'protein-DNA binding site identification', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'SNP analysis', 'technology type': 'DNA microarray', 'protocol type': 'nucleic acid hybridization', 'parameter-like file': 'Array Design REF', 'node name': 'Hybridization Assay Name', 'key in json_configuration': 'microarray_raw_data_file', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Array Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'transcription factor binding site identification', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'protein-DNA binding site identification', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'DNA methylation profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'histone modification profiling', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'genome sequencing', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'metagenome sequencing', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'environmental gene survey', 'technology type': 'nucleic acid sequencing', 'protocol type': 'nucleic acid sequencing', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell sorting', 'technology type': 'flow cytometry', 'protocol type': 'flow cytometry', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell counting', 'technology type': 'flow cytometry', 'protocol type': 'flow cytometry', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'cell migration assay', 'technology type': 'microscopy imaging', 'protocol type': 'migration assay', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'phenotyping', 'technology type': 'imaging', 'protocol type': 'migration assay', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': 'transcription profiling', 'technology type': 'RT-pcr', 'protocol type': 'RT-pcr', 'parameter-like file': None, 'node name': 'Assay Name', 'key in json_configuration': 'raw_data_file (default)', 'raw data file': 'Raw Data File', 'derived data file': 'Derived Data File'}, {'measurement type': None, 'technology type': None, 'protocol type': None, 'parameter-like file': None, 'node name': None, 'key in json_configuration': 'my_favourite_institution_tag (no white space allowed)', 'raw data file': None, 'derived data file': None}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/create/models.py:973: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  _kwargs = {key: val for key, val in kwargs.items() if key is not 'qc_sample_type'}\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "import os\n",
    "import json\n",
    "\n",
    "## ISA-API related imports\n",
    "from isatools.model import Investigation, Study\n",
    "\n",
    "## ISA-API create mode related imports\n",
    "from isatools.create.models import StudyDesign\n",
    "from isatools.create.connectors import generate_study_design_from_config\n",
    "\n",
    "# serializer from ISA Investigation to JSON\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "\n",
    "# ISA-Tab serialisation\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## ISA-API create mode related imports\n",
    "from isatools.create import models\n",
    "from isatools import isajson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load the Study Design JSON configuration\n",
    "\n",
    "First of all we load the study design configurator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'treatmentPlan': {'screen': {'name': 'screen',\n",
       "   'duration': None,\n",
       "   'durationUnit': ''},\n",
       "  'runIn': {'name': 'run-in', 'duration': None, 'durationUnit': ''},\n",
       "  'washout': {'name': 'washout', 'duration': '5', 'durationUnit': 'days'},\n",
       "  'followUp': {'name': 'follow-up', 'duration': 60, 'durationUnit': 'days'},\n",
       "  'treatments': [{'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'hydroxy',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'},\n",
       "   {'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days'}],\n",
       "  'elementParams': {'agents': [],\n",
       "   'intensities': [],\n",
       "   'intensityUnit': '',\n",
       "   'durations': [],\n",
       "   'durationUnit': ''}},\n",
       " 'generatedStudyDesign': {'name': 'test study',\n",
       "  'description': 'this is the verbose description',\n",
       "  'type': {'term': 'crossover design',\n",
       "   'id': 'OBI:0500003',\n",
       "   'iri': 'http://purl.obolibrary.org/obo/OBI_0500003'},\n",
       "  'elements': [{'id': '83978887-be4c-4cac-afb6-0faad3d07c86',\n",
       "    'name': 'treatment_0',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'ibuprofen',\n",
       "    'intensity': 8,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': 'e283a7c5-955f-4d3b-b99e-0053a2af5a61',\n",
       "    'name': 'washout',\n",
       "    'duration': 5,\n",
       "    'durationUnit': 'days'},\n",
       "   {'id': '696a2da4-329a-4b39-a681-e402afb4e89f',\n",
       "    'name': 'treatment_1',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'hydroxy',\n",
       "    'intensity': 5,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '6bf4662d-852f-44ad-9b46-acff5291c47f',\n",
       "    'name': 'treatment_2',\n",
       "    'duration': 10,\n",
       "    'durationUnit': 'days',\n",
       "    'agent': 'paracetamol',\n",
       "    'intensity': 10,\n",
       "    'intensityUnit': 'mg'},\n",
       "   {'id': '28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7',\n",
       "    'name': 'follow-up',\n",
       "    'duration': 60,\n",
       "    'durationUnit': 'days'}],\n",
       "  'events': [],\n",
       "  'arms': [{'id': '#arm/0',\n",
       "    'name': 'Arm_0',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/1',\n",
       "    'name': 'Arm_1',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/2',\n",
       "    'name': 'Arm_2',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/3',\n",
       "    'name': 'Arm_3',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/4',\n",
       "    'name': 'Arm_4',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "   {'id': '#arm/5',\n",
       "    'name': 'Arm_5',\n",
       "    'subjectType': 'human',\n",
       "    'size': 10,\n",
       "    'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "      'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "     {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "     {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "     {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}]},\n",
       " 'selectedArms': [{'id': '#arm/0',\n",
       "   'name': 'Arm_0',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/2',\n",
       "   'name': 'Arm_2',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/4',\n",
       "   'name': 'Arm_4',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]},\n",
       "  {'id': '#arm/3',\n",
       "   'name': 'Arm_3',\n",
       "   'subjectType': 'human',\n",
       "   'size': 10,\n",
       "   'epochs': [{'elements': ['696a2da4-329a-4b39-a681-e402afb4e89f'],\n",
       "     'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['6bf4662d-852f-44ad-9b46-acff5291c47f'], 'events': []},\n",
       "    {'elements': ['e283a7c5-955f-4d3b-b99e-0053a2af5a61'], 'events': []},\n",
       "    {'elements': ['83978887-be4c-4cac-afb6-0faad3d07c86'], 'events': []},\n",
       "    {'elements': ['28e829e2-030c-41ad-b9b9-b7a0e1a9b3e7'], 'events': []}]}],\n",
       " 'samplePlan': [{'sampleType': 'saliva',\n",
       "   'selectedCells': {'Arm_0': [True, False, True, False, True, True],\n",
       "    'Arm_2': [True, False, True, False, True, True],\n",
       "    'Arm_4': [True, False, True, False, True, True],\n",
       "    'Arm_3': [True, False, True, False, True, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [1, 1, 1, 1, 1, 3],\n",
       "    'Arm_2': [1, 1, 1, 1, 1, 3],\n",
       "    'Arm_4': [1, 1, 1, None, 1, 3],\n",
       "    'Arm_3': [1, 1, 1, 1, None, 3]}},\n",
       "  {'sampleType': 'blood',\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, True],\n",
       "    'Arm_2': [False, False, False, False, False, True],\n",
       "    'Arm_4': [False, False, False, False, False, True],\n",
       "    'Arm_3': [False, False, False, False, False, True]},\n",
       "   'sampleTypeSizes': {'Arm_0': [None, None, None, None, None, 1],\n",
       "    'Arm_2': [None, None, None, None, None, 1],\n",
       "    'Arm_4': [None, None, None, None, None, 1],\n",
       "    'Arm_3': [None, None, None, None, None, 1]}}],\n",
       " 'assayConfigs': [{'id': 0,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec.',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'green',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'mass spectrometry',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['polar fraction',\n",
       "        'nonpolar fraction'],\n",
       "       'values': ['polar fraction']},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['labelling', {'#replicates': {'value': 1}}],\n",
       "    ['labelled extract',\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': 'labelled extract type',\n",
       "      'characteristics_value': {'options': [], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['mass spectrometry',\n",
       "     {'#replicates': {'value': 2},\n",
       "      'instrument': {'options': ['Agilent QTQF 6510', 'Agilent QTQF 6520A'],\n",
       "       'values': ['Agilent QTQF 6510']},\n",
       "      'injection_mode': {'options': ['FIA', 'LC'], 'values': ['FIA', 'LC']},\n",
       "      'acquisition_mode': {'options': ['positive mode', 'negative mode'],\n",
       "       'values': ['positive mode']}}],\n",
       "    ['raw spectral data file',\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, True, False, True, True],\n",
       "    'Arm_2': [False, False, True, False, True, True],\n",
       "    'Arm_4': [False, False, True, False, True, True],\n",
       "    'Arm_3': [False, False, True, False, True, True]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 1,\n",
       "   'name': 'Metabolite Profiling - Mass. Spec. (Annotated)',\n",
       "   'icon': 'fa fa-dna',\n",
       "   'color': 'blue',\n",
       "   'measurement_type': {'term': 'metabolite profiling',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000366',\n",
       "    'source': None},\n",
       "   'technology_type': {'term': 'mass spectrometry',\n",
       "    'iri': 'http://purl.obolibrary.org/obo/OBI_0000470',\n",
       "    'source': None},\n",
       "   'workflow': [[{'term': 'extraction',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0302884',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'extract', 'iri': 'http://purl.obolibrary.org/obo/OBI_0000423'},\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': {'term': 'extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [{'term': 'polar fraction',\n",
       "         'iri': None,\n",
       "         'source': None},\n",
       "        {'term': 'lipids', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'labelling',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/CHMO_0001675',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1}}],\n",
       "    [{'term': 'labelled extract',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0000924',\n",
       "      'source': None},\n",
       "     {'node_type': 'labeled extract',\n",
       "      'characteristics_category': {'term': 'labelled extract type',\n",
       "       'iri': None,\n",
       "       'source': None},\n",
       "      'characteristics_value': {'options': [], 'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    [{'term': 'mass spectrometry',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/OBI_0200085',\n",
       "      'source': None},\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': [{'term': 'Agilent QTQF 6510',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000676',\n",
       "         'source': None},\n",
       "        {'term': 'Agilent QTQF 6520A',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000677',\n",
       "         'source': None}],\n",
       "       'values': []},\n",
       "      'injection_mode': {'options': [{'term': 'FIA',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1000058',\n",
       "         'source': None},\n",
       "        {'term': 'LC', 'iri': None, 'source': None}],\n",
       "       'values': []},\n",
       "      'acquisition_mode': {'options': [{'term': 'positive mode',\n",
       "         'iri': 'http://purl.obolibrary.org/obo/MS_1002807',\n",
       "         'source': None}],\n",
       "       'values': []}}],\n",
       "    [{'term': 'raw spectral data file',\n",
       "      'iri': 'http://purl.obolibrary.org/obo/MS_1003083',\n",
       "      'source': None},\n",
       "     {'node_type': 'data file',\n",
       "      'is_input_to_next_protocols': {'value': True}}]],\n",
       "   '@context': {'measurement_type': 'http://purl.obolibrary.org/obo/NCIT_C25209',\n",
       "    'technology_type': 'http://www.ebi.ac.uk/efo/EFO_0005521',\n",
       "    'instrument': 'http://purl.obolibrary.org/obo/MS_1000463',\n",
       "    'injection_mode': 'http://purl.obolibrary.org/obo/MS_1000007',\n",
       "    'acquisition_mode': 'http://purl.obolibrary.org/obo/MS_1000027'},\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}},\n",
       "  {'id': 2,\n",
       "   'name': 'Metabolite Profiling - NMR',\n",
       "   'icon': 'fa fa-atom',\n",
       "   'color': 'orange',\n",
       "   'measurement_type': 'metabolite profiling',\n",
       "   'technology_type': 'NMR spectroscopy',\n",
       "   'workflow': [['extraction', {'#replicates': {'value': 1}}],\n",
       "    ['extract',\n",
       "     {'node_type': 'extract',\n",
       "      'characteristics_category': 'extract type',\n",
       "      'characteristics_value': {'options': ['supernatant', 'pellet'],\n",
       "       'values': []},\n",
       "      'is_input_to_next_protocols': {'value': True}}],\n",
       "    ['nmr_spectroscopy',\n",
       "     {'#replicates': {'value': 1},\n",
       "      'instrument': {'options': ['Bruker AvanceII 1 GHz'], 'values': []},\n",
       "      'acquisition_mode': {'options': ['1D 13C NMR',\n",
       "        '1D 1H NMR',\n",
       "        '2D 13C-13C NMR'],\n",
       "       'values': []},\n",
       "      'pulse_sequence': {'options': ['CPMG', 'TOCSY', 'HOESY', 'watergate'],\n",
       "       'values': []}}],\n",
       "    ['raw_spectral_data_file',\n",
       "     {'node_type': 'data file',\n",
       "      'size': 1,\n",
       "      'is_input_to_next_protocols': {'value': False}}]],\n",
       "   'selectedCells': {'Arm_0': [False, False, False, False, False, False],\n",
       "    'Arm_2': [False, False, False, False, False, False],\n",
       "    'Arm_4': [False, False, False, False, False, False],\n",
       "    'Arm_3': [False, False, False, False, False, False]},\n",
       "   'selectedSampleTypes': {'Arm_0': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']],\n",
       "    'Arm_2': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_4': [['saliva'], [], ['saliva'], [], ['saliva'], ['saliva', 'blood']],\n",
       "    'Arm_3': [['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     [],\n",
       "     ['saliva'],\n",
       "     ['saliva', 'blood']]}}],\n",
       " 'selectedAssayTypes': {'Metabolite Profiling - Mass. Spec.': True,\n",
       "  'Metabolite Profiling - Mass. Spec. (Annotated)': False,\n",
       "  'Metabolite Profiling - NMR': False}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(os.path.abspath(os.path.join(\n",
    "    \"config\", \"study-design-3-repeated-treatment-datascriptor.json\"\n",
    ")), \"r\") as config_file:\n",
    "    study_design_config = json.load(config_file)\n",
    "study_design_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate the ISA Study Design from the JSON configuration\n",
    "To perform the conversion we just need to use the function `generate_isa_study_design_from_config` (name possibly subject to change, should we drop the \"isa\" and \"datascriptor\" qualifiers?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_design = generate_study_design_from_config(study_design_config)\n",
    "assert isinstance(study_design, StudyDesign)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Generate the ISA Study from the StudyDesign and embed it into an ISA Investigation\n",
    "\n",
    "The `StudyDesign.generate_isa_study()` method returns the complete ISA-API `Study` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the study design took 0.36 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "study = study_design.generate_isa_study()\n",
    "end = time()\n",
    "print('The generation of the study design took {:.2f} s.'.format(end - start))\n",
    "assert isinstance(study, Study)\n",
    "investigation = Investigation(studies=[study])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Serialize and save the JSON representation of the generated ISA Investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The JSON serialisation of the ISA investigation took 0.27 s.\n"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "inv_json = json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': '))\n",
    "end = time()\n",
    "print('The JSON serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = os.path.abspath(os.path.join('output'))\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "with open(os.path.abspath(os.path.join('output','isa-investigation-2-arms-nmr-ms.json')), 'w') as out_fp:\n",
    "    json.dump(json.loads(inv_json), out_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Dump the ISA Investigation to ISA-Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-08-20 17:32:00,757 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 270 paths!\n",
      "2020-08-20 17:32:00,773 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 270 paths\n",
      "2020-08-20 17:32:00,778 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 270 rows\n",
      "2020-08-20 17:32:01,036 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 60 paths!\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Protocol REF.mass spectrometry'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-d801917af766>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0misatab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minvestigation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabspath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'output'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'The Tab serialisation of the ISA investigation took {:.2f} s.'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mend\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mdump\u001b[0;34m(isa_obj, output_path, i_file_name, skip_dump_tables, write_factor_values_in_assay_table)\u001b[0m\n\u001b[1;32m   1046\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1047\u001b[0m         \u001b[0mwrite_study_table_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minvestigation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1048\u001b[0;31m         write_assay_table_files(\n\u001b[0m\u001b[1;32m   1049\u001b[0m             investigation, output_path, write_factor_values_in_assay_table)\n\u001b[1;32m   1050\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m in \u001b[0;36mwrite_assay_table_files\u001b[0;34m(inv_obj, output_dir, write_factor_values)\u001b[0m\n\u001b[1;32m   1484\u001b[0m                             \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1485\u001b[0m                         )\n\u001b[0;32m-> 1486\u001b[0;31m                         \u001b[0mdf_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0molabel\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1487\u001b[0m                         \u001b[0moname_label\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1488\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotocol_type\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Protocol REF.mass spectrometry'"
     ]
    }
   ],
   "source": [
    "start = time()\n",
    "isatab.dump(investigation, os.path.abspath(os.path.join('output')))\n",
    "end = time()\n",
    "print('The Tab serialisation of the ISA investigation took {:.2f} s.'.format(end - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/Users/massi/.pyenv/versions/3.8.2/envs/isa-create-notebook-3.8.2/src/isatools/isatools/isatab.py\u001b[0m(1486)\u001b[0;36mwrite_assay_table_files\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m   1484 \u001b[0;31m                            \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   1485 \u001b[0;31m                        )\n",
      "\u001b[0m\u001b[0;32m-> 1486 \u001b[0;31m                        \u001b[0mdf_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0molabel\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   1487 \u001b[0;31m                        \u001b[0moname_label\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   1488 \u001b[0;31m                        \u001b[0;32mif\u001b[0m \u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecutes_protocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotocol_type\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "%debug"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use them on the notebook we can also dump the tables to pandas DataFrames, using the `dump_tables_to_dataframes` function rather than dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes = isatab.dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataframes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Check the correctness of the ISA-Tab DataFrames \n",
    "\n",
    "We have 1 study file and 5 assay files:\n",
    "* 1 assay file for Mass. Spec. (treatmennt arm, third epoch: surgery)\n",
    "* 2 assay files for Mass. Spec. (both arms, fourth epoch: follow-up)\n",
    "* 2 assay files for NMR (both arms, fourth epoch: follow-up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study_01.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_treatment_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_control_3_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_treatment_3_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_treatment_2_ASSAY_GRAPH_000_metabolite-profiling_mass-spectrometry.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_CELL_control_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 8 subjects in the control arm and 24 samples have been collected (3 blood samples per subject during the follow-up epoch)\n",
    "\n",
    "We have 10 subjects in the control arm and 40 samples have been collected (3 blood samples per subject during the follow-up epoch and 1 liver sample per subject during the surgery epoch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 24 samples in the control arm (i.e. group)\n",
      "There are 40 samples in the treatment arm (i.e. group)\n"
     ]
    }
   ],
   "source": [
    "study_frame = dataframes['s_study_01.txt']\n",
    "count_control_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'control' in el)])\n",
    "count_treatment_samples = len(study_frame[study_frame['Source Name'].apply(lambda el: 'treatment' in el)])\n",
    "print(\"There are {} samples in the control arm (i.e. group)\".format(count_control_samples))\n",
    "print(\"There are {} samples in the treatment arm (i.e. group)\".format(count_treatment_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each control samples is fractioned and 2 labelled extracts are produced (i.e. biological replicates).\n",
    "\n",
    "```\n",
    "[\n",
    "    \"labelling\",\n",
    "    {\n",
    "        \"#replicates\": 2\n",
    "    }\n",
    "]\n",
    "```\n",
    "\n",
    "There are 2 possible combinations of Mass Spec Assay specified in our configuration template and 2 techincal replicate are produced for each combinations:\n",
    "\n",
    "```\n",
    "[\n",
    "    \"mass spectrometry\",\n",
    "    {\n",
    "        \"#replicates\": 2,\n",
    "        \"instrument\": [\n",
    "            \"Agilent QTQF 6510\"\n",
    "        ],\n",
    "        \"injection_mode\": [\n",
    "            \"FIA\",\n",
    "            \"LC\"\n",
    "        ],\n",
    "        \"acquisition_mode\": [\n",
    "            \"positive mode\"\n",
    "        ]\n",
    "    }\n",
    "]\n",
    "```\n",
    "\n",
    "Two output raw spectral files are produced as the result of reach run\n",
    "\n",
    "```\n",
    "[\n",
    "    \"raw spectral data file\",\n",
    "    [\n",
    "        {\n",
    "            \"node_type\": \"data file\",\n",
    "            \"size\": 2,\n",
    "            \"is_input_to_next_protocols\": false\n",
    "        }\n",
    "    ]\n",
    "]\n",
    "```\n",
    "\n",
    "As a total we expect\n",
    "\n",
    "$$ N_{rows} = (N_{subjects} \\times N_{samples}) \\times (N_{biorepl} \\times N_{combinations} \\times N_{techrepl}) = (8 \\times 3) \\times (2 \\times 2 \\times 2 \\times 2) = 24 \\times 16 = 384 $$\n",
    "\n",
    "Which we can verify for the mass spectrometry assay file of the control group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Performer</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Characteristics[extract type]</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Parameter Value[instrument]</th>\n",
       "      <th>Parameter Value[pulse_sequence]</th>\n",
       "      <th>Parameter Value[acquisition_mode]</th>\n",
       "      <th>Performer.1</th>\n",
       "      <th>Free Induction Decay Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_27-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_27-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_27-1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_27-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_27-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_27-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_27-1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>watergate</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_27-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_27-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>watergate</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_27-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_11-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_11-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_11-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>1D 13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_11-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_11-1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>watergate</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_11-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_11-18</td>\n",
       "      <td>pellet</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_11-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...</td>\n",
       "      <td>extraction</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>extract_11-1</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>nmr spectroscopy</td>\n",
       "      <td>Bruker AvanceII 1 GHz</td>\n",
       "      <td>CPMG</td>\n",
       "      <td>2D 13C-13C NMR</td>\n",
       "      <td>Ellipsis</td>\n",
       "      <td>raw spectral data file_11-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Sample Name Protocol REF Performer  \\\n",
       "0    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "1    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "2    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "3    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "4    GRP-treatment.SBJ-001.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "..                                                 ...          ...       ...   \n",
       "475  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "476  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "477  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "478  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "479  GRP-treatment.SBJ-010.CEL-CELL_treatment_3.SMP...   extraction  Ellipsis   \n",
       "\n",
       "      Extract Name Characteristics[extract type]    Protocol REF.1  \\\n",
       "0    extract_27-18                        pellet  nmr spectroscopy   \n",
       "1     extract_27-1                   supernatant  nmr spectroscopy   \n",
       "2    extract_27-18                        pellet  nmr spectroscopy   \n",
       "3     extract_27-1                   supernatant  nmr spectroscopy   \n",
       "4    extract_27-18                        pellet  nmr spectroscopy   \n",
       "..             ...                           ...               ...   \n",
       "475  extract_11-18                        pellet  nmr spectroscopy   \n",
       "476  extract_11-18                        pellet  nmr spectroscopy   \n",
       "477   extract_11-1                   supernatant  nmr spectroscopy   \n",
       "478  extract_11-18                        pellet  nmr spectroscopy   \n",
       "479   extract_11-1                   supernatant  nmr spectroscopy   \n",
       "\n",
       "    Parameter Value[instrument] Parameter Value[pulse_sequence]  \\\n",
       "0         Bruker AvanceII 1 GHz                            CPMG   \n",
       "1         Bruker AvanceII 1 GHz                            CPMG   \n",
       "2         Bruker AvanceII 1 GHz                            CPMG   \n",
       "3         Bruker AvanceII 1 GHz                       watergate   \n",
       "4         Bruker AvanceII 1 GHz                       watergate   \n",
       "..                          ...                             ...   \n",
       "475       Bruker AvanceII 1 GHz                            CPMG   \n",
       "476       Bruker AvanceII 1 GHz                            CPMG   \n",
       "477       Bruker AvanceII 1 GHz                       watergate   \n",
       "478       Bruker AvanceII 1 GHz                            CPMG   \n",
       "479       Bruker AvanceII 1 GHz                            CPMG   \n",
       "\n",
       "    Parameter Value[acquisition_mode] Performer.1  \\\n",
       "0                      2D 13C-13C NMR    Ellipsis   \n",
       "1                          1D 13C NMR    Ellipsis   \n",
       "2                          1D 13C NMR    Ellipsis   \n",
       "3                          1D 13C NMR    Ellipsis   \n",
       "4                          1D 13C NMR    Ellipsis   \n",
       "..                                ...         ...   \n",
       "475                    2D 13C-13C NMR    Ellipsis   \n",
       "476                        1D 13C NMR    Ellipsis   \n",
       "477                    2D 13C-13C NMR    Ellipsis   \n",
       "478                    2D 13C-13C NMR    Ellipsis   \n",
       "479                    2D 13C-13C NMR    Ellipsis   \n",
       "\n",
       "    Free Induction Decay Data File  \n",
       "0     raw spectral data file_27-26  \n",
       "1      raw spectral data file_27-3  \n",
       "2     raw spectral data file_27-20  \n",
       "3     raw spectral data file_27-11  \n",
       "4     raw spectral data file_27-30  \n",
       "..                             ...  \n",
       "475   raw spectral data file_11-24  \n",
       "476   raw spectral data file_11-22  \n",
       "477    raw spectral data file_11-7  \n",
       "478   raw spectral data file_11-26  \n",
       "479   raw spectral data file_11-15  \n",
       "\n",
       "[480 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_CELL_treatment_3_ASSAY_GRAPH_001_metabolite-profiling_NMR-spectroscopy.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
